<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Givanna Putri, Thomas O’Neil, John Salamon, Cathal King" />

<meta name="date" content="2024-11-04" />

<title>Setup</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">ASI Spatial Workshop 2024</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Analysis Files
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="setup.html">Setup</a>
    </li>
  </ul>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ghar1821/asi_spatial_workshop_2024">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Setup</h1>
<h4 class="author">Givanna Putri, Thomas O’Neil, John Salamon, Cathal
King</h4>
<h4 class="date">2024-11-04</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-11-15
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>asi_spatial_workshop_2024/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240925code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240925)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240925code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240925)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomghar1821asispatialworkshop2024tree6e2f089dd3ccffadbde411a9b31ee1db209cf7abtargetblank6e2f089a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/tree/6e2f089dd3ccffadbde411a9b31ee1db209cf7ab" target="_blank">6e2f089</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomghar1821asispatialworkshop2024tree6e2f089dd3ccffadbde411a9b31ee1db209cf7abtargetblank6e2f089a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/tree/6e2f089dd3ccffadbde411a9b31ee1db209cf7ab" target="_blank">6e2f089</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    data/.DS_Store
    Ignored:    data/hello.txt
    Ignored:    data/single_cell/
    Ignored:    data/visium_ffpe_human_prostate/
    Ignored:    data/visium_human_crc/
    Ignored:    output/.DS_Store
    Ignored:    visium/

Untracked files:
    Untracked:  analysis/visium_ffpe_prostate.Rmd.bk

Unstaged changes:
    Modified:   analysis/_site.yml
    Modified:   analysis/visium_crc.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/setup.Rmd</code>) and HTML
(<code>docs/setup.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/6e2f089dd3ccffadbde411a9b31ee1db209cf7ab/analysis/setup.Rmd" target="_blank">6e2f089</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-15
</td>
<td>
wflow_publish(c("analysis/setup.Rmd", "analysis/index.Rmd"))
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/202bc797d3a2d9d009ff513d14a861caa6c63995/analysis/setup.Rmd" target="_blank">202bc79</a>
</td>
<td>
John Salamon
</td>
<td>
2024-11-15
</td>
<td>
add dependencies
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/33c8fbf1e3e8b2cccb16aa7b8a08ecd5e4f78596/analysis/setup.Rmd" target="_blank">33c8fbf</a>
</td>
<td>
John Salamon
</td>
<td>
2024-11-15
</td>
<td>
add compressed datasets, update checksums
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/725c2bf2839e1d471ff0e4b10218cc865ecbce15/analysis/setup.Rmd" target="_blank">725c2bf</a>
</td>
<td>
GitHub
</td>
<td>
2024-11-15
</td>
<td>
Update setup.Rmd
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/d1d48408ddab085a7c03d17d02ad7a195c32c0c8/analysis/setup.Rmd" target="_blank">d1d4840</a>
</td>
<td>
GitHub
</td>
<td>
2024-11-15
</td>
<td>
Update setup.Rmd
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/ff8732713a63378165462b58514be9f125c2f2d1/analysis/setup.Rmd" target="_blank">ff87327</a>
</td>
<td>
GitHub
</td>
<td>
2024-11-15
</td>
<td>
Update setup.Rmd
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/4ef335fa502663f04247d27c5620df7cb6d9f746/analysis/setup.Rmd" target="_blank">4ef335f</a>
</td>
<td>
GitHub
</td>
<td>
2024-11-15
</td>
<td>
Update setup.Rmd
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/ad6a58b2eb1a42658a8d8161489d6861e751ec52/analysis/setup.Rmd" target="_blank">ad6a58b</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-15
</td>
<td>
update setup and homepage
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ghar1821/asi_spatial_workshop_2024/ad6a58b2eb1a42658a8d8161489d6861e751ec52/docs/setup.html" target="_blank">ad6a58b</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-15
</td>
<td>
update setup and homepage
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/5b7ca36572a1aa8952c61b168d52fdc08d7ba8e5/analysis/setup.Rmd" target="_blank">5b7ca36</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-15
</td>
<td>
update setup
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ghar1821/asi_spatial_workshop_2024/5b7ca36572a1aa8952c61b168d52fdc08d7ba8e5/docs/setup.html" target="_blank">5b7ca36</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-15
</td>
<td>
update setup
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/9ca58ba9a60f9cd243cae3c92fe87a1407fb1f6b/analysis/setup.Rmd" target="_blank">9ca58ba</a>
</td>
<td>
John Salamon
</td>
<td>
2024-11-13
</td>
<td>
add data files and libraries for stomics
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/0b3e0ccf039db3fd29a9050b32b8d63303058a52/analysis/setup.Rmd" target="_blank">0b3e0cc</a>
</td>
<td>
DrThomasOneil
</td>
<td>
2024-11-13
</td>
<td>
Update setup
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/051b760847b01faf5c93970eed5dbc49d28c741c/analysis/setup.Rmd" target="_blank">051b760</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-12
</td>
<td>
update visium crc workshop and setup page
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ghar1821/asi_spatial_workshop_2024/051b760847b01faf5c93970eed5dbc49d28c741c/docs/setup.html" target="_blank">051b760</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-12
</td>
<td>
update visium crc workshop and setup page
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/695bb87a23b45f1cf5314a89c683a73d2ab0e210/analysis/setup.Rmd" target="_blank">695bb87</a>
</td>
<td>
GitHub
</td>
<td>
2024-11-08
</td>
<td>
Update setup.Rmd
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/analysis/setup.Rmd" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
<td>
Add tutorials for visium
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ghar1821/asi_spatial_workshop_2024/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/setup.html" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
<td>
Add tutorials for visium
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>Please ensure you follow the instructions below
<strong>prior</strong> to attending the workshop.</p>
<p>You can either create a new R script and copy paste the content of
the code blocks or download <a
href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/master/analysis/setup.Rmd">this
Rmd file</a> and run each code block step by step.</p>
<p><strong><em>Do not just blindly run the code! Make sure you read the
description above each of the code block to understand what the code is
doing before running them!</em></strong></p>
<p>If you have any problems, please reach out to either:</p>
<ul>
<li><a href="mailto:felix.marsh-wakefield@sydney.edu.au">Felix
Marsh-Wakefield</a></li>
<li><a href="mailto:thomas.oneil@sydney.edu.au">Thomas O’Neil</a></li>
<li><a href="mailto:putri.g@wehi.edu.au">Givanna Putri</a></li>
<li><a href="mailto:cathal.king@sahmri.com">Cathal King</a></li>
</ul>
<div id="packages" class="section level2">
<h2>Packages</h2>
<p>Please install the packages that we will be using in the workshop by
running the code block below.</p>
<pre class="r"><code>cran_packages &lt;- c(
  &quot;BiocManager&quot;, &quot;Seurat&quot;, &quot;ggplot2&quot;, &quot;scales&quot;, &quot;patchwork&quot;, &quot;qs&quot;,
  &quot;hdf5r&quot;, &quot;viridis&quot;, &quot;devtools&quot;, &quot;cli&quot;, &quot;fs&quot;, &quot;sf&quot;, &quot;terra&quot;, &quot;scales&quot;,
  &quot;plotly&quot;, &quot;hexbin&quot;
)

install.packages(cran_packages)

bioc_packages &lt;- c(
  &quot;glmGamPoi&quot;, &quot;scater&quot;, &quot;scran&quot; &quot;SpatialExperiment&quot;, &quot;ggspavis&quot;, &quot;scran&quot;,
  &quot;nnSVG&quot;, &quot;SpatialFeatureExperiment&quot;, &quot;Voyager&quot;, &quot;RBioFormats&quot;, &quot;BiocParallel&quot;,
  &quot;bluster&quot;, &quot;EBImage&quot;
)

BiocManager::install(bioc_packages)

# RCTD only available via devtools install
devtools::install_github(&quot;dmcable/spacexr&quot;, build_vignettes = FALSE)</code></pre>
<p>Running this chunk will let you know if the packages have been
installed properly.</p>
<pre class="r"><code>checkSetup &lt;- function() {
  library(cli)
  cat(&quot;\n--------------------------------------\n&quot;)
  cat(style_bold(col_magenta(&quot;\n***Installing General Packages***\n\n&quot;)))
  not &lt;- c(); not2 &lt;- c()
  packages1 &lt;- c(cran_packages, bioc_packages)#, &quot;Test&quot;)
  for (i in 1:length(packages1)){
    if(requireNamespace(packages1[i], quietly = TRUE)==F) {
      cat(paste(style_bold(col_red(packages1[i])), &quot;has not been installed\n&quot;))
      not &lt;- c(not,i)
    } else {
      suppressWarnings(suppressMessages(library(as.character(packages1[i]), character.only = TRUE)))
      cat(col_yellow(packages1[i]), &quot;is loaded!\n&quot;)
    }
  }
  cat(&quot;\n--------------------------------------\n&quot;)
  if (length(not) &gt; 0){
    cat(style_bold(bg_red(&quot;\n  **IMPORTANT**  &quot;)),
        style_bold(col_yellow(&quot;\n\nYou need to install: \n&quot;)),
        paste(paste(c(packages1[not]), collapse=&quot;, &quot;)),
        &quot;\n\n--------------------------------------&quot;,
        &quot;\n\n Use:\n - install.packages(),\n - BiocManager::install() or, \n - use Google to find installation instructions.\n\n&quot;, style_bold(col_green(&quot;Then run this function again!\n\n&quot;)))
  } else {
    cat(&quot;&quot;,col_green(style_bold(&quot;\n All packages are loaded!\n\n Happy Coding! :)\n\n&quot;)))
  }
}
checkSetup()</code></pre>
</div>
<div id="folder-structure" class="section level2">
<h2>Folder structure</h2>
<p>The code block below will create a bunch of folders in the location
where this script is stored.</p>
<pre class="r"><code># This sets the working directory for all subsequent code chunks to be run
# to the location of this file.
knitr::opts_knit$set(root.dir = dirname(rstudioapi::getActiveDocumentContext()$path))

if(!dir.exists(&quot;visium&quot;)){dir.create(&quot;visium&quot;)}
if(!dir.exists(&quot;visium/raw&quot;)){dir.create(&quot;visium/raw&quot;)}
if(!dir.exists(&quot;visium/data&quot;)){dir.create(&quot;visium/data&quot;)}

if(!dir.exists(&quot;xenium&quot;)){dir.create(&quot;xenium&quot;)}
if(!dir.exists(&quot;xenium/raw&quot;)){dir.create(&quot;xenium/raw&quot;)}
if(!dir.exists(&quot;xenium/data&quot;)){dir.create(&quot;xenium/data&quot;)}

if(!dir.exists(&quot;stomics&quot;)){dir.create(&quot;stomics&quot;)}
if(!dir.exists(&quot;stomics/raw&quot;)){dir.create(&quot;stomics/raw&quot;)}
if(!dir.exists(&quot;stomics/data&quot;)){dir.create(&quot;stomics/data&quot;)}</code></pre>
</div>
<div id="data-download" class="section level2 tabset">
<h2 class="tabset">Data Download</h2>
<p>Please source the code block below to load a function to check the
integrity of the downloaded files.</p>
<pre class="r"><code>check_md5 &lt;- function(files_expected_md5sum) {
    # files_expected_md5sum must be a named vector where the name is the path to the file
    # and the value is the expected md5sum
    library(cli)
    library(tools)
    
    error &lt;- combine_ansi_styles(&quot;red&quot;, &quot;bold&quot;)
    
    need_redownload &lt;- FALSE
    for (filename in names(files_expected_md5sum)) {
        actual_md5sum &lt;- md5sum(filename)
        expected_md5sum &lt;- files_expected_md5sum[filename]
        if (actual_md5sum != expected_md5sum) {
            cat(error(paste(
                filename, &quot;is corrupted. Actual md5sum:&quot;, actual_md5sum, &quot;!= Expected md5sum:&quot;, expected_md5sum, &quot;\n&quot;
                )
            ))
            
            need_redownload &lt;- TRUE
        }
    }
    if (need_redownload) {
        cat(error(&quot;Some files are corrupted. Please redownload&quot;))
    }
}</code></pre>
<div id="visium" class="section level3">
<h3>Visium</h3>
<p>The code block below will download the count matrix and spatial
information required to load the visium data in.</p>
<pre class="r"><code>download.file(&quot;https://cf.10xgenomics.com/samples/spatial-exp/3.0.0/Visium_V2_Human_Colon_Cancer_P2/Visium_V2_Human_Colon_Cancer_P2_filtered_feature_bc_matrix.h5&quot;, 
              destfile = &quot;visium/raw/Visium_V2_Human_Colon_Cancer_P2_filtered_feature_bc_matrix.h5&quot;, 
              method = &quot;curl&quot;)


download.file(&quot;https://cf.10xgenomics.com/samples/spatial-exp/3.0.0/Visium_V2_Human_Colon_Cancer_P2/Visium_V2_Human_Colon_Cancer_P2_spatial.tar.gz&quot;, 
              destfile = &quot;visium/raw/spatial.tar.gz&quot;, 
              method = &quot;curl&quot;)</code></pre>
<p>The code below will check the integrity of the downloaded files. If
you get any red warning, please re-download the files that were found to
be corrupted using the code block above.</p>
<pre class="r"><code># check md5sum of the downloaded files.
md5sum_str &lt;- c(
    &quot;visium/raw/Visium_V2_Human_Colon_Cancer_P2_filtered_feature_bc_matrix.h5&quot; = &quot;0a9733cf0dfd3dfaa3b6c1c5c24bcbed&quot;,
    &quot;visium/raw/spatial.tar.gz&quot; = &quot;a32edc825ab7089fca7848fc80cd5113&quot;
)
check_md5(md5sum_str)</code></pre>
<p>If there are no warning from the code block above, proceed to untar
the <code>spatial.tar.gz</code> file using the code block below.</p>
<pre class="r"><code>untar(&quot;visium/raw/spatial.tar.gz&quot;, exdir =&quot;visium/raw&quot;)</code></pre>
<div id="optional-loupe-browser-file" class="section level4">
<h4>Optional loupe browser file</h4>
<p>The following file is optional and is only required if you would like
to try out the demo we will be showing on how to use loupe browser to
make some QC step simpler.</p>
<pre class="r"><code>download.file(
    &quot;https://cf.10xgenomics.com/samples/spatial-exp/3.0.0/Visium_V2_Human_Colon_Cancer_P2/Visium_V2_Human_Colon_Cancer_P2_cloupe.cloupe&quot;,
    destfile = &quot;visium/raw/Visium_V2_Human_Colon_Cancer_P2_cloupe.cloupe&quot;,
    method = &quot;curl&quot;
)
md5sum_str &lt;- c(
    &quot;visium/raw/Visium_V2_Human_Colon_Cancer_P2_cloupe.cloupe&quot; = &quot;c5726e203c21a2fc3432831ea7c74252&quot;
)
check_md5(md5sum_str)</code></pre>
<p>Install Loupe Browser at this link: <a
href="https://www.10xgenomics.com/support/software/loupe-browser/latest"
class="uri">https://www.10xgenomics.com/support/software/loupe-browser/latest</a></p>
</div>
</div>
<div id="stomics" class="section level3">
<h3>STOmics</h3>
<pre class="r"><code># LISTA paper, processed tissue slice (222MB)
download.file(&quot;https://filesender.aarnet.edu.au/download.php?token=ec632836-0700-4921-9699-668bfe9657d2&amp;files_ids=23608389&quot;,
              destfile = &quot;stomics/raw/DY1_D0_stereo-seq.h5ad&quot;)

# LISTA single cell dataset (540MB)
download.file(&quot;https://filesender.aarnet.edu.au/download.php?token=ec632836-0700-4921-9699-668bfe9657d2&amp;files_ids=23608386&quot;,
              destfile = &quot;stomics/raw/Homeostasis_hepatic_cell_scRNAseq.h5ad&quot;)</code></pre>
<pre class="r"><code># check md5sum of the downloaded files.
md5sum_str &lt;- c(
    &quot;stomics/raw/DY1_D0_stereo-seq.h5ad&quot; = &quot;f63df815d75548034cb659bc3491f51a&quot;,
    &quot;stomics/raw/Homeostasis_hepatic_cell_scRNAseq.h5ad&quot; = &quot;b1a7fd3ffa58e0420083fa8651cb7fec&quot;
)
check_md5(md5sum_str)</code></pre>
<div id="optional-raw-data-download-from-saw" class="section level4">
<h4>OPTIONAL raw data download from SAW</h4>
<p>File size: 4.6GB</p>
<pre class="r"><code># Raw data from SAW (4.6GB, optional download!)
download.file(&quot;https://filesender.aarnet.edu.au/download.php?token=32ffcb63-24df-47e2-bdce-7341fe86082a&amp;files_ids=23595450&quot;,
              destfile = &quot;stomics/raw/visualization.tar.gz&quot;)

# check md5sum of the downloaded files.
md5sum_str &lt;- c(
    &quot;stomics/raw/visualization.tar.gz&quot; = &quot;7d3daa224de7ca1244711531270546d3&quot;
)
check_md5(md5sum_str)</code></pre>
<p>If there are no warning from the code block above, proceed to untar
the <code>visualization.tar.gz</code> file using the code block
below.</p>
<pre class="r"><code>untar(&quot;stomics/raw/visualization.tar.gz&quot;, exdir =&quot;stomics/raw&quot;)</code></pre>
</div>
</div>
<div id="xenium" class="section level3">
<h3>Xenium</h3>
<p>Install Xenium Explorer v3.1.1 at this link: <a
href="https://www.10xgenomics.com/support/software/xenium-explorer/downloads"
class="uri">https://www.10xgenomics.com/support/software/xenium-explorer/downloads</a></p>
<p>Please also install the raw xenium data by using the below code. The
file size is large so it might take a few minutes to complete.</p>
<pre class="r"><code># Sample P5 CRC
# This file is ~12GB so will take a few minutes to download.
download.file(
  url = &quot;https://cf.10xgenomics.com/samples/xenium/2.0.0/Xenium_V1_Human_Colon_Cancer_P5_CRC_Add_on_FFPE/Xenium_V1_Human_Colon_Cancer_P5_CRC_Add_on_FFPE_outs.zip&quot;, 
  destfile = &quot;xenium/raw/Xenium_V1_Human_Colon_Cancer_P5_CRC_Add_on_FFPE_outs.zip&quot;, 
  method = &quot;curl&quot;
)

# check md5sum of the downloaded files.
md5sum_str &lt;- c(
    &quot;xenium/raw/Xenium_V1_Human_Colon_Cancer_P5_CRC_Add_on_FFPE_outs.zip&quot; = &quot;1069cb6b67a596e0011a4b3f5d3b7c08&quot;
)
check_md5(md5sum_str)</code></pre>
<p>Make sure there are no warning from the code block above. If there
are, please redownload the file by re-running <code>download.file</code>
command.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Australia/Melbourne
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] vctrs_0.6.5       httr_1.4.7        cli_3.6.3         knitr_1.48       
 [5] rlang_1.1.4       xfun_0.47         stringi_1.8.4     processx_3.8.4   
 [9] promises_1.3.0    jsonlite_1.8.9    glue_1.7.0        rprojroot_2.0.4  
[13] git2r_0.33.0      htmltools_0.5.8.1 httpuv_1.6.15     ps_1.8.0         
[17] sass_0.4.9        fansi_1.0.6       rmarkdown_2.28    jquerylib_0.1.4  
[21] tibble_3.2.1      evaluate_1.0.0    fastmap_1.2.0     yaml_2.3.10      
[25] lifecycle_1.0.4   whisker_0.4.1     stringr_1.5.1     compiler_4.4.1   
[29] fs_1.6.4          pkgconfig_2.0.3   Rcpp_1.0.13       rstudioapi_0.16.0
[33] later_1.3.2       digest_0.6.37     R6_2.5.1          utf8_1.2.4       
[37] pillar_1.9.0      callr_3.7.6       magrittr_2.0.3    bslib_0.8.0      
[41] tools_4.4.1       cachem_1.1.0      getPass_0.2-4    </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
