<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="John Salamon" />

<meta name="date" content="2024-11-20" />

<title>STOmics analysis</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">ASI Spatial Workshop 2024</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Analysis Files
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="setup.html">Setup</a>
    </li>
  </ul>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ghar1821/asi_spatial_workshop_2024">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">STOmics analysis</h1>
<h4 class="author">John Salamon</h4>
<h4 class="date">2024-11-20</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-11-20
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>asi_spatial_workshop_2024/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of
the R Markdown file created these results, you’ll want to first commit
it to the Git repo. If you’re still working on the analysis, you can
ignore this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240925code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240925)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240925code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240925)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomghar1821asispatialworkshop2024treed159369db7afd7d2b68b6afd118f5eba5dc4dbaetargetblankd159369a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/tree/d159369db7afd7d2b68b6afd118f5eba5dc4dbae" target="_blank">d159369</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomghar1821asispatialworkshop2024treed159369db7afd7d2b68b6afd118f5eba5dc4dbaetargetblankd159369a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/tree/d159369db7afd7d2b68b6afd118f5eba5dc4dbae" target="_blank">d159369</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    xenium/

Untracked files:
    Untracked:  stomics/
    Untracked:  visium/

Unstaged changes:
    Modified:   analysis/index.Rmd
    Modified:   analysis/stomics_01_intro.Rmd
    Modified:   analysis/stomics_02_basics.Rmd
    Modified:   analysis/stomics_03_analysis.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/stomics_03_analysis.Rmd</code>)
and HTML (<code>docs/stomics_03_analysis.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/69bd8e961402e283331e8fb67d151b6e9f589c32/analysis/stomics_03_analysis.Rmd" target="_blank">69bd8e9</a>
</td>
<td>
John Salamon
</td>
<td>
2024-11-19
</td>
<td>
show some of the immune cell annotations
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/818a4e90cdf0df2cef9767e7cfe678afa408c38a/analysis/stomics_03_analysis.Rmd" target="_blank">818a4e9</a>
</td>
<td>
John Salamon
</td>
<td>
2024-11-18
</td>
<td>
changes to make html more readable
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/e9213d994393dd19c224e3de9c1067ef73446e1c/analysis/stomics_03_analysis.Rmd" target="_blank">e9213d9</a>
</td>
<td>
John Salamon
</td>
<td>
2024-11-18
</td>
<td>
Merge stomics commits
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/0a0195c9daf56f96dfa6df818a6674bb21d19098/analysis/stomics_03_analysis.Rmd" target="_blank">0a0195c</a>
</td>
<td>
John Salamon
</td>
<td>
2024-11-18
</td>
<td>
more details
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/ac0b4d3fbac281bf71c398d91a8f12b59b0703ce/analysis/stomics_03_analysis.Rmd" target="_blank">ac0b4d3</a>
</td>
<td>
John Salamon
</td>
<td>
2024-11-18
</td>
<td>
add more comments
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/534489df238c52cfd3630a9c7133f40359025db8/analysis/stomics_03_analysis.Rmd" target="_blank">534489d</a>
</td>
<td>
John Salamon
</td>
<td>
2024-11-18
</td>
<td>
add in some content for RCTD
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/ff451dc02d76584c47614fe9aec8b3d04d1b8c81/analysis/stomics_03_analysis.Rmd" target="_blank">ff451dc</a>
</td>
<td>
John Salamon
</td>
<td>
2024-11-17
</td>
<td>
updates
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/997618e43a1d359b4b30ac99cf5f5d0b65569784/analysis/stomics_03_analysis.Rmd" target="_blank">997618e</a>
</td>
<td>
John Salamon
</td>
<td>
2024-11-16
</td>
<td>
making some adjustments
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<pre class="r"><code># For RCTD
library(spacexr)

# spatial experiment
suppressMessages(library(SpatialExperiment))
library(SpatialExperiment)

# Easier plotting with spe
library(ggspavis)</code></pre>
<pre><code>Loading required package: ggplot2</code></pre>
<pre><code>Registered S3 method overwritten by &#39;ggside&#39;:
  method from   
  +.gg   ggplot2</code></pre>
<pre class="r"><code># Custom functions
source(&quot;code/stomicsFunctions.R&quot;)</code></pre>
<pre><code>
Attaching package: &#39;Matrix&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:S4Vectors&#39;:

    expand</code></pre>
<div id="lista-data-dy1_d0_stereo-seq.h5ad" class="section level1">
<h1>LISTA data (DY1_D0_stereo-seq.h5ad)</h1>
<div id="spatial-area" class="section level2">
<h2>Spatial area</h2>
<p>DY1_D0 represents <span class="math inline">\(48.143
mm^2\)</span></p>
<p>From the paper: each bin contains <span
class="math inline">\(50x50\)</span> nanoballs.</p>
<p>Center-to-center distance of <span class="math inline">\(715
nm\)</span> between nanoballs, aka bin-resolution of <span
class="math inline">\(36x36 um\)</span>.</p>
<p>Liver cells are larger, so this is probably capturing close to a
single-cell diameter.</p>
</div>
<div id="experimental-design" class="section level2">
<h2>Experimental design</h2>
<ul>
<li>In total, 30 whole liver lobe sections</li>
<li>Time series (Day 0,1,2,3,7)</li>
<li>Matched scRNA-seq.</li>
</ul>
<p>This experiment has matched and annotated single cell data. We’ll be
using this to generate Figure 1A from the paper using robust cell-type
decomposition (RCTD).</p>
</div>
<div id="loading-the-lista-dataset" class="section level2">
<h2>Loading the LISTA dataset</h2>
<p>As they are stored as AnnData, we can import the files from this
experiment in the same way we would files directly output from SAW.</p>
<pre class="r"><code># SPE
h5ad_file &lt;- &quot;stomics/raw/DY1_D0_stereo-seq.h5ad&quot;

# Read in the matrix data (CSR format)
matrix_data &lt;- h5adMatrixLoad(h5ad_file)</code></pre>
<pre><code>Matrix is in CSR format, performing conversion</code></pre>
<pre class="r"><code>spatial_coords &lt;- h5read(h5ad_file, &quot;/obsm/spatial&quot;)

rowdata &lt;- anndataDataframe(h5ad_file, &quot;var&quot;)
coldata &lt;- anndataDataframe(h5ad_file, &quot;obs&quot;)


spatial_coords &lt;- data.matrix(data.frame(
  x=spatial_coords[1,],
  y=spatial_coords[2,],
  row.names = rownames(coldata)
))

spe &lt;- SpatialExperiment(
  assays = list(counts = matrix_data),
  colData = coldata, # spatial bins
  rowData = rowdata, # features
  spatialCoords = spatial_coords
)

spe</code></pre>
<pre><code>class: SpatialExperiment 
dim: 29075 37669 
metadata(0):
assays(1): counts
rownames(29075): 0610005C13Rik 0610006L08Rik ... Gm28487 Gm28679
rowData names(0):
colnames(37669): DY1_D0_249_210 DY1_D0_147_141 ... DY1_D0_122_149
  DY1_D0_199_70
colData names(2): annotation sample_id
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
spatialCoords names(2) : x y
imgData names(0):</code></pre>
<p>We can quickly plot this to verify it looks correct using
<code>plot</code>. <code>ggplot</code> is much more powerful but
sometimes we just want to verify the data is there and more or less
looks correct.</p>
<pre class="r"><code># R&#39;s default plot function is actually pretty smart
# but ggplot gives more functionality
plot(spatialCoords(spe))</code></pre>
<p><img src="figure/stomics_03_analysis.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" />
We can flip the y axis and set the aspect ratio correctly, for a
start.</p>
<pre class="r"><code># flip y axis, and add colours
ggplot(spatialCoords(spe), aes(x, y)) +
  geom_point(size = 0.5) +
  theme(aspect.ratio=1) +
  scale_y_reverse()</code></pre>
<p><img src="figure/stomics_03_analysis.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="examing-the-annotation" class="section level2">
<h2>Examing the annotation</h2>
<p>A pre-set list of genes was used to define zonation layers in this
tissue, we’ve already put this annotation in our SPE.</p>
<pre class="r"><code># Plot the current annotation
# These were created using AddModuleScore in Seurat with lists of genes
# https://github.com/haoshijie13/LISTA/blob/main/00.cut_zonation_layer_and_pathway_module_score.R

# ggspavis plotSpots can be useful
# plotSpots(spe, annotate = &quot;annotation&quot;, in_tissue = NULL) +
  
ggplot(spatialCoords(spe), aes(x, y, color = spe$annotation)) +
  geom_point() +
  scale_color_viridis_d() +
  theme(aspect.ratio=1) +
  scale_y_reverse()</code></pre>
<p><img src="figure/stomics_03_analysis.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Plotting each point can actually obscure the real picture if many
points are overlapping. We could reduce the point size, or simply bin
our data further. A good option for binning spatial data is to use
<code>hexbin</code>:</p>
<pre class="r"><code>ggplot(spatialCoords(spe), aes(x, y)) +
  geom_hex(bins=80, mapping = aes(fill = spe$annotation)) +
  scale_fill_viridis_d() +
  theme(aspect.ratio=1) +
  scale_y_reverse()</code></pre>
<p><img src="figure/stomics_03_analysis.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="rctd" class="section level2">
<h2>RCTD</h2>
<p>Robust cell type decomposition (<a
href="https://www.nature.com/articles/s41587-021-00830-w">RCTD</a>) can
be used to determine the likely cell types comprising each bin of our
spatial dataset.</p>
<div id="loading-single-cell-reference" class="section level3">
<h3>Loading single cell reference</h3>
<p>The reference data for LISTA are also stored in AnnData format, which
we can store in a SingleCellExperiment object.</p>
<pre class="r"><code># examine the file
# h5ls(sc_reference)
# str(h5read(sc_reference, &quot;var&quot;))

sc_reference &lt;- &quot;stomics/raw/Homeostasis_hepatic_cell_scRNAseq.h5ad&quot;

# Load the raw data
sc_matrix_data &lt;- h5adMatrixLoad(sc_reference, group = &quot;/raw/X&quot;)</code></pre>
<pre><code>Matrix is in CSR format, performing conversion</code></pre>
<pre class="r"><code>obs &lt;- anndataDataframe(sc_reference, &quot;obs&quot;)
var &lt;-anndataDataframe(sc_reference, &quot;var&quot;)</code></pre>
<pre><code>Warning in anndataDataframe(sc_reference, &quot;var&quot;): No _index found in data,
using sequential row names</code></pre>
<pre class="r"><code># make sure col and rownames are set
rownames(sc_matrix_data) &lt;- var$gene
colnames(sc_matrix_data) &lt;- rownames(obs)

# Construct SCE
sce &lt;- SingleCellExperiment(
  assays = list(counts = sc_matrix_data),
  colData = obs, # spatial bins
  rowData = var # features
)

sce</code></pre>
<pre><code>class: SingleCellExperiment 
dim: 34945 51275 
metadata(0):
assays(1): counts
rownames(34945): Cox7a2 Syf2 ... AC168090.1 Gm26742
rowData names(2): features gene
colnames(51275): D0_H_M1_batch1_CELL516_N2_1 D0_H_M1_batch1_CELL81_N2_1
  ... TTTGTCATCCCATTAT-1_3 TTTGTCATCCTATTCA-1_3
colData names(7): Library annotation ... orig.ident pct_counts_mt
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
<p>We can see how many cells we have per annotation:</p>
<pre class="r"><code>table(sce$annotation)</code></pre>
<pre><code>
           BC       Cd4+ TC       Cd8+ TC     Cholangio            DC 
         4695          6926           959           462           876 
           FB           HSC       Kupffer          LSEC          LVEC 
          163          4721          1363          9654          1640 
         Mono            NK           NKT Naive Cd4+ TC Naive Cd8+ TC 
          879          1112          1406          1225          1269 
        Neutr        PC hep        PP hep 
          214          6851          6860 </code></pre>
<p>Because this will take quite some time with our full dataset, we’ll
actually just take a random subsample of cells for this example to speed
things up.</p>
<pre class="r"><code># Make sure we get a usable seed...
set.seed(123)

# Take a random subsample
sce &lt;- sce[ ,sample(dim(sce)[2], 8000)]</code></pre>
<p>Finally, we can create our reference for RCTD.</p>
<pre class="r"><code># Extract single cell annotations
cell_types &lt;- sce$annotation
names(cell_types) &lt;- colnames(sce)

# Create RCTD object
reference &lt;- Reference(assays(sce)$counts, cell_types)</code></pre>
</div>
<div id="preparing-spatial-data" class="section level3">
<h3>Preparing spatial data</h3>
<p>First, we’ll also subset our spatial data for efficiency. We don’t
want to just randomly subsample this however, as we’ll end up with a
very strangle looking dataset. Instead, let’s select a particular subset
based on our xy coordinates.</p>
<pre class="r"><code># Let&#39;s select a little circle from the middle of the tissue
midpoint &lt;- as.integer(colMeans(spatialCoords(spe)))

# center with midpoint around origin
adjusted_coords &lt;- sweep(spatialCoords(spe), 2, midpoint, &quot;-&quot;)

# define some regions
spe$circle &lt;- sqrt(rowSums((adjusted_coords)^2)) &lt; 40 # A circle, 50 pts in diameter
spe$square &lt;- (rowSums(abs(adjusted_coords) &lt; 50)) == 2 # A square

# We can plot more easily if we have x and y in coldata..
spe$x_pos &lt;- spatialCoords(spe)[,1]
spe$y_pos &lt;- spatialCoords(spe)[,2]

# Take the circle
spe_circle &lt;- spe[, spe$circle]

# Plot...
ggplot(colData(spe_circle), aes(x_pos, y_pos)) +
  geom_hex(bins=30, mapping = aes(fill = annotation)) +
  scale_fill_viridis_d() +
  theme(aspect.ratio=1) +
  scale_y_reverse()</code></pre>
<p><img src="figure/stomics_03_analysis.Rmd/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Now that we have our subset, we can set up the</p>
<pre class="r"><code>coords &lt;- spatialCoords(spe_circle)
rownames(coords) &lt;- colnames(spe_circle)
puck &lt;- SpatialRNA(data.frame(coords), assays(spe_circle)$counts)</code></pre>
<p>Now, let’s try running RCTD. I’ve commented this out for now because
it will take a very long time to run.</p>
<pre class="r"><code># Prepare for RCTD analysis
# The max_cores thing seems broken, it just uses total n cores regardless
RCTD &lt;- create.RCTD(puck, reference, max_cores = 1)</code></pre>
<pre><code>Begin: process_cell_type_info</code></pre>
<pre><code>process_cell_type_info: number of cells in reference: 8000</code></pre>
<pre><code>process_cell_type_info: number of genes in reference: 34945</code></pre>
<pre><code>
           BC       Cd4+ TC       Cd8+ TC     Cholangio            DC 
          699          1127           159            77           123 
           FB           HSC       Kupffer          LSEC          LVEC 
           29           734           204          1500           219 
         Mono            NK           NKT Naive Cd4+ TC Naive Cd8+ TC 
          142           192           219           185           191 
        Neutr        PC hep        PP hep 
           26          1070          1104 </code></pre>
<pre><code>End: process_cell_type_info</code></pre>
<pre><code>create.RCTD: getting regression differentially expressed genes: </code></pre>
<pre><code>get_de_genes: BC found DE genes: 174</code></pre>
<pre><code>get_de_genes: Cd4+ TC found DE genes: 139</code></pre>
<pre><code>get_de_genes: Cd8+ TC found DE genes: 106</code></pre>
<pre><code>get_de_genes: Cholangio found DE genes: 215</code></pre>
<pre><code>get_de_genes: DC found DE genes: 248</code></pre>
<pre><code>get_de_genes: FB found DE genes: 315</code></pre>
<pre><code>get_de_genes: HSC found DE genes: 345</code></pre>
<pre><code>get_de_genes: Kupffer found DE genes: 336</code></pre>
<pre><code>get_de_genes: LSEC found DE genes: 328</code></pre>
<pre><code>get_de_genes: LVEC found DE genes: 334</code></pre>
<pre><code>get_de_genes: Mono found DE genes: 145</code></pre>
<pre><code>get_de_genes: NK found DE genes: 181</code></pre>
<pre><code>get_de_genes: NKT found DE genes: 142</code></pre>
<pre><code>get_de_genes: Naive Cd4+ TC found DE genes: 135</code></pre>
<pre><code>get_de_genes: Naive Cd8+ TC found DE genes: 127</code></pre>
<pre><code>get_de_genes: Neutr found DE genes: 521</code></pre>
<pre><code>get_de_genes: PC hep found DE genes: 458</code></pre>
<pre><code>get_de_genes: PP hep found DE genes: 394</code></pre>
<pre><code>get_de_genes: total DE genes: 2598</code></pre>
<pre><code>create.RCTD: getting platform effect normalization differentially expressed genes: </code></pre>
<pre><code>get_de_genes: BC found DE genes: 439</code></pre>
<pre><code>get_de_genes: Cd4+ TC found DE genes: 386</code></pre>
<pre><code>get_de_genes: Cd8+ TC found DE genes: 247</code></pre>
<pre><code>get_de_genes: Cholangio found DE genes: 466</code></pre>
<pre><code>get_de_genes: DC found DE genes: 526</code></pre>
<pre><code>get_de_genes: FB found DE genes: 659</code></pre>
<pre><code>get_de_genes: HSC found DE genes: 626</code></pre>
<pre><code>get_de_genes: Kupffer found DE genes: 670</code></pre>
<pre><code>get_de_genes: LSEC found DE genes: 648</code></pre>
<pre><code>get_de_genes: LVEC found DE genes: 638</code></pre>
<pre><code>get_de_genes: Mono found DE genes: 338</code></pre>
<pre><code>get_de_genes: NK found DE genes: 557</code></pre>
<pre><code>get_de_genes: NKT found DE genes: 378</code></pre>
<pre><code>get_de_genes: Naive Cd4+ TC found DE genes: 332</code></pre>
<pre><code>get_de_genes: Naive Cd8+ TC found DE genes: 279</code></pre>
<pre><code>get_de_genes: Neutr found DE genes: 877</code></pre>
<pre><code>get_de_genes: PC hep found DE genes: 672</code></pre>
<pre><code>get_de_genes: PP hep found DE genes: 620</code></pre>
<pre><code>get_de_genes: total DE genes: 4399</code></pre>
<pre class="r"><code># Multi mode allows us to define &gt;2 cell types per tissue
# RCTD &lt;- run.RCTD(RCTD, doublet_mode = &#39;multi&#39;)

# Save the output as it can take a while to run
# saveRDS(RCTD, &quot;stomics/data/RCTD.rds&quot;)</code></pre>
<p>We’ll load our RDS file in case we weren’t able to run this (or we
don’t want to wait):</p>
<pre class="r"><code>RCTD &lt;- readRDS(&quot;stomics/data/RCTD.rds&quot;)</code></pre>
<p>Let’s inspect the results.</p>
<p>Each spot will have a list of identified cell types. It’ll also have
a list of identified cells. But we can extract the weights assigned to
every cell type in the reference and plot that.</p>
<pre class="r"><code># We can extract &quot;all weights&quot; to see the score assigned per type
celltype_list &lt;- lapply(RCTD@results, \(x) x$all_weights)
celltype_df &lt;- as.data.frame(do.call(rbind, celltype_list))
celltype_df &lt;- normalize_weights(celltype_df)
rownames(celltype_df) &lt;- colnames(RCTD@spatialRNA@counts)

boxplot(celltype_df)</code></pre>
<p><img src="figure/stomics_03_analysis.Rmd/unnamed-chunk-15-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The main cell types assigned are “PC hep” (pericentral hepatocyte)
and “PP hep” (periportal hepatocyte). If we were to only take the most
highly weighted result, this is all we’d see.</p>
<p>The liver cells are on average much larger, and perhaps expression is
higher. If we want to highlight these immune cells, we’ll need to treat
them differently.</p>
<pre class="r"><code># Get a closer look at the other cell type scores
boxplot(celltype_df[, !(names(celltype_df) %in% c(&quot;PC hep&quot;, &quot;PP hep&quot;))])</code></pre>
<p><img src="figure/stomics_03_analysis.Rmd/unnamed-chunk-16-1.png" width="672" style="display: block; margin: auto;" />
Now we can see the liver sinusoidal endothelial cells (LSECs), and a
small number of immune cells. Let’s extract all of these results so we
can plot them more easily.</p>
<pre class="r"><code># Define a dataframe with the correct number of rows
df &lt;- data.frame(row.names = colnames(spe_circle))
# Set the results
df[rownames(celltype_df),colnames(celltype_df)] &lt;- celltype_df
dim(df)</code></pre>
<pre><code>[1] 5009   18</code></pre>
<p>We’ll reshape this into “long” format.</p>
<pre class="r"><code># Reshape into long format for plotting
df_long &lt;- data.frame(
  variable = rep(names(df), each = nrow(df)),
  value = unlist(df),
  x_pos = rep(colData(spe_circle)$x_pos, times = ncol(df)),
  y_pos = rep(colData(spe_circle)$y_pos, times = ncol(df))
)</code></pre>
<p>We can see that our main two cell types correspond well to the
zonation layers that have already been annotated:</p>
<pre class="r"><code># Plot the PC hep and PP hep weights
df_long[df_long$variable %in% c(&quot;PC hep&quot;, &quot;PP hep&quot;),] |&gt; 
  ggplot(aes(x_pos, y_pos, color=value)) +
  geom_point(size=1.0) +
  scale_color_viridis_c() +
  theme(aspect.ratio=1) +
  scale_y_reverse() +
  facet_wrap(~variable)</code></pre>
<p><img src="figure/stomics_03_analysis.Rmd/unnamed-chunk-19-1.png" width="672" style="display: block; margin: auto;" />
Let’s exclude these two, and just see the distribution of other cell
types, to identify some specific immune cell hotspots.</p>
<pre class="r"><code># Plot weights of each cell type
df_long[!df_long$variable %in% c(&quot;PC hep&quot;, &quot;PP hep&quot;),] |&gt; 
  ggplot(aes(x_pos, y_pos, color=value)) +
  geom_point(size=1.0) +
  scale_color_viridis_c() +
  theme(aspect.ratio=1) +
  scale_y_reverse() +
  facet_wrap(~variable)</code></pre>
<p><img src="figure/stomics_03_analysis.Rmd/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="extra-steps" class="section level2">
<h2>Extra steps</h2>
<p>In the remaining time, perhaps we could do:</p>
<ul>
<li>Go into some depth with clustering? (PCA/UMAP/kmeans)</li>
<li>Define some more spatially variable genes?</li>
</ul>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.2 (2024-10-31 ucrt)
Platform: x86_64-w64-mingw32/x64
Running under: Windows 11 x64 (build 22631)

Matrix products: default


locale:
[1] LC_COLLATE=English_Australia.utf8  LC_CTYPE=English_Australia.utf8   
[3] LC_MONETARY=English_Australia.utf8 LC_NUMERIC=C                      
[5] LC_TIME=English_Australia.utf8    

time zone: Australia/Adelaide
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] Matrix_1.7-1                rhdf5_2.50.0               
 [3] ggspavis_1.12.0             ggplot2_3.5.1              
 [5] SpatialExperiment_1.16.0    SingleCellExperiment_1.28.1
 [7] SummarizedExperiment_1.36.0 Biobase_2.66.0             
 [9] GenomicRanges_1.58.0        GenomeInfoDb_1.42.0        
[11] IRanges_2.40.0              S4Vectors_0.44.0           
[13] BiocGenerics_0.52.0         MatrixGenerics_1.18.0      
[15] matrixStats_1.4.1           spacexr_2.2.1              
[17] workflowr_1.7.1            

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.1        viridisLite_0.4.2       dplyr_1.1.4            
 [4] farver_2.1.2            fastmap_1.2.0           promises_1.3.0         
 [7] digest_0.6.37           lifecycle_1.0.4         processx_3.8.4         
[10] magrittr_2.0.3          compiler_4.4.2          rlang_1.1.4            
[13] sass_0.4.9              tools_4.4.2             utf8_1.2.4             
[16] yaml_2.3.10             knitr_1.49              S4Arrays_1.6.0         
[19] labeling_0.4.3          DelayedArray_0.32.0     RColorBrewer_1.1-3     
[22] abind_1.4-8             withr_3.0.2             grid_4.4.2             
[25] fansi_1.0.6             git2r_0.35.0            colorspace_2.1-1       
[28] Rhdf5lib_1.28.0         scales_1.3.0            iterators_1.0.14       
[31] cli_3.6.3               rmarkdown_2.29          crayon_1.5.3           
[34] generics_0.1.3          rstudioapi_0.17.1       httr_1.4.7             
[37] rjson_0.2.23            cachem_1.1.0            stringr_1.5.1          
[40] zlibbioc_1.52.0         parallel_4.4.2          XVector_0.46.0         
[43] vctrs_0.6.5             jsonlite_1.8.9          callr_3.7.6            
[46] ggrepel_0.9.6           magick_2.8.5            foreach_1.5.2          
[49] hexbin_1.28.5           jquerylib_0.1.4         glue_1.8.0             
[52] ggside_0.3.1            codetools_0.2-20        ps_1.8.1               
[55] stringi_1.8.4           gtable_0.3.6            later_1.3.2            
[58] UCSC.utils_1.2.0        munsell_0.5.1           tibble_3.2.1           
[61] pillar_1.9.0            htmltools_0.5.8.1       rhdf5filters_1.18.0    
[64] GenomeInfoDbData_1.2.13 R6_2.5.1                doParallel_1.0.17      
[67] rprojroot_2.0.4         evaluate_1.0.1          lattice_0.22-6         
[70] httpuv_1.6.15           bslib_0.8.0             Rcpp_1.0.13-1          
[73] SparseArray_1.6.0       whisker_0.4.1           xfun_0.49              
[76] fs_1.6.5                getPass_0.2-4           pkgconfig_2.0.3        </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
