<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Givanna Putri" />

<meta name="date" content="2024-10-30" />

<title>Visium FFPE Colon Tissue Analysis</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">ASI Spatial Workshop 2024</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ghar1821/asi_spatial_workshop_2024">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Visium FFPE Colon Tissue Analysis</h1>
<h4 class="author">Givanna Putri</h4>
<h4 class="date">2024-10-30</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-11-06
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>asi_spatial_workshop_2024/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of
the R Markdown file created these results, you’ll want to first commit
it to the Git repo. If you’re still working on the analysis, you can
ignore this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240925code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240925)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240925code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240925)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomghar1821asispatialworkshop2024tree1a2b848a31540145d2c082a2f897d90c61dc8540targetblank1a2b848a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/tree/1a2b848a31540145d2c082a2f897d90c61dc8540" target="_blank">1a2b848</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomghar1821asispatialworkshop2024tree1a2b848a31540145d2c082a2f897d90c61dc8540targetblank1a2b848a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/tree/1a2b848a31540145d2c082a2f897d90c61dc8540" target="_blank">1a2b848</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    data/.DS_Store
    Ignored:    data/hello.txt
    Ignored:    data/single_cell/
    Ignored:    data/visium_ffpe_human_prostate/
    Ignored:    data/visium_human_crc/
    Ignored:    output/.DS_Store

Untracked files:
    Untracked:  analysis/visium_ffpe_prostate.Rmd.bk
    Untracked:  output/visium_data_qced.qs
    Untracked:  output/visium_data_qced.rds

Unstaged changes:
    Modified:   analysis/visium_crc.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/visium_crc.Rmd</code>) and HTML
(<code>docs/visium_crc.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/analysis/visium_crc.Rmd" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
<td>
Add tutorials for visium
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ghar1821/asi_spatial_workshop_2024/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/visium_crc.html" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
<td>
Add tutorials for visium
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="setup" class="section level2">
<h2>Setup</h2>
<pre class="r"><code>library(Seurat)</code></pre>
<pre><code>Loading required package: SeuratObject</code></pre>
<pre><code>Loading required package: sp</code></pre>
<pre><code>&#39;SeuratObject&#39; was built under R 4.4.0 but the current version is
4.4.1; it is recomended that you reinstall &#39;SeuratObject&#39; as the ABI
for R may have changed</code></pre>
<pre><code>
Attaching package: &#39;SeuratObject&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    intersect, t</code></pre>
<pre class="r"><code># for plotting
library(ggplot2)
library(scales)
library(patchwork)

# for exporting object
library(qs)</code></pre>
<pre><code>qs 0.27.2. Announcement: https://github.com/qsbase/qs/issues/103</code></pre>
</div>
<div id="loading-data" class="section level2">
<h2>Loading data</h2>
<p>The H5 file and the spatial folder came from SpaceRanger.</p>
<p>The spatial folder is required for seurat to find the following extra
information:</p>
<ul>
<li><code>tissue_lowres_image.png</code>: the image of the slide</li>
<li><code>scalefactors_json.json</code>: the scale factors to convert
the high-resolution coordinates and the scaled-down coordinates - need
to check whether this is the conversion from the high res image to low
res.</li>
<li><code>tissue_positions_list.csv</code>: coordinates for each
spot</li>
</ul>
<p><em>NOTE</em> the function requires <code>hdf5r</code> package.</p>
<p>First up, let’s import our data into a seurat object.</p>
<pre class="r"><code>dat &lt;- Load10X_Spatial(
  data.dir = &#39;data/visium_human_crc&#39;,
  filename = &#39;filtered_feature_bc_matrix.h5&#39;
)</code></pre>
</div>
<div id="inspecting-seurat-object" class="section level2">
<h2>Inspecting Seurat object</h2>
<p><strong><em>TODO</em></strong>: add a diagram to show what seurat
object looks like, if there is any?</p>
<p>We can inspect the object to see the number of spots and genes there
are in the dataset.</p>
<pre class="r"><code># 4,269 spots - samples in the seurat object description
# 18,085 genes.
dat</code></pre>
<pre><code>An object of class Seurat 
18085 features across 4269 samples within 1 assay 
Active assay: Spatial (18085 features, 0 variable features)
 1 layer present: counts
 1 spatial field of view present: slice1</code></pre>
<p>A Seurat object can contain multiple Assays. The idea behind assay is
so we can store data from different omics technologies in one
object.</p>
<p>Within an assay, we can have multiple Layers, each containing one
spot by gene matrix. In subsequent sections, when we perform some
transformations to the data, we shall see how storing the same data but
transformed it in different manner in different Layers can come in
handy.</p>
<p>By default, when we run <code>Load10x_Spatial</code>, Seurat will
store the raw spot by gene matrix in <code>Spatial</code> assay under
<code>counts</code> Layer. You can change the assay name by specifying
the <code>assay</code> parameter when running the
<code>Load10x_Spatial</code> function.</p>
<p>We can retrieve the spot by gene matrix like so:</p>
<pre class="r"><code># showing only 20 genes for 5 spots
head(dat[[&quot;Spatial&quot;]]$counts, c(20, 5))</code></pre>
<pre><code>20 x 5 sparse Matrix of class &quot;dgCMatrix&quot;
         AACAATGTGCTCCGAG-1 AACACCATTCGCATAC-1 AACACGACAACGGAGT-1
SAMD11                    1                  1                  .
NOC2L                     4                  .                 12
KLHL17                    .                  .                  1
PLEKHN1                   2                  .                 11
PERM1                     .                  .                  .
HES4                      2                  1                  3
ISG15                    10                  3                 18
AGRN                      6                  .                 25
RNF223                    .                  .                  1
C1orf159                  3                  .                  1
TTLL10                    1                  .                  .
TNFRSF18                  1                  .                  2
TNFRSF4                   .                  .                  1
SDF4                     29                  3                 45
B3GALT6                   6                  .                 12
C1QTNF12                  1                  .                  .
UBE2J2                    6                  2                 14
SCNN1D                    .                  .                  .
ACAP3                    14                  1                  8
PUSL1                     3                  .                  4
         AACACGCAGATAACAA-1 AACACGTTGATACCGC-1
SAMD11                    .                  .
NOC2L                     .                  .
KLHL17                    1                  1
PLEKHN1                   .                  2
PERM1                     .                  1
HES4                      .                  .
ISG15                     .                  1
AGRN                      .                  .
RNF223                    .                  .
C1orf159                  .                  .
TTLL10                    .                  .
TNFRSF18                  .                  1
TNFRSF4                   .                  2
SDF4                      5                  7
B3GALT6                   .                  2
C1QTNF12                  .                  .
UBE2J2                    1                  2
SCNN1D                    .                  1
ACAP3                     .                  5
PUSL1                     1                  .</code></pre>
<p>By default, genes (features) are stored as rows and spots (or cells)
are stored as columns.</p>
<p>Different assay can have different features, and as of Seurat v5,
different assay and layer can have different spots (or cells).</p>
<div id="gene-and-spot-metadata" class="section level3">
<h3>Gene and spot metadata</h3>
<p>We can retrieve the name of the genes we have in the data like
so:</p>
<pre class="r"><code>head(rownames(dat), 20)</code></pre>
<pre><code> [1] &quot;SAMD11&quot;   &quot;NOC2L&quot;    &quot;KLHL17&quot;   &quot;PLEKHN1&quot;  &quot;PERM1&quot;    &quot;HES4&quot;    
 [7] &quot;ISG15&quot;    &quot;AGRN&quot;     &quot;RNF223&quot;   &quot;C1orf159&quot; &quot;TTLL10&quot;   &quot;TNFRSF18&quot;
[13] &quot;TNFRSF4&quot;  &quot;SDF4&quot;     &quot;B3GALT6&quot;  &quot;C1QTNF12&quot; &quot;UBE2J2&quot;   &quot;SCNN1D&quot;  
[19] &quot;ACAP3&quot;    &quot;PUSL1&quot;   </code></pre>
<pre class="r"><code>head(Features(dat), 20)</code></pre>
<pre><code> [1] &quot;SAMD11&quot;   &quot;NOC2L&quot;    &quot;KLHL17&quot;   &quot;PLEKHN1&quot;  &quot;PERM1&quot;    &quot;HES4&quot;    
 [7] &quot;ISG15&quot;    &quot;AGRN&quot;     &quot;RNF223&quot;   &quot;C1orf159&quot; &quot;TTLL10&quot;   &quot;TNFRSF18&quot;
[13] &quot;TNFRSF4&quot;  &quot;SDF4&quot;     &quot;B3GALT6&quot;  &quot;C1QTNF12&quot; &quot;UBE2J2&quot;   &quot;SCNN1D&quot;  
[19] &quot;ACAP3&quot;    &quot;PUSL1&quot;   </code></pre>
<p>And the unique name assigned to each spot like so:</p>
<pre class="r"><code>head(colnames(dat), 20)</code></pre>
<pre><code> [1] &quot;AACAATGTGCTCCGAG-1&quot; &quot;AACACCATTCGCATAC-1&quot; &quot;AACACGACAACGGAGT-1&quot;
 [4] &quot;AACACGCAGATAACAA-1&quot; &quot;AACACGTTGATACCGC-1&quot; &quot;AACACTCGTGAGCTTC-1&quot;
 [7] &quot;AACAGCCTCCTGACTA-1&quot; &quot;AACAGGAATTCTGTGA-1&quot; &quot;AACAGGCCAGGCCGCT-1&quot;
[10] &quot;AACAGGCCATTGTCAC-1&quot; &quot;AACAGTCGAGGTAGAT-1&quot; &quot;AACAGTCTGACGAATT-1&quot;
[13] &quot;AACATACTCCAGCTGA-1&quot; &quot;AACATAGAAGACTGTT-1&quot; &quot;AACATAGAAGGTGAGT-1&quot;
[16] &quot;AACATAGAAGTGCGCA-1&quot; &quot;AACATAGGAGGCGTCC-1&quot; &quot;AACATAGGTCCATAGC-1&quot;
[19] &quot;AACATAGGTTCCGCAC-1&quot; &quot;AACATAGGTTGGCACC-1&quot;</code></pre>
<pre class="r"><code>head(Cells(dat), 20)</code></pre>
<pre><code> [1] &quot;AACAATGTGCTCCGAG-1&quot; &quot;AACACCATTCGCATAC-1&quot; &quot;AACACGACAACGGAGT-1&quot;
 [4] &quot;AACACGCAGATAACAA-1&quot; &quot;AACACGTTGATACCGC-1&quot; &quot;AACACTCGTGAGCTTC-1&quot;
 [7] &quot;AACAGCCTCCTGACTA-1&quot; &quot;AACAGGAATTCTGTGA-1&quot; &quot;AACAGGCCAGGCCGCT-1&quot;
[10] &quot;AACAGGCCATTGTCAC-1&quot; &quot;AACAGTCGAGGTAGAT-1&quot; &quot;AACAGTCTGACGAATT-1&quot;
[13] &quot;AACATACTCCAGCTGA-1&quot; &quot;AACATAGAAGACTGTT-1&quot; &quot;AACATAGAAGGTGAGT-1&quot;
[16] &quot;AACATAGAAGTGCGCA-1&quot; &quot;AACATAGGAGGCGTCC-1&quot; &quot;AACATAGGTCCATAGC-1&quot;
[19] &quot;AACATAGGTTCCGCAC-1&quot; &quot;AACATAGGTTGGCACC-1&quot;</code></pre>
<p>We can also query the seurat object as if it is a table by using
functions like <code>ncol</code>, <code>nrow</code>, and
<code>dim</code>:</p>
<pre class="r"><code># how many spots
ncol(dat)</code></pre>
<pre><code>[1] 4269</code></pre>
<pre class="r"><code># how many genes
nrow(dat)</code></pre>
<pre><code>[1] 18085</code></pre>
<pre class="r"><code># how many genes and spots (rows and cols respectively)
dim(dat)</code></pre>
<pre><code>[1] 18085  4269</code></pre>
<p>Metadata for the spots can be retrieved like so:</p>
<pre class="r"><code>head(dat[[]], 20)</code></pre>
<pre><code>                      orig.ident nCount_Spatial nFeature_Spatial
AACAATGTGCTCCGAG-1 SeuratProject         159480            12673
AACACCATTCGCATAC-1 SeuratProject           5719             3393
AACACGACAACGGAGT-1 SeuratProject         244735            13370
AACACGCAGATAACAA-1 SeuratProject          12387             6003
AACACGTTGATACCGC-1 SeuratProject          58894            11198
AACACTCGTGAGCTTC-1 SeuratProject          55548            10947
AACAGCCTCCTGACTA-1 SeuratProject         270292            13726
AACAGGAATTCTGTGA-1 SeuratProject          14250             5628
AACAGGCCAGGCCGCT-1 SeuratProject         146950            13120
AACAGGCCATTGTCAC-1 SeuratProject         120019            12746
AACAGTCGAGGTAGAT-1 SeuratProject         200220            13057
AACAGTCTGACGAATT-1 SeuratProject         136015            12656
AACATACTCCAGCTGA-1 SeuratProject          65435            12270
AACATAGAAGACTGTT-1 SeuratProject         200960            12695
AACATAGAAGGTGAGT-1 SeuratProject           8817             4874
AACATAGAAGTGCGCA-1 SeuratProject         233746            13541
AACATAGGAGGCGTCC-1 SeuratProject          54057            11153
AACATAGGTCCATAGC-1 SeuratProject         117304            12415
AACATAGGTTCCGCAC-1 SeuratProject          11350             5506
AACATAGGTTGGCACC-1 SeuratProject         146714            12390</code></pre>
<p>Here, each row is a spot. <code>nCount_Spatial</code> shows the total
number of transcripts (specifically total number of unique UMIs) for
that spot. <code>nFeature_Spatial</code> shows how many unique genes is
expressed per spot. <code>orig.ident</code> refers to the “original
identity” of the sample. By default, the <code>Load10x_Spatial</code>
will pre-fill this with “SeuratProject”.</p>
<p>The metadata in seurat object is actually stored as a data.frame.</p>
<pre class="r"><code>class(dat[[]])</code></pre>
<pre><code>[1] &quot;data.frame&quot;</code></pre>
<p>Thus, we can add more information into it using functions that work
for data.frame. For example, we can add a separate column identifying
the patient ID the data originates from:</p>
<pre class="r"><code>dat$patient_id &lt;- &quot;P2&quot;
head(dat[[]], 20)</code></pre>
<pre><code>                      orig.ident nCount_Spatial nFeature_Spatial patient_id
AACAATGTGCTCCGAG-1 SeuratProject         159480            12673         P2
AACACCATTCGCATAC-1 SeuratProject           5719             3393         P2
AACACGACAACGGAGT-1 SeuratProject         244735            13370         P2
AACACGCAGATAACAA-1 SeuratProject          12387             6003         P2
AACACGTTGATACCGC-1 SeuratProject          58894            11198         P2
AACACTCGTGAGCTTC-1 SeuratProject          55548            10947         P2
AACAGCCTCCTGACTA-1 SeuratProject         270292            13726         P2
AACAGGAATTCTGTGA-1 SeuratProject          14250             5628         P2
AACAGGCCAGGCCGCT-1 SeuratProject         146950            13120         P2
AACAGGCCATTGTCAC-1 SeuratProject         120019            12746         P2
AACAGTCGAGGTAGAT-1 SeuratProject         200220            13057         P2
AACAGTCTGACGAATT-1 SeuratProject         136015            12656         P2
AACATACTCCAGCTGA-1 SeuratProject          65435            12270         P2
AACATAGAAGACTGTT-1 SeuratProject         200960            12695         P2
AACATAGAAGGTGAGT-1 SeuratProject           8817             4874         P2
AACATAGAAGTGCGCA-1 SeuratProject         233746            13541         P2
AACATAGGAGGCGTCC-1 SeuratProject          54057            11153         P2
AACATAGGTCCATAGC-1 SeuratProject         117304            12415         P2
AACATAGGTTCCGCAC-1 SeuratProject          11350             5506         P2
AACATAGGTTGGCACC-1 SeuratProject         146714            12390         P2</code></pre>
<p>To access the metadata associated with the genes, we can use similar
notation and also include which assay we want to retrieve the metadata
from.</p>
<pre class="r"><code>head(dat[[&#39;Spatial&#39;]][[]], 20)</code></pre>
<pre><code>data frame with 0 columns and 20 rows</code></pre>
</div>
</div>
<div id="visualising-the-spots" class="section level2">
<h2>Visualising the spots</h2>
<p>Seurat offers function to visualise the spots captured on the tissue.
One that is very commonly used is <code>SpatialFeaturePlot</code> which
place the spots on top of the tissue image and colour the spots by
whatever features you have in the Seurat object.</p>
<p>For example, let’s visualise the library size of the spots.</p>
<pre class="r"><code>SpatialFeaturePlot(dat, features = &quot;nCount_Spatial&quot;)</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-12-1">
Past versions of unnamed-chunk-12-1.png
</button>
</p>
<div id="fig-unnamed-chunk-12-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-12-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>You can also visualise multiple features at a time, e.g., let’s also
visualise the number of genes expressed for each spot as well as the
library size:</p>
<pre class="r"><code>SpatialFeaturePlot(dat, features = c(&quot;nCount_Spatial&quot;, &quot;nFeature_Spatial&quot;))</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-13-1">
Past versions of unnamed-chunk-13-1.png
</button>
</p>
<div id="fig-unnamed-chunk-13-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-13-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>By default, the plot will focus on the the area of the tissue where
spots containing spots. Disabling this will show entire tissue.</p>
<pre class="r"><code>SpatialFeaturePlot(dat, features = &quot;nCount_Spatial&quot;, crop=FALSE)</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-14-1">
Past versions of unnamed-chunk-14-1.png
</button>
</p>
<div id="fig-unnamed-chunk-14-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-14-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Often, spots may look too small, like in this case. We can increase
it by increasing the <code>pt.size.factor</code> parameter.</p>
<pre class="r"><code># store it so we can use the same number later on
pt_size &lt;- 5
SpatialFeaturePlot(dat, features = c(&quot;nCount_Spatial&quot;, &quot;nFeature_Spatial&quot;), 
                   pt.size.factor = pt_size)</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-15-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-15-1">
Past versions of unnamed-chunk-15-1.png
</button>
</p>
<div id="fig-unnamed-chunk-15-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-15-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>To visualise qualitative features, we can use
<code>SpatialDimPlot</code>. For example, let’s visualise the spot by
the patient ID metadata we added before.</p>
<pre class="r"><code>SpatialDimPlot(dat, group.by = &#39;patient_id&#39;, pt.size.factor = pt_size)</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-16-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-16-1">
Past versions of unnamed-chunk-16-1.png
</button>
</p>
<div id="fig-unnamed-chunk-16-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-16-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We can override the spot colour by overriding the <code>cols</code>
parameter with a named vector mapping the discrete category in the data
against the colour.</p>
<pre class="r"><code># pretend we have three patients in the data
dat[[]]$pretend_patient_id &lt;- c(
  rep(&quot;Adam&quot;, 3000),
  rep(&quot;Jane&quot;, 1000),
  rep(&quot;Clare&quot;, 269)
)

SpatialDimPlot(dat, group.by = &#39;pretend_patient_id&#39;, 
               pt.size.factor = pt_size, 
               cols = c(&quot;Adam&quot; = &quot;blue&quot;, &quot;Jane&quot; = &quot;orange&quot;, &quot;Clare&quot; = &quot;black&quot;))</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-17-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-17-1">
Past versions of unnamed-chunk-17-1.png
</button>
</p>
<div id="fig-unnamed-chunk-17-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-17-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="tweaking-spatial-plots" class="section level3">
<h3>Tweaking spatial plots</h3>
<p>Often, it is helpful to tweak the opacity of the spots or the
background image to help us better visualise the spots and their
features.</p>
<p>You can adjust the opacity of the spots by lowering the
<code>alpha</code> parameter:</p>
<pre class="r"><code>SpatialFeaturePlot(
  dat, 
  features = &quot;nCount_Spatial&quot;,
  pt.size.factor = pt_size,
  alpha = 0.5
)</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-18-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-18-1">
Past versions of unnamed-chunk-18-1.png
</button>
</p>
<div id="fig-unnamed-chunk-18-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-18-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Alpha parameter can take either a single numerical value or a range
(a vector with 2 values). If the latter, it represents the minimum and
maximum transparency and default to c(1,1), i.e. same transparency for
both low and high value. Thus, we can accentuate the spots with high
value by lowering the minimum value we pass as alpha:</p>
<pre class="r"><code>SpatialFeaturePlot(
  dat, 
  features = &quot;nCount_Spatial&quot;,
  pt.size.factor = 5,
  alpha = c(0.1, 1)
)</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-19-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-19-1">
Past versions of unnamed-chunk-19-1.png
</button>
</p>
<div id="fig-unnamed-chunk-19-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-19-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The lower the minimum alpha value, the stronger the high expression
will look.</p>
<p>However, it is often the case that the background image is too
overpowering, making it very hard to view the spots. We can make the
background more transparent by lowering the <code>image.alpha</code>
parameter.</p>
<pre class="r"><code>SpatialFeaturePlot(
  dat, 
  features = &quot;nCount_Spatial&quot;,
  pt.size.factor = pt_size,
  image.alpha = 0.2
)</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-20-1">
Past versions of unnamed-chunk-20-1.png
</button>
</p>
<div id="fig-unnamed-chunk-20-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-20-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>You can also completely forgo the tissue image by setting the
<code>image.alpha</code> to 0.</p>
<pre class="r"><code>SpatialFeaturePlot(
  dat, 
  features = &quot;nCount_Spatial&quot;,
  pt.size.factor = 4,
  image.alpha = 0
)</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-21-1">
Past versions of unnamed-chunk-21-1.png
</button>
</p>
<div id="fig-unnamed-chunk-21-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-21-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Both <code>SpatialFeaturePlot</code> and <code>SpatialDimPlot</code>
will return a <code>ggplot</code> object, which we can modify using the
functionalities built into <code>ggplot2</code> package.</p>
<p>As an example, let’s update the plot showing library size so we:</p>
<ol style="list-style-type: decimal">
<li>Change the colour scheme to use viridis</li>
<li>Rename <code>nCount_Spatial</code> to <code>library_size</code></li>
<li>Have a black background and axes</li>
</ol>
<p>Let’s update the plot so we have a black background and white letters
and axes.</p>
<pre class="r"><code>plt_dark_bg &lt;- SpatialFeaturePlot(
  dat, 
  features = &quot;nCount_Spatial&quot;,
  pt.size.factor = pt_size,
  image.alpha = 0
) + scale_fill_viridis_c(option = &#39;viridis&#39;) + 
  labs(fill=&#39;library_size&#39;) +
  theme_minimal() + 
  theme(
  legend.text = element_text(colour = &quot;white&quot;),
  legend.title = element_text(colour = &quot;white&quot;),
  legend.background = element_rect(fill = &quot;black&quot;),
  plot.background = element_rect(fill = &quot;black&quot;),
  panel.grid = element_blank(),
  axis.text = element_text(colour = &quot;white&quot;),
  axis.line = element_line(colour = &quot;white&quot;)
) </code></pre>
<pre><code>Scale for fill is already present.
Adding another scale for fill, which will replace the existing scale.</code></pre>
<pre class="r"><code>plt_dark_bg</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-22-1">
Past versions of unnamed-chunk-22-1.png
</button>
</p>
<div id="fig-unnamed-chunk-22-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-22-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Since the plot is a ggplot object, we can export them using either
the export button in Rstudio’s panel, or the ggsave function:</p>
<pre class="r"><code>ggsave(filename=&quot;output/spatial_plot_darkbg.png&quot;, plot=plt_dark_bg)</code></pre>
<blockquote>
<p>Good to know: When there are multiple features drawn using
<code>SpatialFeaturePlot</code>, the resulting object is no longer a
ggplot object but rather a “Large patchwork” object, an object generated
by the <a
href="https://patchwork.data-imaginist.com/index.html">patchwork
package</a>. This package is commonly used for combining plots in R. In
actual fact, the large patchwork object itself is a list where each
element is a ggplot object. Thus, to modify the plot, you will just need
to override the object like above.</p>
</blockquote>
</div>
</div>
<div id="qc" class="section level2">
<h2>QC</h2>
<p>We can divide the QC into two sections, QC the spots, and QC the
genes.</p>
<div id="qc-genes" class="section level3">
<h3>QC genes</h3>
<p>By default, <code>Load10X_Spatial</code> will include all genes in
the data. However, we may have genes that are only expressed in a
handful of spots and may well be noise.</p>
<pre class="r"><code># Count how many spots each gene is expressed in.
spots_per_gene &lt;- rowSums(dat[[&quot;Spatial&quot;]]$counts &gt; 0)

# Plot as histogram
# Have to convert spots_per_gene to data.frame first, the column is spots
ggplot(data.frame(spots = spots_per_gene), aes(x=spots)) + 
  geom_histogram(binwidth = 1, colour = &#39;blue&#39;) +
  geom_vline(xintercept = 100, colour = &#39;red&#39;) +
  theme_minimal() +
  labs(title = &quot;Number of Spots per Gene&quot;,
       x = &quot;Number of Spots&quot;,
       y = &quot;Number of Genes&quot;) +
  scale_x_continuous(breaks = pretty_breaks(10)) +
  scale_y_continuous(breaks = pretty_breaks(10))</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-24-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-24-1">
Past versions of unnamed-chunk-24-1.png
</button>
</p>
<div id="fig-unnamed-chunk-24-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-24-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>From the histogram, we can see that we have some genes that are
expressed only by a handful of spots, e.g., there are 12 genes that are
not expressed by any spots at all.</p>
<p>For this dataset, we will set the threshold to 100, i.e. for a gene
to be included, it needs to be expressed by at least 100 spots, which is
around 2.3% of total spots we have in the data.</p>
<pre class="r"><code># names will give you the gene names
genes_to_keep &lt;- names(spots_per_gene[spots_per_gene &gt;= 100])

# Subset the Seurat object
dat_fil &lt;- subset(dat, features = genes_to_keep)</code></pre>
<pre><code>Warning: Not validating Centroids objects
Not validating Centroids objects</code></pre>
<pre><code>Warning: Not validating FOV objects
Not validating FOV objects
Not validating FOV objects
Not validating FOV objects
Not validating FOV objects
Not validating FOV objects</code></pre>
<pre><code>Warning: Not validating Seurat objects</code></pre>
<blockquote>
<p>Good to know: The threshold 100 is not fixed. This will vary
depending on the dataset you have on hand. So adjust this as you see
fit.</p>
</blockquote>
</div>
<div id="qc-spots" class="section level3">
<h3>QC spots</h3>
<p>In addition to QC the genes, we should also QC the spots. The idea is
to make sure the spots overlap with tissue sections that are stained
with the fluorescence dyes, and that we do not have spots that lie
outside the tissue.</p>
<blockquote>
<p>Tom O’neil demo using ImageJ</p>
</blockquote>
<p>To check whether the spots lie on top of the tissues, we can
interactively inspect <code>SpatialDimPlot</code> and note the ID of the
spots we need to remove.</p>
<p>The default spot ID is a nucleotide sequence that is hard to remember
and note. Thus, to make life easier, let’s assign a new numerical ID to
the spots.</p>
<pre class="r"><code>dat_fil[[]]$numeric_spot_id &lt;- paste0(&quot;spot_&quot;, seq(ncol(dat_fil)))

# backup the original id
dat_fil[[]]$orig_spot_id &lt;- colnames(dat_fil)

# assign the new numeric id as names of the spots
colnames(dat_fil) &lt;- dat_fil[[]]$numeric_spot_id

head(dat_fil[[]], 20)</code></pre>
<pre><code>           orig.ident nCount_Spatial nFeature_Spatial patient_id
spot_1  SeuratProject         159478            12671         P2
spot_2  SeuratProject           5719             3393         P2
spot_3  SeuratProject         244734            13369         P2
spot_4  SeuratProject          12387             6003         P2
spot_5  SeuratProject          58892            11196         P2
spot_6  SeuratProject          55548            10947         P2
spot_7  SeuratProject         270289            13723         P2
spot_8  SeuratProject          14250             5628         P2
spot_9  SeuratProject         146948            13118         P2
spot_10 SeuratProject         120018            12745         P2
spot_11 SeuratProject         200219            13056         P2
spot_12 SeuratProject         136014            12655         P2
spot_13 SeuratProject          65431            12267         P2
spot_14 SeuratProject         200958            12694         P2
spot_15 SeuratProject           8816             4873         P2
spot_16 SeuratProject         233742            13537         P2
spot_17 SeuratProject          54057            11153         P2
spot_18 SeuratProject         117303            12414         P2
spot_19 SeuratProject          11350             5506         P2
spot_20 SeuratProject         146713            12389         P2
        pretend_patient_id numeric_spot_id       orig_spot_id
spot_1                Adam          spot_1 AACAATGTGCTCCGAG-1
spot_2                Adam          spot_2 AACACCATTCGCATAC-1
spot_3                Adam          spot_3 AACACGACAACGGAGT-1
spot_4                Adam          spot_4 AACACGCAGATAACAA-1
spot_5                Adam          spot_5 AACACGTTGATACCGC-1
spot_6                Adam          spot_6 AACACTCGTGAGCTTC-1
spot_7                Adam          spot_7 AACAGCCTCCTGACTA-1
spot_8                Adam          spot_8 AACAGGAATTCTGTGA-1
spot_9                Adam          spot_9 AACAGGCCAGGCCGCT-1
spot_10               Adam         spot_10 AACAGGCCATTGTCAC-1
spot_11               Adam         spot_11 AACAGTCGAGGTAGAT-1
spot_12               Adam         spot_12 AACAGTCTGACGAATT-1
spot_13               Adam         spot_13 AACATACTCCAGCTGA-1
spot_14               Adam         spot_14 AACATAGAAGACTGTT-1
spot_15               Adam         spot_15 AACATAGAAGGTGAGT-1
spot_16               Adam         spot_16 AACATAGAAGTGCGCA-1
spot_17               Adam         spot_17 AACATAGGAGGCGTCC-1
spot_18               Adam         spot_18 AACATAGGTCCATAGC-1
spot_19               Adam         spot_19 AACATAGGTTCCGCAC-1
spot_20               Adam         spot_20 AACATAGGTTGGCACC-1</code></pre>
<p>Let’s interactively visualise the tissue and the spot:</p>
<pre class="r"><code># unfortunately, the spot size and colour cannot be changed here. 
SpatialDimPlot(
  dat_fil,
  group.by = &quot;patient_id&quot;,
  interactive = TRUE
)</code></pre>
<p>Note the ID of the spots we want to remove and then remove them.</p>
<pre class="r"><code>spots_to_remove &lt;- paste0(&quot;spot_&quot;, c(
  4030, 1810, 1168, 3110, 44, 513, 2278, 3126, 2803, 3828, 2147, 1879, 3828, 2021, 
  4215, 3631, 3225, 3631, 3843, 4035, 1076, 93, 2615, 2041, 749, 3629, 3155, 3683, 
  560, 2953, 2211, 59, 1502, 1789, 2244, 649, 2720, 3803, 3087, 2717, 663, 2882, 
  2934, 3440, 1540, 3830, 471, 1035, 3221, 3257, 3039, 1570, 3560, 758, 40, 1154, 
  2275, 3841, 3475, 3361, 3760, 3592, 1983, 1862, 671, 4234, 622, 2583, 573, 3832, 3959, 1426,
  3746, 959, 1228, 3391, 2332, 2785, 4243, 2315, 1107, 1504, 1133, 3505, 3504, 670
))

# add it as a metadata
dat_fil[[]]$remove_spot &lt;- Cells(dat_fil) %in% spots_to_remove

dat_fil &lt;- subset(x = dat_fil, remove_spot == FALSE)</code></pre>
<pre><code>Warning: Not validating Centroids objects
Not validating Centroids objects</code></pre>
<pre><code>Warning: Not validating FOV objects
Not validating FOV objects
Not validating FOV objects
Not validating FOV objects
Not validating FOV objects
Not validating FOV objects</code></pre>
<pre><code>Warning: Not validating Seurat objects</code></pre>
</div>
<div id="check-spots-library-size-and-genes-expressed"
class="section level3">
<h3>Check spots’ library size and genes expressed</h3>
<p>We should also remove spots which express low library size and/or low
number of genes expressed.</p>
<p>Few plots we can draw to check this are violin plots, histogram, and
plain scatter plots:</p>
<pre class="r"><code>plt1 &lt;- ggplot(dat_fil[[]], aes(x=nCount_Spatial)) + geom_histogram(bins = 100) +
  theme_classic() +
  scale_y_continuous(breaks = pretty_breaks(10)) +
  scale_x_continuous(breaks = pretty_breaks(10)) 
plt2 &lt;- ggplot(dat_fil[[]], aes(x=nFeature_Spatial)) + geom_histogram(bins = 100) +
  theme_classic() +
  scale_y_continuous(breaks = pretty_breaks(10)) +
  scale_x_continuous(breaks = pretty_breaks(10)) 

plt3 &lt;- VlnPlot(dat_fil, features = c(&quot;nCount_Spatial&quot;, &quot;nFeature_Spatial&quot;))</code></pre>
<pre><code>Warning: Default search for &quot;data&quot; layer in &quot;Spatial&quot; assay yielded no results;
utilizing &quot;counts&quot; layer instead.</code></pre>
<pre class="r"><code>plt4 &lt;- FeatureScatter(dat_fil, feature1 = &quot;nCount_Spatial&quot;, feature2 = &quot;nFeature_Spatial&quot;) +
  scale_y_continuous(breaks = pretty_breaks(10)) +
  scale_x_continuous(breaks = pretty_breaks(10)) +
  theme_bw()

plt1 + plt2 + plt3 + plt4</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-29-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-29-1">
Past versions of unnamed-chunk-29-1.png
</button>
</p>
<div id="fig-unnamed-chunk-29-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-29-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The histograms and violin plots are great for determining the minimum
cutoff for library size and/or number of genes expressed, while the
scatter plot is handy to check whether the spots with low library size
also has low number of genes expressed, which is a clear sign of
potential noise.</p>
<p>For this workshop, we will exclude spots that express less than 3000
genes. Notably, this threshold will also automatically exclude spots
with small library size.</p>
<pre class="r"><code>dat_fil &lt;- subset(x = dat_fil, nFeature_Spatial &gt;= 3000)</code></pre>
<pre><code>Warning: Not validating Centroids objects
Not validating Centroids objects</code></pre>
<pre><code>Warning: Not validating FOV objects
Not validating FOV objects
Not validating FOV objects
Not validating FOV objects
Not validating FOV objects
Not validating FOV objects</code></pre>
<pre><code>Warning: Not validating Seurat objects</code></pre>
<pre class="r"><code>dat_fil</code></pre>
<pre><code>An object of class Seurat 
17991 features across 4147 samples within 1 assay 
Active assay: Spatial (17991 features, 0 variable features)
 1 layer present: counts
 1 spatial field of view present: slice1</code></pre>
<blockquote>
<p>Good to know: Unlike single cell in which cells with high library
size are often removed because they may well represent doublets or
multiplets, for visium data, spots with high library size may not
necessarily mean bad spots as visium data is not single cell resolution.
Those spots may well just contain an assortment of cell types. Removing
them may well be removing areas with important biology!</p>
</blockquote>
</div>
<div id="qc-spots-based-on-mitochondria-genes-expression"
class="section level3">
<h3>QC spots based on mitochondria genes expression</h3>
<p>Just like in single cell, we can also QC the spots based on the
percentage of mitochondria genes expressed. Spots that have excessively
high expression of mitochondria genes may contain low quality or dying
cells.</p>
<pre class="r"><code># This function will compute, for each spot, percentage of transcripts that mapped to mitochondria genes
dat_fil[[&#39;percent_mt&#39;]] &lt;- PercentageFeatureSet(dat_fil, pattern = &#39;^MT-&#39;)

# Visualize them
plt1 &lt;- VlnPlot(dat_fil, features = &quot;percent_mt&quot;)</code></pre>
<pre><code>Warning: Default search for &quot;data&quot; layer in &quot;Spatial&quot; assay yielded no results;
utilizing &quot;counts&quot; layer instead.</code></pre>
<pre class="r"><code>plt2 &lt;- FeatureScatter(dat_fil, feature1 = &#39;nCount_Spatial&#39;, feature2 = &#39;percent_mt&#39;)

plt1 + plt2</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-31-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-31-1">
Past versions of unnamed-chunk-31-1.png
</button>
</p>
<div id="fig-unnamed-chunk-31-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-31-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We will remove spots that have &gt; 12% mitochondria genes.</p>
<pre class="r"><code>dat_fil &lt;- subset(dat_fil, subset = percent_mt &lt; 12)</code></pre>
<pre><code>Warning: Not validating Centroids objects
Not validating Centroids objects</code></pre>
<pre><code>Warning: Not validating FOV objects
Not validating FOV objects
Not validating FOV objects
Not validating FOV objects
Not validating FOV objects
Not validating FOV objects</code></pre>
<pre><code>Warning: Not validating Seurat objects</code></pre>
<pre class="r"><code>dat_fil</code></pre>
<pre><code>An object of class Seurat 
17991 features across 4137 samples within 1 assay 
Active assay: Spatial (17991 features, 0 variable features)
 1 layer present: counts
 1 spatial field of view present: slice1</code></pre>
<p>After QC, we ended up with 4,215 spots.</p>
</div>
<div id="export-data-out" class="section level3">
<h3>Export data out</h3>
<p>There are two options, save as RDS or using QS library.</p>
<pre class="r"><code>saveRDS(dat_fil, &quot;output/visium_data_qced.rds&quot;)

qsave(dat_fil, &quot;output/visium_data_qced.qs&quot;)</code></pre>
<blockquote>
<p>Good to know: There are QC tools purposely developed for QCing
spatial data like SpotSweeper that can QC spots based on the their
neighbouring spots.</p>
</blockquote>
</div>
</div>
<div id="normalisation" class="section level2">
<h2>Normalisation</h2>
<p>Why is this necessary? The same reason as any other RNA seq based
protocol (including single cell). Differences in library size may well
be due to variances in sampling of RNA molecules. During library
preparation, it is not possible to capture all molecules and sequence
them. Thus, the unevenness of library sizes and differences in gene
expression across spots, may well be due to variance in sampling. There
are also other technical variations that contributed to the differences
that need to be accounted for, e.g., Probe hybridization efficiency and
PCR amplification biases.</p>
<p>Normalisation aims to adjust differences introduced due to technical
variations. In single cell, there are many methods that have been
developed to do this (see <a
href="https://www.nature.com/articles/s41592-023-01814-1">here</a> for
benchmarking paper). Admittedly, not all of them will be suitable for
visium data. For this tutorial, we will demonstrate classic log
normalisation and SCTransform.</p>
<div id="running-sctransform" class="section level3">
<h3>Running SCTransform</h3>
<p>Briefly SCTransform fits a regularised negative binomial model to the
raw count matrix, using sequencing depth as a covariate.</p>
<p>Note, SCT requires glmGamPoi package:
<code>BiocManager::install('glmGamPoi')</code></p>
<pre class="r"><code># Might need to increase the size limit for global variables so parallel workers
# can access it. 2GB should be sufficient.
options(future.globals.maxSize = 2000 * 1024^2)


# Note, return.only.var.genes can be set to TRUE if we want to get back 
# the expression of only highly variable genes in scale.data
dat_fil &lt;- SCTransform(dat_fil, assay = &quot;Spatial&quot;, verbose = TRUE, return.only.var.genes = FALSE)</code></pre>
<pre><code>Running SCTransform on assay: Spatial</code></pre>
<pre><code>Running SCTransform on layer: counts</code></pre>
<pre><code>vst.flavor=&#39;v2&#39; set. Using model with fixed slope and excluding poisson genes.</code></pre>
<pre><code>Variance stabilizing transformation of count matrix of size 17991 by 4137</code></pre>
<pre><code>Model formula is y ~ log_umi</code></pre>
<pre><code>Get Negative Binomial regression parameters per gene</code></pre>
<pre><code>Using 2000 genes, 4137 cells</code></pre>
<pre><code>Found 25 outliers - those will be ignored in fitting/regularization step</code></pre>
<pre><code>Second step: Get residuals using fitted parameters for 17991 genes</code></pre>
<pre><code>Computing corrected count matrix for 17991 genes</code></pre>
<pre><code>Calculating gene attributes</code></pre>
<pre><code>Wall clock passed: Time difference of 11.48468 secs</code></pre>
<pre><code>Determine variable features</code></pre>
<pre><code>Centering data matrix</code></pre>
<pre><code>Set default assay to SCT</code></pre>
<pre class="r"><code># Inspect after
dat_fil</code></pre>
<pre><code>An object of class Seurat 
35982 features across 4137 samples within 2 assays 
Active assay: SCT (17991 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: Spatial
 1 spatial field of view present: slice1</code></pre>
<p>By default, SCTransform will store the output in a different assay
called “SCT”.</p>
<pre class="r"><code>head(dat_fil[[&#39;SCT&#39;]]$data, c(20, 5))</code></pre>
<pre><code>20 x 5 sparse Matrix of class &quot;dgCMatrix&quot;
            spot_1    spot_2    spot_3    spot_4    spot_6
SAMD11   .         1.0986123 .         0.6931472 0.6931472
NOC2L    1.0986123 1.3862944 1.6094379 1.0986123 1.6094379
KLHL17   .         0.6931472 .         1.0986123 1.3862944
PLEKHN1  0.6931472 0.6931472 1.6094379 0.6931472 .        
PERM1    .         .         .         .         .        
HES4     0.6931472 1.6094379 .         0.6931472 1.7917595
ISG15    1.7917595 2.8903718 1.9459101 1.3862944 2.3978953
AGRN     1.0986123 1.7917595 2.1972246 1.6094379 2.3025851
RNF223   .         .         .         .         .        
C1orf159 1.0986123 0.6931472 .         0.6931472 .        
TTLL10   0.6931472 .         .         .         .        
TNFRSF18 .         0.6931472 0.6931472 0.6931472 1.3862944
TNFRSF4  .         0.6931472 .         0.6931472 .        
SDF4     2.7080502 3.3672958 2.7725887 3.2958369 2.7725887
B3GALT6  1.3862944 1.0986123 1.6094379 1.0986123 .        
C1QTNF12 0.6931472 .         .         .         .        
UBE2J2   1.3862944 2.1972246 1.7917595 1.7917595 1.0986123
SCNN1D   .         .         .         .         .        
ACAP3    2.0794415 2.1972246 1.0986123 1.3862944 2.3978953
PUSL1    1.0986123 0.6931472 0.6931472 1.3862944 .        </code></pre>
<blockquote>
<p>Note on layers: As previously discussed in the beginning, an assay
can contain multiple layers. After running SCTransfrom, you shall notice
a new assay call SCT containing SCTransformed counts. Specifically,
counts layer contains corrected UMI counts, data contains log1p - log(1
+ counts) of the corrected UMI counts, scale.data contains the counts in
the data layer transformed using pearson residuals calculated by
SCTransform.</p>
</blockquote>
</div>
<div id="comparison-with-log-normalisation" class="section level3">
<h3>Comparison with log normalisation</h3>
<p>Log normalisation is commonly used for single cell data. It is also
widely used for normalising visium data where for each spot and gene, we
divide the expression by the spot’s library size and multiply the value
by a size factor (say 10,000). Then we perform log1p operation where we
run logarithmic transformation after adding value of 1. Why the
addition, because log of 0 is infinity.</p>
<pre class="r"><code># By default, this will run log CPTT
# The data layer will contain the normalised counts.
# The counts layer will contain the unnormalised count.
dat_fil &lt;- NormalizeData(dat_fil, verbose = FALSE, assay = &quot;Spatial&quot;)
dat_fil</code></pre>
<pre><code>An object of class Seurat 
35982 features across 4137 samples within 2 assays 
Active assay: SCT (17991 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: Spatial
 1 spatial field of view present: slice1</code></pre>
<p>Upon running this, you will notice an additional layer in the Spatial
assay call <code>data</code> which contains the output of log
normalisation function.</p>
<p>We can plot the output of SCTransform, log normalisation, and raw
count for a gene (say COL1A1), side by side to see what they look
like.</p>
<pre class="r"><code># Important to change assay before plotting to make sure it is the right one.

DefaultAssay(object = dat_fil) &lt;- &quot;SCT&quot;
plt_sct &lt;- SpatialFeaturePlot(
  dat_fil,
  features = &quot;COL1A1&quot;,
  image.alpha = 0.5,
  pt.size.factor = 4
) + ggtitle(&quot;SCT&quot;)


DefaultAssay(object = dat_fil) &lt;- &quot;Spatial&quot;
plt_lognorm &lt;- SpatialFeaturePlot(
  dat_fil,
  features = &quot;COL1A1&quot;,
  image.alpha = 0.5,
  pt.size.factor = 4
) + ggtitle(&quot;Log Norm&quot;)

# Plot raw data for comparison
plt_raw &lt;- SpatialFeaturePlot(
  dat_fil,
  features = &quot;COL1A1&quot;,
  image.alpha = 0.5,
  pt.size.factor = 4,
  slot = &#39;counts&#39;
) + ggtitle(&quot;Raw&quot;)

# Put all 3 plots side by side for comparison
plt_raw + plt_sct + plt_lognorm</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-37-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-37-1">
Past versions of unnamed-chunk-37-1.png
</button>
</p>
<div id="fig-unnamed-chunk-37-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-37-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<blockquote>
<p>Good to know: There are pros and cons on choosing either SCT or log
normalisation. From the plot, you can see they are different.
Literatures say SCT mitigates the correlation between library size and
normalised counts which log normalisation failed to do so. On the other
hand, log normalisation is very simple (you can compute it by hand if
you like!) and fast to compute, very well established, intuitive, and
compatible with a variety of analysis tools and pipelines. While SCT may
be more robust for data with high variability in sequencing depth, at
the end of the day, choose whichever method that you are comfortable
with.</p>
</blockquote>
<p>For this tutorial, we will use the normalised counts generated by
SCTransform.</p>
<pre class="r"><code># Set the default assay to SCT
DefaultAssay(dat_fil) &lt;- &quot;SCT&quot;

# save so we can reload if we need to
qsave(dat_fil, &quot;output/visium_data_qced.qs&quot;)
# dat_fil &lt;- qread(&quot;output/visium_data_qced.qs&quot;)</code></pre>
</div>
</div>
<div id="identifying-cell-types-in-different-spots"
class="section level2">
<h2>Identifying cell types in different spots</h2>
<p>There are many ways to do this, e.g., clustering or deconvolution.
Clustering which can be controversial for spot based data like visium as
a spot may contain an assortment of cell types. Deconvolution will
estimate how many cell types in a spot using a single cell data as a
reference, which is not always readily available. Another simple yet
efficient method is using module scoring (<code>AddModuleScore</code>
function).</p>
<p><code>AddModuleScore</code> calculates the average expression levels
of predefined set of genes (module) for each spot, subtracting the
background expression of control genes, which by default, is set to the
whole library. This results in a “module score” for each module in each
spot. High score suggests high likelihood of a given cell type (defined
by a set of marker genes in a module) being present in that spot.</p>
<p>Module scoring is a simple yet powerful way of resolving what cell
types are likely to be present in which spot if you know for sure what
the marker genes for those cell types are. A spot that scores high in
multiple modules is likely to contain a mixture of those cell types,
while a score scoring high in only a single module, is likely to contain
mostly a cell type.</p>
<p>Care need to be taken when interpreting the module score as the
specificity of a given module is highly dependent on the quality of the
gene lists. Moreover, it does not provide absolute quantification of
cell types (exactly how many cell types are present in a spot).</p>
<pre class="r"><code># for finding genes
# Features(dat_fil)[grep(&quot;S100B.*&quot;, Features(dat_fil))]

# The following marker genes are obtained from combination of
# https://doi.org/10.1038/s43018-024-00807-z
# https://doi.org/10.1038/s41467-024-49916-4
# and perplexity AI.
genes_to_score &lt;- list(
  ec = c(&quot;PLVAP&quot;, &quot;VWF&quot;, &quot;PECAM1&quot;, &quot;ERG&quot;),
  malignant = c(&quot;EPCAM&quot;, &quot;EGFR&quot;),
  glial = c(&quot;CLU&quot;, &quot;CRYAB&quot;, &quot;S100B&quot;, &quot;GFAP&quot;, &quot;SOX10&quot;),
  fibroblast = c(&quot;COL1A1&quot;, &quot;COL1A2&quot;, &quot;VIM&quot;, &quot;FAP&quot;, &quot;PDGFRA&quot;, &quot;THY1&quot;, &quot;S100A4&quot;),
  bcell = c(&quot;PTPRC&quot;, &quot;MS4A1&quot;, &quot;CD79A&quot;, &quot;MZB1&quot;),
  tcell = c(&quot;PTPRC&quot;, &quot;CD3D&quot;, &quot;CD3E&quot;, &quot;CD3G&quot;, &quot;CCL5&quot;, &quot;GZMA&quot;, &quot;TRBC2&quot;, &quot;CD2&quot;, &quot;TRAC&quot;, &quot;CD7&quot;, &quot;KLRB1&quot;, &quot;GNLY&quot;),
  myeloid = c(&quot;PTPRC&quot;, &quot;ITGAX&quot;, &quot;ITGAM&quot;, &quot;CD14&quot;, &quot;LYZ&quot;, &quot;LILRA4&quot;, &quot;IRF7&quot;),
  hev_ec = c(&quot;CXCL10&quot;, &quot;CXCL11&quot;, &quot;GBP1&quot;, &quot;CXCL9&quot;, &quot;ISG15&quot;, &quot;GBP4&quot;, &quot;WARS&quot;, &quot;IL32&quot;, &quot;CCL2&quot;, &quot;CTSS&quot;, 
             &quot;IGFBP5&quot;, &quot;PECAM1&quot;, &quot;VWF&quot;),
  macro = c(&quot;CD68&quot;, &quot;FCER1G&quot;, &quot;CD14&quot;, &quot;TYROBP&quot;)
)

dat_fil &lt;- AddModuleScore(
  object = dat_fil, features = genes_to_score,
  name = &quot;cell_scores&quot;
)

# rename the modules (basically copy to new column..)
names(dat_fil[[]])[names(dat_fil[[]]) == &#39;cell_scores1&#39;] &lt;- &quot;ec_score&quot;
names(dat_fil[[]])[names(dat_fil[[]]) == &#39;cell_scores2&#39;] &lt;- &quot;malignant_score&quot;
names(dat_fil[[]])[names(dat_fil[[]]) == &#39;cell_scores3&#39;] &lt;- &quot;glial_score&quot;
names(dat_fil[[]])[names(dat_fil[[]]) == &#39;cell_scores4&#39;] &lt;- &quot;fibroblast_score&quot;
names(dat_fil[[]])[names(dat_fil[[]]) == &#39;cell_scores5&#39;] &lt;- &quot;bcell_score&quot;
names(dat_fil[[]])[names(dat_fil[[]]) == &#39;cell_scores6&#39;] &lt;- &quot;tcell_score&quot;
names(dat_fil[[]])[names(dat_fil[[]]) == &#39;cell_scores7&#39;] &lt;- &quot;myeloid_score&quot;
names(dat_fil[[]])[names(dat_fil[[]]) == &#39;cell_scores8&#39;] &lt;- &quot;hev_ec_score&quot;
names(dat_fil[[]])[names(dat_fil[[]]) == &#39;cell_scores9&#39;] &lt;- &quot;macro_score&quot;</code></pre>
<p>After scoring the modules, we can look at a variety of different
biology of interest, including those that are derived from previous
publication, e.g., from <a
href="https://doi.org/10.1038/s43018-024-00807-z">this paper</a>.</p>
<div id="co-localisation-of-different-cell-types"
class="section level3">
<h3>Co-localisation of different cell types</h3>
<p>For example, the paper found the presence of rare endothelial cells
(HEV-CXCL10) expressing HEV marker (SELP) and CXCL9/10/11 that play
important role in recruiting T cells. By plotting the module scores for
Endothelial cells (<code>ec_score</code>), HEV-CXCL10 endothelial cells
(<code>hev_ec_score</code>), and T cells (<code>tcell_score</code>), we
can see whether there exists spots in this CRC tissue are have high
score for all modules or whether there are spots with high
<code>hev_ec_score</code> next to spots with high
<code>tcell_score</code>. The former will indicate a spot where there
may be a mix of T cells and HEV-CXCL10 endothelial cells while the
former will indicate there are regions where they are close to one
another.</p>
<pre class="r"><code>SpatialFeaturePlot(
  dat_fil,
  features = c(
    &quot;ec_score&quot;, &quot;hev_ec_score&quot;, &quot;tcell_score&quot;
  ),
  pt.size.factor = pt_size,
  image.alpha = 0.3
)</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-40-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-40-1">
Past versions of unnamed-chunk-40-1.png
</button>
</p>
<div id="fig-unnamed-chunk-40-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-40-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We do have some overlap on the top left and middle right. Having
generic endothelial module ensure that the HEV-CXCL10 subsets lie on the
appropriate region (where endothelial cells are).</p>
<p>The paper also discovered myeloid regions neighbouring fibroblast
regions which suggests positive correlation between macrophages and
fibroblast in tumour samples. Also positive correlation between
macro-ISG15 and macro-SPP1 with tumour cells.</p>
<pre class="r"><code>SpatialFeaturePlot(
  dat_fil,
  features = c(
    &quot;malignant_score&quot;, &quot;fibroblast_score&quot;, &quot;macro_score&quot;, &quot;ISG15&quot;, &quot;SPP1&quot;
  ),
  pt.size.factor = pt_size,
  image.alpha = 0.3
)</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-41-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-41-1">
Past versions of unnamed-chunk-41-1.png
</button>
</p>
<div id="fig-unnamed-chunk-41-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-41-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="inferring-cell-cell-communications" class="section level3">
<h3>Inferring cell-cell communications</h3>
<p>SIRPA and CD47 is a common receptor ligand pair between macrophages
and tumour cells. CD47 ligand is commonly sent out by tumour cells to
tell macrophages to not eat them. This is quite a well known immune
mechanism employed by tumour cells.</p>
<pre class="r"><code>SpatialFeaturePlot(
  dat_fil,
  features = c(
    &quot;macro_score&quot;, &quot;malignant_score&quot;, &quot;SIRPA&quot;, &quot;CD47&quot;
  ),
  ncol = 2,
  pt.size.factor = pt_size,
  image.alpha = 0.3
)</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-42-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-42-1">
Past versions of unnamed-chunk-42-1.png
</button>
</p>
<div id="fig-unnamed-chunk-42-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-42-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>From the plot, we can infer the following:</p>
<ol style="list-style-type: decimal">
<li>There are regions where macrophages and tumour cells co-exist (top
right) where there are high SIRPA and CD47 expression. As the SIRPA
signal is highly expressed, macrophages’ inhibitory receptor is
activated, signalling it against phagocytosis.</li>
<li>There are regions in middle bottom (the strip) where there are
tumour cells with low expression of CD47, and macrophages with low
expression of SIRPA. This is the region where phagocytosis might happen
because the tumour cells are lacking the CD47 ligand.</li>
</ol>
</div>
<div id="pca-and-umap" class="section level3">
<h3>PCA and UMAP</h3>
<p>Spatial plot may not be optimal to show regions where different cell
types colocalised in a spot. To help with this, we can draw a UMAP plot
coloured by the module scores we have previously calculated. Spots (one
dot in a UMAP) with high score for two modules may indicate
colocalisation of two different subsets.</p>
<pre class="r"><code>dat_fil &lt;- RunPCA(dat_fil, assay=&#39;SCT&#39;)</code></pre>
<pre><code>PC_ 1 
Positive:  FABP1, CEACAM6, LCN2, TRIM31, EREG, CXCL3, CXCL2, MYC, SELENBP1, LRATD1 
       GPX2, CEACAM5, AREG, CD24, KRT23, PI3, CCL20, MCM4, NQO1, ELF3 
       CLDN4, S100P, GPRC5A, ST14, HIST1H1B, EPCAM, PHGR1, SPINK1, FERMT1, TK1 
Negative:  MGP, TAGLN, C3, ACTA2, GREM1, DCN, COL6A3, COL6A2, CALD1, MYL9 
       SFRP2, C1S, MMP2, ACTG2, LUM, COL3A1, COL1A2, C1R, FBN1, CNN1 
       COL1A1, TIMP2, AEBP1, ELN, MYLK, IGHG1, FBLN1, TNC, THBS2, IGFBP5 
PC_ 2 
Positive:  PIGR, FCGBP, TFF3, MUC2, JCHAIN, IGHA1, MUC4, REG4, PLA2G2A, CA2 
       CLCA1, TSPAN1, IGHM, KLF4, PARM1, ZG16, MLPH, CCL28, ADTRP, IGKC 
       SPINK4, MT2A, MT1X, PADI2, B3GALT5, AGR2, APOBR, CDHR5, NXPE1, IGLC1 
Negative:  FN1, KRT23, SPARC, TIMP3, EREG, COL1A1, COL1A2, SLC22A31, MMP11, MYC 
       MME, MCM4, COL3A1, IGFBP7, COL6A3, CALD1, ASCL2, COL5A2, NKX2-1, COL5A1 
       INHBA, HIST1H1B, LY6G6D, BGN, THY1, DAB2, LAPTM4B, SULF1, LUM, APCDD1 
PC_ 3 
Positive:  PI3, CCL20, LCN2, NOS2, MMP12, TRIM31, CXCL8, CXCL1, CXCL3, CEACAM1 
       CEACAM7, ID1, MUC13, S100A9, UBD, CXCL2, CFB, ATP1B1, WARS, DUOX2 
       SLC2A1, PLAUR, PDZK1IP1, MXRA5, IFI6, NAMPT, NCOA7, MMP1, SELENBP1, HMOX1 
Negative:  LGR5, ADH1C, MAOA, SORBS2, ABAT, HMGCS2, MME, AGR2, AMACR, SLC22A31 
       NXPE4, LEFTY1, ERN2, AIFM3, KIF12, RASSF10, CYP2W1, ATP2C2, TCEA3, PTPRN2 
       WNK2, ALDH1A1, ZBTB16, PRAC1, OR4D5, EPHB3, PTGDR, KLK1, FMO5, OR8D4 
PC_ 4 
Positive:  COL11A1, MMP11, CTHRC1, INHBA, ISLR, SFRP4, IGFBP3, ITGA11, TIMP3, THBS2 
       ADAM12, ITGBL1, COMP, AEBP1, SPARC, COL5A1, CDH11, COL8A1, COL5A2, POSTN 
       COL12A1, CDKN2B, SPON2, ELN, BGN, CTSK, COL1A1, COL10A1, HTRA3, ADAMTS2 
Negative:  DES, SYNM, SYNPO2, PRUNE2, CHRDL2, MYH11, SMTN, CSRP1, SORBS1, IGHG1 
       SPARCL1, HSPB6, ITGA5, PPP1R12B, CLU, KCNMA1, IGKC, TNS1, LMOD1, TNC 
       HSPB8, PLEKHO1, TPM2, MGP, FLNA, PLN, MYLK, ACTG2, CNN1, A2M 
PC_ 5 
Positive:  JCHAIN, MZB1, IGHG3, IGKC, IGHM, PIM2, IGHG1, PTGDS, DERL3, IGHA1 
       TENT5C, PECAM1, POU2AF1, LSP1, CYTIP, LYZ, EGFL7, IKZF1, LCP1, IL7R 
       IRF4, OLFM4, VWF, XBP1, CORO1A, CD38, CXCR4, FCRL5, TSC22D3, CD79A 
Negative:  MYLK, CKB, CNN1, MYL9, CSRP1, FLNA, TPM2, CALD1, ACTG2, LMOD1 
       FABP1, PALLD, SORBS1, TPM1, MYH11, ACTA2, TAGLN, TNS1, SMTN, SYNM 
       SYNPO2, CRYAB, PDLIM7, LIMS2, PRUNE2, FXYD3, PPP1R12B, DPYSL3, MSRB3, MAB21L2 </code></pre>
<pre class="r"><code># just to pick how many PCs to use to calculate UMAP.
ElbowPlot(dat_fil, ndims=50)</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-43-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-43-1">
Past versions of unnamed-chunk-43-1.png
</button>
</p>
<div id="fig-unnamed-chunk-43-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-43-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>dat_fil &lt;- RunUMAP(dat_fil, reduction = &#39;pca&#39;, dims=1:20)</code></pre>
<pre><code>Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to &#39;umap-learn&#39; and metric to &#39;correlation&#39;
This message will be shown once per session</code></pre>
<pre><code>10:49:28 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
<pre><code>10:49:28 Read 4137 rows and found 20 numeric columns</code></pre>
<pre><code>10:49:28 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
<pre><code>10:49:28 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
<pre><code>0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
<pre><code>[----|----|----|----|----|----|----|----|----|----|</code></pre>
<pre><code>**************************************************|
10:49:28 Writing NN index file to temp file /var/folders/7q/lqx0fzgn1zqfmlc5tx1c0n0r0004z0/T//RtmpnyRGZ5/file538732943bf7
10:49:28 Searching Annoy index using 1 thread, search_k = 3000
10:49:29 Annoy recall = 100%
10:49:29 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30
10:49:29 Initializing from normalized Laplacian + noise (using RSpectra)
10:49:29 Commencing optimization for 500 epochs, with 162824 positive edges
10:49:32 Optimization finished</code></pre>
<pre class="r"><code>FeaturePlot(
  dat_fil, 
  features = c(
    &quot;macro_score&quot;, &quot;malignant_score&quot;, &quot;SIRPA&quot;, &quot;CD47&quot;
  ),
  cols = viridis::viridis(50)
)</code></pre>
<p><img src="figure/visium_crc.Rmd/unnamed-chunk-44-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-44-1">
Past versions of unnamed-chunk-44-1.png
</button>
</p>
<div id="fig-unnamed-chunk-44-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/1a2b848a31540145d2c082a2f897d90c61dc8540/docs/figure/visium_crc.Rmd/unnamed-chunk-44-1.png" target="_blank">1a2b848</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="extra-stuff" class="section level2">
<h2>Extra stuff</h2>
<p>Maybe spatially variable genes? and deconvolution - if there is a
nice single cell data we can use. Maybe the h5ad cancer data from <a
href="https://www.nature.com/articles/s43018-024-00807-z">this paper</a>
can be used. Also merging two tissues?</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Australia/Melbourne
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] qs_0.27.2          patchwork_1.3.0    scales_1.3.0       ggplot2_3.5.1     
[5] Seurat_5.1.0       SeuratObject_5.0.2 sp_2.1-4           workflowr_1.7.1   

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3          rstudioapi_0.16.0          
  [3] jsonlite_1.8.9              magrittr_2.0.3             
  [5] spatstat.utils_3.1-0        farver_2.1.2               
  [7] rmarkdown_2.28              zlibbioc_1.50.0            
  [9] fs_1.6.4                    vctrs_0.6.5                
 [11] ROCR_1.0-11                 DelayedMatrixStats_1.26.0  
 [13] spatstat.explore_3.3-2      S4Arrays_1.4.1             
 [15] htmltools_0.5.8.1           SparseArray_1.4.8          
 [17] sass_0.4.9                  sctransform_0.4.1          
 [19] parallelly_1.38.0           KernSmooth_2.23-24         
 [21] bslib_0.8.0                 htmlwidgets_1.6.4          
 [23] ica_1.0-3                   plyr_1.8.9                 
 [25] plotly_4.10.4               zoo_1.8-12                 
 [27] cachem_1.1.0                whisker_0.4.1              
 [29] igraph_2.0.3                mime_0.12                  
 [31] lifecycle_1.0.4             pkgconfig_2.0.3            
 [33] Matrix_1.7-0                R6_2.5.1                   
 [35] fastmap_1.2.0               GenomeInfoDbData_1.2.12    
 [37] MatrixGenerics_1.16.0       fitdistrplus_1.2-1         
 [39] future_1.34.0               shiny_1.9.1                
 [41] digest_0.6.37               colorspace_2.1-1           
 [43] S4Vectors_0.42.1            ps_1.8.0                   
 [45] rprojroot_2.0.4             tensor_1.5                 
 [47] RSpectra_0.16-2             irlba_2.3.5.1              
 [49] GenomicRanges_1.56.2        labeling_0.4.3             
 [51] progressr_0.14.0            fansi_1.0.6                
 [53] spatstat.sparse_3.1-0       httr_1.4.7                 
 [55] polyclip_1.10-7             abind_1.4-8                
 [57] compiler_4.4.1              bit64_4.5.2                
 [59] withr_3.0.1                 viridis_0.6.5              
 [61] fastDummies_1.7.4           highr_0.11                 
 [63] MASS_7.3-61                 DelayedArray_0.30.1        
 [65] tools_4.4.1                 lmtest_0.9-40              
 [67] httpuv_1.6.15               future.apply_1.11.2        
 [69] goftest_1.2-3               glmGamPoi_1.16.0           
 [71] glue_1.7.0                  callr_3.7.6                
 [73] nlme_3.1-166                promises_1.3.0             
 [75] grid_4.4.1                  Rtsne_0.17                 
 [77] getPass_0.2-4               cluster_2.1.6              
 [79] reshape2_1.4.4              generics_0.1.3             
 [81] hdf5r_1.3.11                gtable_0.3.5               
 [83] spatstat.data_3.1-2         tidyr_1.3.1                
 [85] RApiSerialize_0.1.4         data.table_1.16.0          
 [87] XVector_0.44.0              stringfish_0.16.0          
 [89] utf8_1.2.4                  BiocGenerics_0.50.0        
 [91] spatstat.geom_3.3-3         RcppAnnoy_0.0.22           
 [93] ggrepel_0.9.6               RANN_2.6.2                 
 [95] pillar_1.9.0                stringr_1.5.1              
 [97] spam_2.10-0                 RcppHNSW_0.6.0             
 [99] later_1.3.2                 splines_4.4.1              
[101] dplyr_1.1.4                 lattice_0.22-6             
[103] bit_4.5.0                   survival_3.7-0             
[105] deldir_2.0-4                tidyselect_1.2.1           
[107] miniUI_0.1.1.1              pbapply_1.7-2              
[109] knitr_1.48                  git2r_0.33.0               
[111] gridExtra_2.3               IRanges_2.38.1             
[113] SummarizedExperiment_1.34.0 scattermore_1.2            
[115] stats4_4.4.1                xfun_0.47                  
[117] Biobase_2.64.0              matrixStats_1.4.1          
[119] UCSC.utils_1.0.0            stringi_1.8.4              
[121] lazyeval_0.2.2              yaml_2.3.10                
[123] evaluate_1.0.0              codetools_0.2-20           
[125] tibble_3.2.1                cli_3.6.3                  
[127] RcppParallel_5.1.9          uwot_0.2.2                 
[129] xtable_1.8-4                reticulate_1.39.0          
[131] munsell_0.5.1               processx_3.8.4             
[133] jquerylib_0.1.4             GenomeInfoDb_1.40.1        
[135] Rcpp_1.0.13                 globals_0.16.3             
[137] spatstat.random_3.3-2       png_0.1-8                  
[139] spatstat.univar_3.0-1       parallel_4.4.1             
[141] dotCall64_1.1-1             sparseMatrixStats_1.16.0   
[143] listenv_0.9.1               viridisLite_0.4.2          
[145] ggridges_0.5.6              crayon_1.5.3               
[147] leiden_0.4.3.1              purrr_1.0.2                
[149] rlang_1.1.4                 cowplot_1.1.3              </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
