<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Givanna Putri" />

<meta name="date" content="2024-11-20" />

<title>Visium part 2</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">ASI Spatial Workshop 2024</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Analysis Files
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="setup.html">Setup</a>
    </li>
    <li>
      <a href="stomics_01_intro.html">STOmics intro</a>
    </li>
    <li>
      <a href="stomics_02_basics.html">STOmics basics</a>
    </li>
    <li>
      <a href="stomics_03_analysis.html">STOmics analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ghar1821/asi_spatial_workshop_2024">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Visium part 2</h1>
<h4 class="author">Givanna Putri</h4>
<h4 class="date">2024-11-20</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-11-20
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>asi_spatial_workshop_2024/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R
Markdown file created these results, you’ll want to first commit it to
the Git repo. If you’re still working on the analysis, you can ignore
this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240925code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240925)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240925code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240925)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomghar1821asispatialworkshop2024tree85ecc5f6d2c8ff8802c79c687c0fdf13f15b890atargetblank85ecc5fa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/tree/85ecc5f6d2c8ff8802c79c687c0fdf13f15b890a" target="_blank">85ecc5f</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomghar1821asispatialworkshop2024tree85ecc5f6d2c8ff8802c79c687c0fdf13f15b890atargetblank85ecc5fa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/tree/85ecc5f6d2c8ff8802c79c687c0fdf13f15b890a" target="_blank">85ecc5f</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    data/.DS_Store
    Ignored:    data/single_cell/
    Ignored:    data/visium/
    Ignored:    output/visium/
    Ignored:    visium/

Untracked files:
    Untracked:  analysis/visium_02.Rmd

Unstaged changes:
    Modified:   analysis/visium_01.Rmd
    Modified:   analysis/visium_crc.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with
<code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>In this part of the visium workshop, we will introduce steps to
perform Quality Control (QC) and normalisation on visium data.</p>
</div>
<div id="load-libraries-and-data" class="section level2">
<h2>Load libraries and data</h2>
<pre class="r"><code>library(Seurat)
library(qs)
library(ggplot2)
library(scales)</code></pre>
<p>We will be using the Seurat object that we have previously loaded in
<a href="visium_01.html">part 1</a>.</p>
<pre class="r"><code>dat &lt;- qread(&quot;output/visium/visium_seurat.qs&quot;)
dat</code></pre>
<pre><code>An object of class Seurat 
18085 features across 4269 samples within 1 assay 
Active assay: Spatial (18085 features, 0 variable features)
 1 layer present: counts
 1 spatial field of view present: slice1</code></pre>
</div>
<div id="quality-control-qc" class="section level2">
<h2>Quality Control (QC)</h2>
<p>QC is part of data pre-processing process. The main purpose of it is
to remove stuff (e.g., spots and genes) that are of low quality. If not
removed, these can negatively impact your downstream analyses, and
potentially leading to false findings and interpretations.</p>
<p>Performing QC on the data can be seen as a delicate balancing act.
Overly aggressive QC will yield very clean data, but at the same token,
can remove stuff that would have otherwise led to an interesting
biology. Overly liberal QC may include many dud stuff (or just plain
noise) that makes interpretation hard. Thus, it is always important to
inspect the result of a given QC strategy, and fine tune as
required.</p>
<div id="qc-genes" class="section level3">
<h3>QC genes</h3>
<p>By default, <code>Load10X_Spatial</code> will load up all the genes
in the data. However, it is highly likely for some genes to only be
expressed (have raw count &gt; 0) in only a small number of spots. It
may well be that there are only really a tiny number of spots that
expressed this gene that it may well be a noise. Thus, it is best to
remove them as otherwise you may be running the risk of them negatively
impacting further downstream analyses.</p>
<p>To work out the number of spots expressing a given gene, you can use
a histogram.</p>
<pre class="r"><code># Count how many spots each gene is expressed in.
spots_per_gene &lt;- rowSums(dat[[&quot;Spatial&quot;]]$counts &gt; 0)

# Plot as histogram
# Have to convert spots_per_gene to data.frame first, the column is spots
ggplot(data.frame(spots = spots_per_gene), aes(x=spots)) + 
  geom_histogram(binwidth = 1, colour = &#39;blue&#39;) +
  geom_vline(xintercept = 100, colour = &#39;red&#39;) +
  theme_minimal() +
  labs(title = &quot;Number of Spots per Gene&quot;,
       x = &quot;Number of Spots&quot;,
       y = &quot;Number of Genes&quot;) +
  scale_x_continuous(breaks = pretty_breaks(10)) +
  scale_y_continuous(breaks = pretty_breaks(10))</code></pre>
<p><img src="figure/visium_02.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The histogram above illustrates how many genes (y-axis) are expressed
in how many spots (x-axis). For example, the left most bar shows there
are 12 genes that are not expressed in any spots at all.</p>
<p>From the histogram, we can see the distribution of the gene
expressions, and work out a sensible threshold to filter out genes. In
other words, identifying the minimum number of spots a gene must be
expressed for it to be included in our data. For this dataset, we will
set the threshold to 100, i.e. for a gene to be included, it needs to be
expressed in at least 100 spots, which is around 2.3% of total spots we
have in the data.</p>
<pre class="r"><code># names will give you the gene names
genes_to_keep &lt;- names(spots_per_gene[spots_per_gene &gt;= 100])

# Subset the Seurat object
dat_fil &lt;- subset(dat, features = genes_to_keep)</code></pre>
<hr>
<details>
<summary>
Setting the threshold
</summary>
<hr>
<p>The threshold 100 is not fixed. This will vary depending on the
dataset you have on hand. So adjust this as you see fit.</p>
<p>You may notice that the value 100 seems high, especially if you have
some experience QCing some scRNAseq data. However, it is worth
remembering that visium is a spot based technology and that each spot
captures more than 1 cell.</p>
</details>
<hr>
</div>
<div id="qc-spots" class="section level3">
<h3>QC spots</h3>
<p>In addition to QC the genes, we should also QC the spots to make sure
that we remove spots that are low quality. These could be spots that lie
outside of the tissue, or have low library size or low number of genes
expressed, or expressed high mitochondrial genes.</p>
<div id="using-imagej-to-remove-spots-outside-of-the-tissue"
class="section level4">
<h4>Using ImageJ to remove spots outside of the tissue</h4>
<p>The vanilla 10X SpaceRanger pipeline automatically identifies the
spots that overlap with tissue. This is not always consistent or
reliable. Therefore, there is no reason why</p>
</div>
<div id="using-loupe-browser-to-remove-spots-outside-of-the-tissue"
class="section level4">
<h4>Using Loupe Browser to remove spots outside of the tissue</h4>
<p>To identify spots that lie outside of the tissue, we can use the
Loupe Browser software. Loupe browser is a visualisation software
produced by 10x to make it easier for users to interact with their data.
It can be downloaded from <a
href="https://www.10xgenomics.com/support/software/loupe-browser/latest">10x
website</a>.</p>
<p>To load the data up to Loupe browser, you need to have on hand the
Loupe browser file generated by spaceranger count. For this dataset, you
can download it from 10x website (choose Visium CytoAssist v2 P2 crc on
the right panel): <a
href="https://www.10xgenomics.com/products/visium-hd-spatial-gene-expression/dataset-human-crc"
class="uri">https://www.10xgenomics.com/products/visium-hd-spatial-gene-expression/dataset-human-crc</a>.</p>
<p>Using the loop browser, you can see which spots lie outside of the
tissue, mark them, and export their barcodes out:</p>
<video width="600" height="500" controls>
<source src="assets/visium_loupe_demo.mp4" type="video/mp4">
</video>
<hr>
<details>
<summary>
Removing spots
</summary>
<hr>
<p>If you happen to make a mistake when using the lasso tool, and
accidentally selected spots that you meant to keep, use the eraser tool
to “erase” those spots from the list by selecting them.</p>
</details>
<hr>
<p>After exporting the spots barcode in a csv file, we can load it up
and remove them from the seurat object:</p>
<pre class="r"><code>spots_to_remove &lt;- read.csv(&quot;data/visium/spots_to_remove_v2.csv&quot;)

# add it as a metadata
dat_fil[[]]$remove_spot_from_loupe &lt;- Cells(dat_fil) %in% spots_to_remove$Barcode

# store as separate object just for demo
dat_fil &lt;- subset(x = dat_fil, remove_spot_from_loupe == FALSE)
dat_fil</code></pre>
<pre><code>An object of class Seurat 
17991 features across 4006 samples within 1 assay 
Active assay: Spatial (17991 features, 0 variable features)
 1 layer present: counts
 1 spatial field of view present: slice1</code></pre>
</div>
</div>
<div id="qc-spots-based-on-library-size-and-genes-expressed"
class="section level3">
<h3>QC spots based on library size and genes expressed</h3>
<p>We should also remove spots which express low library size and/or low
number of genes expressed.</p>
<p>Few plots we can draw to check this are violin plots, histogram, and
plain scatter plots:</p>
<pre class="r"><code>plt1 &lt;- ggplot(dat_fil[[]], aes(x=nCount_Spatial)) + geom_histogram(bins = 100) +
  theme_classic() +
  scale_y_continuous(breaks = pretty_breaks(10)) +
  scale_x_continuous(breaks = pretty_breaks(10)) 
plt2 &lt;- ggplot(dat_fil[[]], aes(x=nFeature_Spatial)) + geom_histogram(bins = 100) +
  theme_classic() +
  scale_y_continuous(breaks = pretty_breaks(10)) +
  scale_x_continuous(breaks = pretty_breaks(10)) 

plt3 &lt;- VlnPlot(dat_fil, features = c(&quot;nCount_Spatial&quot;, &quot;nFeature_Spatial&quot;))</code></pre>
<pre><code>Warning: Default search for &quot;data&quot; layer in &quot;Spatial&quot; assay yielded no results;
utilizing &quot;counts&quot; layer instead.</code></pre>
<pre class="r"><code>plt4 &lt;- FeatureScatter(dat_fil, feature1 = &quot;nCount_Spatial&quot;, feature2 = &quot;nFeature_Spatial&quot;) +
  scale_y_continuous(breaks = pretty_breaks(10)) +
  scale_x_continuous(breaks = pretty_breaks(10)) +
  theme_bw()

plt1 + plt2 + plt3 + plt4</code></pre>
<p><img src="figure/visium_02.Rmd/unnamed-chunk-6-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>The histograms and violin plots are great for determining the minimum
cutoff for library size and/or number of genes expressed, while the
scatter plot is handy to check whether the spots with low library size
also has low number of genes expressed, which make them the prime
candidates for removal.</p>
<p>For this workshop, we will exclude spots that express less than 3000
genes. Incidentally, this threshold will also automatically exclude
spots with small library size.</p>
<pre class="r"><code>dat_fil &lt;- subset(x = dat_fil, nFeature_Spatial &gt;= 3000)
dat_fil</code></pre>
<pre><code>An object of class Seurat 
17991 features across 3969 samples within 1 assay 
Active assay: Spatial (17991 features, 0 variable features)
 1 layer present: counts
 1 spatial field of view present: slice1</code></pre>
<hr>
<details>
<summary>
Good know
</summary>
<hr>
<h3>
Keeping spots with high library size
</h3>
<p>Unlike single cell in which cells that have high library size are
often removed because they may well represent doublets or multiplets,
for visium data, spots with high library size may not necessarily mean
bad spots as visium data is not single cell resolution. Those spots may
well just contain an assortment of cell types which may lead to
interesting biology!</p>
</details>
<hr>
</div>
<div id="qc-spots-based-on-mitochondria-genes-expression"
class="section level3">
<h3>QC spots based on mitochondria genes expression</h3>
<p>Just like in single cell, we can also QC the spots based on the
percentage of mitochondria genes expressed. Spots that have excessively
high expression of mitochondria genes may contain low quality or dying
cells.</p>
<pre class="r"><code># This function will compute, for each spot, percentage of transcripts that mapped to mitochondria genes
dat_fil[[&#39;percent_mt&#39;]] &lt;- PercentageFeatureSet(dat_fil, pattern = &#39;^MT-&#39;)

# Visualize them
plt1 &lt;- VlnPlot(dat_fil, features = &quot;percent_mt&quot;)</code></pre>
<pre><code>Warning: Default search for &quot;data&quot; layer in &quot;Spatial&quot; assay yielded no results;
utilizing &quot;counts&quot; layer instead.</code></pre>
<pre class="r"><code>plt2 &lt;- FeatureScatter(dat_fil, feature1 = &#39;nCount_Spatial&#39;, feature2 = &#39;percent_mt&#39;)

plt1 + plt2</code></pre>
<p><img src="figure/visium_02.Rmd/unnamed-chunk-8-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>We will remove spots that have &gt; 12% mitochondria genes.</p>
<pre class="r"><code>dat_fil &lt;- subset(dat_fil, subset = percent_mt &lt; 12)
dat_fil</code></pre>
<pre><code>An object of class Seurat 
17991 features across 3959 samples within 1 assay 
Active assay: Spatial (17991 features, 0 variable features)
 1 layer present: counts
 1 spatial field of view present: slice1</code></pre>
<p>After QC, we ended up with 3,950 spots.</p>
</div>
<div id="export-data-out" class="section level3">
<h3>Export data out</h3>
<pre class="r"><code>qsave(dat_fil, &quot;output/visium_seurat_qced.qs&quot;)</code></pre>
</div>
</div>
<div id="normalisation" class="section level2">
<h2>Normalisation</h2>
<p>Why is this necessary? The same reason as any other RNA seq based
protocol (including single cell). Differences in library size may well
be due to variances in sampling of RNA molecules. During library
preparation, it is not possible to capture all molecules and sequence
them. Thus, the unevenness of library sizes and differences in gene
expression across spots, may well be due to variance in sampling. There
are also other technical variations that contributed to the differences
that need to be accounted for, e.g., Probe hybridization efficiency and
PCR amplification biases.</p>
<p>Normalisation aims to adjust differences introduced due to technical
variations. In single cell, there are many methods that have been
developed to do this (see <a
href="https://www.nature.com/articles/s41592-023-01814-1">here</a> for
benchmarking paper). Admittedly, not all of them will be suitable for
visium data. For this tutorial, we will demonstrate classic log
normalisation and SCTransform.</p>
<div id="running-sctransform" class="section level3">
<h3>Running SCTransform</h3>
<p>Briefly SCTransform fits a regularised negative binomial model to the
raw count matrix, using sequencing depth as a covariate.</p>
<p>Note, SCT requires glmGamPoi package:
<code>BiocManager::install('glmGamPoi')</code></p>
<pre class="r"><code># Might need to increase the size limit for global variables so parallel workers
# can access it. 2GB should be sufficient.
options(future.globals.maxSize = 2000 * 1024^2)


# Note, return.only.var.genes can be set to TRUE if we want to get back 
# the expression of only highly variable genes in scale.data
dat_fil &lt;- SCTransform(dat_fil, assay = &quot;Spatial&quot;, verbose = TRUE, return.only.var.genes = FALSE)
# Inspect after
dat_fil</code></pre>
<pre><code>An object of class Seurat 
35982 features across 3959 samples within 2 assays 
Active assay: SCT (17991 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: Spatial
 1 spatial field of view present: slice1</code></pre>
<p>By default, SCTransform will store the output in a different assay
called “SCT”.</p>
<pre class="r"><code>head(dat_fil[[&#39;SCT&#39;]]$data, c(20, 5))</code></pre>
<pre><code>20 x 5 sparse Matrix of class &quot;dgCMatrix&quot;
         AACAATGTGCTCCGAG-1 AACACCATTCGCATAC-1 AACACGACAACGGAGT-1
SAMD11            .                  1.0986123          .        
NOC2L             0.6931472          1.3862944          1.6094379
KLHL17            .                  0.6931472          .        
PLEKHN1           0.6931472          0.6931472          1.6094379
PERM1             .                  .                  .        
HES4              0.6931472          1.6094379          .        
ISG15             1.7917595          2.8332133          1.9459101
AGRN              1.0986123          1.7917595          2.1972246
RNF223            .                  .                  .        
C1orf159          1.0986123          0.6931472          .        
TTLL10            0.6931472          .                  .        
TNFRSF18          .                  0.6931472          0.6931472
TNFRSF4           .                  0.6931472          .        
SDF4              2.7080502          3.3672958          2.7080502
B3GALT6           1.3862944          1.0986123          1.6094379
C1QTNF12          0.6931472          .                  .        
UBE2J2            1.3862944          2.1972246          1.7917595
SCNN1D            .                  .                  .        
ACAP3             2.0794415          2.1972246          0.6931472
PUSL1             1.0986123          0.6931472          0.6931472
         AACACGCAGATAACAA-1 AACACTCGTGAGCTTC-1
SAMD11            0.6931472          0.6931472
NOC2L             1.0986123          1.6094379
KLHL17            1.0986123          1.3862944
PLEKHN1           0.6931472          .        
PERM1             .                  .        
HES4              0.6931472          1.7917595
ISG15             1.3862944          2.3025851
AGRN              1.6094379          2.1972246
RNF223            .                  .        
C1orf159          0.6931472          .        
TTLL10            .                  .        
TNFRSF18          0.6931472          1.3862944
TNFRSF4           0.6931472          .        
SDF4              3.2958369          2.7080502
B3GALT6           1.0986123          .        
C1QTNF12          .                  .        
UBE2J2            1.7917595          1.0986123
SCNN1D            .                  .        
ACAP3             1.3862944          2.3025851
PUSL1             1.3862944          .        </code></pre>
<hr>
<details>
<summary>
Note on layers
</summary>
<hr>
<h3>
Keeping spots with high library size
</h3>
<p>As previously discussed in the beginning, an assay can contain
multiple layers. After running SCTransfrom, you shall notice a new assay
call SCT containing SCTransformed counts. Specifically, counts layer
contains corrected UMI counts, data contains log1p - log(1 + counts) of
the corrected UMI counts, scale.data contains the counts in the data
layer transformed using pearson residuals calculated by SCTransform.</p>
</details>
<hr>
</div>
<div id="comparison-with-log-normalisation" class="section level3">
<h3>Comparison with log normalisation</h3>
<p>Log normalisation is commonly used for single cell data. It is also
widely used for normalising visium data where for each spot and gene, we
divide the expression by the spot’s library size and multiply the value
by a size factor (say 10,000). Then we perform log1p operation where we
run logarithmic transformation after adding value of 1. Why the
addition, because log of 0 is infinity.</p>
<pre class="r"><code># By default, this will run log CPTT
# The data layer will contain the normalised counts.
# The counts layer will contain the unnormalised count.
dat_fil &lt;- NormalizeData(dat_fil, verbose = FALSE, assay = &quot;Spatial&quot;)
dat_fil</code></pre>
<pre><code>An object of class Seurat 
35982 features across 3959 samples within 2 assays 
Active assay: SCT (17991 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: Spatial
 1 spatial field of view present: slice1</code></pre>
<p>Upon running this, you will notice an additional layer in the Spatial
assay call <code>data</code> which contains the output of log
normalisation function.</p>
<p>We can plot the output of SCTransform, log normalisation, and raw
count for a gene (say COL1A1), side by side to see what they look
like.</p>
<pre class="r"><code># Important to change assay before plotting to make sure it is the right one.

DefaultAssay(object = dat_fil) &lt;- &quot;SCT&quot;
plt_sct &lt;- SpatialFeaturePlot(
  dat_fil,
  features = &quot;COL1A1&quot;,
  image.alpha = 0.5,
  pt.size.factor = 4
) + ggtitle(&quot;SCT&quot;)


DefaultAssay(object = dat_fil) &lt;- &quot;Spatial&quot;
plt_lognorm &lt;- SpatialFeaturePlot(
  dat_fil,
  features = &quot;COL1A1&quot;,
  image.alpha = 0.5,
  pt.size.factor = 4
) + ggtitle(&quot;Log Norm&quot;)

# Plot raw data for comparison
plt_raw &lt;- SpatialFeaturePlot(
  dat_fil,
  features = &quot;COL1A1&quot;,
  image.alpha = 0.5,
  pt.size.factor = 4,
  slot = &#39;counts&#39;
) + ggtitle(&quot;Raw&quot;)

# Put all 3 plots side by side for comparison
plt_raw + plt_sct + plt_lognorm</code></pre>
<p><img src="figure/visium_02.Rmd/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;" /></p>
<hr>
<details>
<summary>
SCTransform vs Log Norm
</summary>
<hr>
<p>There are pros and cons on choosing either SCT or log normalisation.
From the plot, you can see they are different. Literatures say SCT
mitigates the correlation between library size and normalised counts
which log normalisation failed to do so. On the other hand, log
normalisation is very simple (you can compute it by hand if you like!)
and fast to compute, very well established, intuitive, and compatible
with a variety of analysis tools and pipelines. While SCT may be more
robust for data with high variability in sequencing depth, at the end of
the day, choose whichever method that you are comfortable with.</p>
</details>
<hr>
<p>For this tutorial, we will use the normalised counts generated by
SCTransform.</p>
<pre class="r"><code># Set the default assay to SCT
DefaultAssay(dat_fil) &lt;- &quot;SCT&quot;

# save so we can reload if we need to
qsave(dat_fil, &quot;output/visium_seurat_qced_norm.qs&quot;)</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Australia/Sydney
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] scales_1.3.0       ggplot2_3.5.1      qs_0.27.2          Seurat_5.1.0      
[5] SeuratObject_5.0.2 sp_2.1-4           workflowr_1.7.1   

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3          rstudioapi_0.16.0          
  [3] jsonlite_1.8.9              magrittr_2.0.3             
  [5] ggbeeswarm_0.7.2            spatstat.utils_3.1-0       
  [7] farver_2.1.2                rmarkdown_2.28             
  [9] zlibbioc_1.50.0             fs_1.6.5                   
 [11] vctrs_0.6.5                 ROCR_1.0-11                
 [13] DelayedMatrixStats_1.26.0   spatstat.explore_3.3-2     
 [15] S4Arrays_1.4.1              htmltools_0.5.8.1          
 [17] SparseArray_1.4.8           sass_0.4.9                 
 [19] sctransform_0.4.1           parallelly_1.38.0          
 [21] KernSmooth_2.23-24          bslib_0.8.0                
 [23] htmlwidgets_1.6.4           ica_1.0-3                  
 [25] plyr_1.8.9                  plotly_4.10.4              
 [27] zoo_1.8-12                  cachem_1.1.0               
 [29] whisker_0.4.1               igraph_2.0.3               
 [31] mime_0.12                   lifecycle_1.0.4            
 [33] pkgconfig_2.0.3             Matrix_1.7-0               
 [35] R6_2.5.1                    fastmap_1.2.0              
 [37] GenomeInfoDbData_1.2.12     MatrixGenerics_1.16.0      
 [39] fitdistrplus_1.2-1          future_1.34.0              
 [41] shiny_1.9.1                 digest_0.6.37              
 [43] colorspace_2.1-1            S4Vectors_0.42.1           
 [45] patchwork_1.3.0             ps_1.8.0                   
 [47] rprojroot_2.0.4             tensor_1.5                 
 [49] RSpectra_0.16-2             irlba_2.3.5.1              
 [51] GenomicRanges_1.56.2        labeling_0.4.3             
 [53] progressr_0.14.0            fansi_1.0.6                
 [55] spatstat.sparse_3.1-0       httr_1.4.7                 
 [57] polyclip_1.10-7             abind_1.4-8                
 [59] compiler_4.4.1              withr_3.0.1                
 [61] fastDummies_1.7.4           highr_0.11                 
 [63] MASS_7.3-61                 DelayedArray_0.30.1        
 [65] tools_4.4.1                 vipor_0.4.7                
 [67] lmtest_0.9-40               beeswarm_0.4.0             
 [69] httpuv_1.6.15               future.apply_1.11.2        
 [71] goftest_1.2-3               glmGamPoi_1.16.0           
 [73] glue_1.7.0                  callr_3.7.6                
 [75] nlme_3.1-166                promises_1.3.0             
 [77] grid_4.4.1                  Rtsne_0.17                 
 [79] getPass_0.2-4               cluster_2.1.6              
 [81] reshape2_1.4.4              generics_0.1.3             
 [83] gtable_0.3.5                spatstat.data_3.1-2        
 [85] tidyr_1.3.1                 RApiSerialize_0.1.4        
 [87] data.table_1.16.0           XVector_0.44.0             
 [89] stringfish_0.16.0           utf8_1.2.4                 
 [91] BiocGenerics_0.50.0         spatstat.geom_3.3-3        
 [93] RcppAnnoy_0.0.22            ggrepel_0.9.6              
 [95] RANN_2.6.2                  pillar_1.9.0               
 [97] stringr_1.5.1               spam_2.10-0                
 [99] RcppHNSW_0.6.0              later_1.3.2                
[101] splines_4.4.1               dplyr_1.1.4                
[103] lattice_0.22-6              survival_3.7-0             
[105] deldir_2.0-4                tidyselect_1.2.1           
[107] miniUI_0.1.1.1              pbapply_1.7-2              
[109] knitr_1.48                  git2r_0.33.0               
[111] gridExtra_2.3               IRanges_2.38.1             
[113] SummarizedExperiment_1.34.0 scattermore_1.2            
[115] stats4_4.4.1                xfun_0.47                  
[117] Biobase_2.64.0              matrixStats_1.4.1          
[119] UCSC.utils_1.0.0            stringi_1.8.4              
[121] lazyeval_0.2.2              yaml_2.3.10                
[123] evaluate_1.0.0              codetools_0.2-20           
[125] tibble_3.2.1                cli_3.6.3                  
[127] RcppParallel_5.1.9          uwot_0.2.2                 
[129] xtable_1.8-4                reticulate_1.39.0          
[131] munsell_0.5.1               processx_3.8.4             
[133] jquerylib_0.1.4             GenomeInfoDb_1.40.1        
[135] Rcpp_1.0.13                 globals_0.16.3             
[137] spatstat.random_3.3-2       png_0.1-8                  
[139] ggrastr_1.0.2               spatstat.univar_3.0-1      
[141] parallel_4.4.1              dotCall64_1.1-1            
[143] sparseMatrixStats_1.16.0    listenv_0.9.1              
[145] viridisLite_0.4.2           ggridges_0.5.6             
[147] crayon_1.5.3                leiden_0.4.3.1             
[149] purrr_1.0.2                 rlang_1.1.4                
[151] cowplot_1.1.3              </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
