<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Givanna Putri, Thomas O’neil" />


<title>Visium Part 4</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">ASI Spatial Workshop 2024</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Analysis Files
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="setup.html">Setup</a>
    </li>
    <li>
      <a href="visium_03.html">Visium part 3</a>
    </li>
    <li>
      <a href="stomics_01_intro.html">STOmics intro</a>
    </li>
    <li>
      <a href="stomics_02_basics.html">STOmics basics</a>
    </li>
    <li>
      <a href="stomics_03_analysis.html">STOmics analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ghar1821/asi_spatial_workshop_2024">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Visium Part 4</h1>
<h4 class="author">Givanna Putri, Thomas O’neil</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-11-21
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>asi_spatial_workshop_2024/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R
Markdown file created these results, you’ll want to first commit it to
the Git repo. If you’re still working on the analysis, you can ignore
this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240925code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240925)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240925code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240925)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomghar1821asispatialworkshop2024tree58ca9b12df602abb333c276a65decf5aa5cb4a6ftargetblank58ca9b1a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/tree/58ca9b12df602abb333c276a65decf5aa5cb4a6f" target="_blank">58ca9b1</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomghar1821asispatialworkshop2024tree58ca9b12df602abb333c276a65decf5aa5cb4a6ftargetblank58ca9b1a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/tree/58ca9b12df602abb333c276a65decf5aa5cb4a6f" target="_blank">58ca9b1</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    data/.DS_Store
    Ignored:    data/single_cell/
    Ignored:    data/visium/
    Ignored:    output/visium/
    Ignored:    visium/

Untracked files:
    Untracked:  analysis/visium_04.Rmd

Unstaged changes:
    Modified:   analysis/visium_crc.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with
<code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>In this part of the workshop, we will learn how to use module scoring
and deconvolution to resolve the cell types in each spot. From there, we
will then explore what interesting biology visium data can offer,
including finding spatially variable genes.</p>
</div>
<div id="load-packages-and-data" class="section level2">
<h2>Load packages and data</h2>
<pre class="r"><code>library(Seurat)
library(qs)
library(ggplot2)
library(scales)
library(spacexr)</code></pre>
<p>We will be using the Seurat object that we have QCed and normalised
in <a href="visium_02.html">part 2</a>.</p>
<pre class="r"><code>dat &lt;- qread(&quot;output/visium_seurat_qced_norm.qs&quot;)
dat</code></pre>
<pre><code>An object of class Seurat 
35982 features across 3959 samples within 2 assays 
Active assay: SCT (17991 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: Spatial
 1 spatial field of view present: slice1</code></pre>
</div>
<div id="module-scoring" class="section level2">
<h2>Module scoring</h2>
<p>Module scoring is a simple yet powerful way of
<strong><em>inferring</em></strong> what cell types are likely to be
present in which spot based on a set of signatures/marker genes (a
module). It uses predefined sets of signature or marker genes (modules)
to calculate scores that reflect the expression levels of these genes
across different spots.</p>
<p>It is particularly useful when you have predefined cell types and
their corresponding marker genes, allowing you to assess the presence or
absence of these cell types in your data. Module scoring can be
advantageous when you have the signatures of the cell types that are of
interest, or if you have previously analysed data from different omics
(e.g., bulk RNA seq) and derived a set of signatures you want to focus
on.</p>
<pre class="r"><code># for finding genes
# Features(dat_fil)[grep(&quot;S100B.*&quot;, Features(dat_fil))]

genes_to_score &lt;- list(
  endothelial = c(&quot;PLVAP&quot;, &quot;VWF&quot;, &quot;PECAM1&quot;, &quot;ERG&quot;),
  tumour = c(&quot;EPCAM&quot;, &quot;EGFR&quot;, &quot;CEACAM6&quot;, &quot;MUC1&quot;),
  fibroblast = c(&quot;COL1A1&quot;, &quot;COL1A2&quot;, &quot;VIM&quot;, &quot;FAP&quot;, &quot;PDGFRA&quot;, &quot;THY1&quot;, &quot;S100A4&quot;),
  bcell = c(&quot;PTPRC&quot;, &quot;MS4A1&quot;, &quot;CD79A&quot;, &quot;MZB1&quot;),
  tcell = c(&quot;PTPRC&quot;, &quot;CD3D&quot;, &quot;CD3E&quot;, &quot;CD3G&quot;),
  myeloid = c(&quot;PTPRC&quot;, &quot;ITGAX&quot;, &quot;ITGAM&quot;, &quot;CD14&quot;, &quot;LYZ&quot;, &quot;LILRA4&quot;, &quot;IRF7&quot;)
)

dat &lt;- AddModuleScore(
  object = dat, features = genes_to_score,
  seed = 42
)

head(dat[[]])</code></pre>
<pre><code>                      orig.ident nCount_Spatial nFeature_Spatial patient_id
AACAATGTGCTCCGAG-1 SeuratProject         159478            12671         P2
AACACCATTCGCATAC-1 SeuratProject           5719             3393         P2
AACACGACAACGGAGT-1 SeuratProject         244734            13369         P2
AACACGCAGATAACAA-1 SeuratProject          12387             6003         P2
AACACTCGTGAGCTTC-1 SeuratProject          55548            10947         P2
AACAGCCTCCTGACTA-1 SeuratProject         270289            13723         P2
                   remove_spot_from_loupe percent_mt nCount_SCT nFeature_SCT
AACAATGTGCTCCGAG-1                  FALSE   3.507023      78899        11013
AACACCATTCGCATAC-1                  FALSE   2.080783      71715        12165
AACACGACAACGGAGT-1                  FALSE   3.377939      78551        10577
AACACGCAGATAACAA-1                  FALSE   2.882054      71631        12235
AACACTCGTGAGCTTC-1                  FALSE   2.385324      76821        10931
AACAGCCTCCTGACTA-1                  FALSE   3.331212      79104        10992
                      Cluster1    Cluster2   Cluster3    Cluster4    Cluster5
AACAATGTGCTCCGAG-1 -0.48331890  0.78324471 -0.8296040 -0.52937960 -0.35567254
AACACCATTCGCATAC-1  0.41389772 -0.04350017  1.5741999 -0.05605231  0.11432363
AACACGACAACGGAGT-1  0.22576106  1.09322111 -0.1339130 -0.30391066 -0.35184381
AACACGCAGATAACAA-1  1.21610364 -0.16191345  1.8153155 -0.23086647 -0.08703452
AACACTCGTGAGCTTC-1 -0.03386751  0.45520543  1.6151196 -0.06641174  0.37234615
AACAGCCTCCTGACTA-1  0.36436415  1.03264208  0.1069596 -0.17954790 -0.38298954
                       Cluster6
AACAATGTGCTCCGAG-1 -0.913391613
AACACCATTCGCATAC-1  0.179063881
AACACGACAACGGAGT-1 -0.377623635
AACACGCAGATAACAA-1  0.504274609
AACACTCGTGAGCTTC-1 -0.114673812
AACAGCCTCCTGACTA-1  0.007111593</code></pre>
<p>For a given module, a score is computed as the average expression of
genes in the module, adjusted by the aggregated expression of a control
set of genes (defaults to all genes in the data). A positive score would
suggest that this module of genes is expressed in a particular spot more
highly than would be expected (given the average expression of this
module across the population). Importantly, the scores are not absolute
scores, meaning you should not compare them across different
experiments.</p>
<p>However, it is important to note that the score produces only
represent the <strong><em>relative enrichment</em></strong> of a set of
genes in a spot, and that it represents neither the proportion nor the
absolute count of a cell type therein. The score only provides the
probability for a given cell type to be present in a spot.</p>
<p>In the example above, we have 6 modules defined in the
<code>genes_to_score</code> list. By default, Seurat will store the
scores as new columns in the spot metadata, named Cluster1, Cluster2, …,
Cluster6. Each column corresponds to 1 module. Cluster1 denotes score
for endothelial cells, Cluster2 for tumour cells, etc. For ease of
subsequent interpretation, we should rename these.</p>
<pre class="r"><code>dat[[names(genes_to_score)]] &lt;- dat[[paste0(&quot;Cluster&quot;, seq(6))]]
head(dat[[]])</code></pre>
<pre><code>                      orig.ident nCount_Spatial nFeature_Spatial patient_id
AACAATGTGCTCCGAG-1 SeuratProject         159478            12671         P2
AACACCATTCGCATAC-1 SeuratProject           5719             3393         P2
AACACGACAACGGAGT-1 SeuratProject         244734            13369         P2
AACACGCAGATAACAA-1 SeuratProject          12387             6003         P2
AACACTCGTGAGCTTC-1 SeuratProject          55548            10947         P2
AACAGCCTCCTGACTA-1 SeuratProject         270289            13723         P2
                   remove_spot_from_loupe percent_mt nCount_SCT nFeature_SCT
AACAATGTGCTCCGAG-1                  FALSE   3.507023      78899        11013
AACACCATTCGCATAC-1                  FALSE   2.080783      71715        12165
AACACGACAACGGAGT-1                  FALSE   3.377939      78551        10577
AACACGCAGATAACAA-1                  FALSE   2.882054      71631        12235
AACACTCGTGAGCTTC-1                  FALSE   2.385324      76821        10931
AACAGCCTCCTGACTA-1                  FALSE   3.331212      79104        10992
                      Cluster1    Cluster2   Cluster3    Cluster4    Cluster5
AACAATGTGCTCCGAG-1 -0.48331890  0.78324471 -0.8296040 -0.52937960 -0.35567254
AACACCATTCGCATAC-1  0.41389772 -0.04350017  1.5741999 -0.05605231  0.11432363
AACACGACAACGGAGT-1  0.22576106  1.09322111 -0.1339130 -0.30391066 -0.35184381
AACACGCAGATAACAA-1  1.21610364 -0.16191345  1.8153155 -0.23086647 -0.08703452
AACACTCGTGAGCTTC-1 -0.03386751  0.45520543  1.6151196 -0.06641174  0.37234615
AACAGCCTCCTGACTA-1  0.36436415  1.03264208  0.1069596 -0.17954790 -0.38298954
                       Cluster6 endothelial      tumour fibroblast       bcell
AACAATGTGCTCCGAG-1 -0.913391613 -0.48331890  0.78324471 -0.8296040 -0.52937960
AACACCATTCGCATAC-1  0.179063881  0.41389772 -0.04350017  1.5741999 -0.05605231
AACACGACAACGGAGT-1 -0.377623635  0.22576106  1.09322111 -0.1339130 -0.30391066
AACACGCAGATAACAA-1  0.504274609  1.21610364 -0.16191345  1.8153155 -0.23086647
AACACTCGTGAGCTTC-1 -0.114673812 -0.03386751  0.45520543  1.6151196 -0.06641174
AACAGCCTCCTGACTA-1  0.007111593  0.36436415  1.03264208  0.1069596 -0.17954790
                         tcell      myeloid
AACAATGTGCTCCGAG-1 -0.35567254 -0.913391613
AACACCATTCGCATAC-1  0.11432363  0.179063881
AACACGACAACGGAGT-1 -0.35184381 -0.377623635
AACACGCAGATAACAA-1 -0.08703452  0.504274609
AACACTCGTGAGCTTC-1  0.37234615 -0.114673812
AACAGCCTCCTGACTA-1 -0.38298954  0.007111593</code></pre>
<p>We can visualise the scores using SpatialFeaturePlot:</p>
<pre class="r"><code>SpatialFeaturePlot(
  dat,
  features = names(genes_to_score),
  pt.size.factor = 5,
  image.alpha = 0.5
)</code></pre>
<p><img src="figure/visium_04.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="cell-type-deconvolution" class="section level2">
<h2>Cell type deconvolution</h2>
<p>Cell type deconvolution is very commonly used for analysing visium
data to infer the cell types that are present in a spot and their
proportions. It uses an annotated scRNAseq data as a reference to do
this. Importantly, for this to work as intended, it is important to make
sure that both the visium data and the scRNAseq data are of the same
biology (they do not need to be from the same sample).</p>
<p>For this workshop, we will use the <a
href="https://www.nature.com/articles/s41587-021-00830-w">RCTD
deconvolution algorithm</a> provided by the <a
href="https://github.com/dmcable/spacexr">spacexr package</a>. For the
reference scRNAseq data, we will use one of the scRNAseq data that
accompanies the visium data.</p>
<pre class="r"><code>sc_data &lt;- qread(&quot;data/single_cell/seurat_object_10x.qs&quot;)
DimPlot(sc_data, reduction = &quot;umap&quot;, group.by = c(&quot;Level1&quot;, &quot;Level2&quot;, &quot;Patient&quot;), ncol=2)</code></pre>
<pre><code>Rasterizing points since number of points exceeds 100,000.
To disable this behavior set `raster=FALSE`
Rasterizing points since number of points exceeds 100,000.
To disable this behavior set `raster=FALSE`
Rasterizing points since number of points exceeds 100,000.
To disable this behavior set `raster=FALSE`</code></pre>
<p><img src="figure/visium_04.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Prepare the data - note that: RCTD requires raw counts.</p>
<pre class="r"><code># Prepare the single cell data
sc_counts &lt;- sc_data[[&#39;RNA&#39;]]$counts
lbls &lt;- as.factor(sc_data$Level1)
nUMI &lt;- sc_data$nCount_originalexp

ref_sc &lt;- Reference(
  counts = sc_counts, 
  cell_types = lbls,
  nUMI = nUMI
)</code></pre>
<pre><code>Warning in Reference(counts = sc_counts, cell_types = lbls, nUMI = nUMI):
Reference: nUMI does not match colSums of counts. If this is unintended, please
correct this discrepancy. If this is intended, there is no problem.</code></pre>
<pre><code>Warning in Reference(counts = sc_counts, cell_types = lbls, nUMI = nUMI):
Reference: number of cells per cell type is 64750, larger than maximum
allowable of 10000. Downsampling number of cells to: 10000</code></pre>
<pre class="r"><code># Prepare the visium data
spat_counts &lt;- dat[[&#39;Spatial&#39;]]$counts
spat_nUMI &lt;- colSums(spat_counts)
# Note GetTissueCoordinates is needed because RCTD wants the x,y coordinates of the spots
spot_coords &lt;- GetTissueCoordinates(dat)[, 1:2]

query_spat &lt;- SpatialRNA(
  coords = spot_coords,
  counts = spat_counts,
  nUMI = spat_nUMI
)</code></pre>
<p>Run RCTD - this can take some time!:</p>
<pre class="r"><code># set max_cores to a reasonable number, don&#39;t overdo it!
RCTD &lt;- create.RCTD(
  spatialRNA = query_spat, 
  reference = ref_sc, 
  max_cores = 13
)</code></pre>
<pre><code>Begin: process_cell_type_info</code></pre>
<pre><code>process_cell_type_info: number of cells in reference: 66066</code></pre>
<pre><code>process_cell_type_info: number of genes in reference: 18076</code></pre>
<pre><code>
              B cells           Endothelial            Fibroblast 
                10000                  4872                 10000 
Intestinal Epithelial               Myeloid              Neuronal 
                 6618                 10000                   555 
        Smooth Muscle               T cells                 Tumor 
                 4021                 10000                 10000 </code></pre>
<pre><code>Warning in asMethod(object): sparse-&gt;dense coercion: allocating vector of size
1.3 GiB
Warning in asMethod(object): sparse-&gt;dense coercion: allocating vector of size
1.3 GiB
Warning in asMethod(object): sparse-&gt;dense coercion: allocating vector of size
1.3 GiB
Warning in asMethod(object): sparse-&gt;dense coercion: allocating vector of size
1.3 GiB
Warning in asMethod(object): sparse-&gt;dense coercion: allocating vector of size
1.3 GiB</code></pre>
<pre><code>End: process_cell_type_info</code></pre>
<pre><code>create.RCTD: getting regression differentially expressed genes: </code></pre>
<pre><code>get_de_genes: B cells found DE genes: 110</code></pre>
<pre><code>get_de_genes: Endothelial found DE genes: 343</code></pre>
<pre><code>get_de_genes: Fibroblast found DE genes: 234</code></pre>
<pre><code>get_de_genes: Intestinal Epithelial found DE genes: 304</code></pre>
<pre><code>get_de_genes: Myeloid found DE genes: 367</code></pre>
<pre><code>get_de_genes: Neuronal found DE genes: 252</code></pre>
<pre><code>get_de_genes: Smooth Muscle found DE genes: 267</code></pre>
<pre><code>get_de_genes: T cells found DE genes: 502</code></pre>
<pre><code>get_de_genes: Tumor found DE genes: 242</code></pre>
<pre><code>get_de_genes: total DE genes: 2207</code></pre>
<pre><code>create.RCTD: getting platform effect normalization differentially expressed genes: </code></pre>
<pre><code>get_de_genes: B cells found DE genes: 241</code></pre>
<pre><code>get_de_genes: Endothelial found DE genes: 660</code></pre>
<pre><code>get_de_genes: Fibroblast found DE genes: 441</code></pre>
<pre><code>get_de_genes: Intestinal Epithelial found DE genes: 655</code></pre>
<pre><code>get_de_genes: Myeloid found DE genes: 697</code></pre>
<pre><code>get_de_genes: Neuronal found DE genes: 562</code></pre>
<pre><code>get_de_genes: Smooth Muscle found DE genes: 543</code></pre>
<pre><code>get_de_genes: T cells found DE genes: 1114</code></pre>
<pre><code>get_de_genes: Tumor found DE genes: 703</code></pre>
<pre><code>get_de_genes: total DE genes: 4214</code></pre>
<pre class="r"><code># Note, full_mode option is reasonable for visium as we don&#39;t really want to restrict the
# number of cell types per spot.
# ‘doublet mode’ (at most 1-2 cell types per pixel), 
# ‘full mode’ (no restrictions on number of cell types), or 
# ‘multi mode’ (finitely many cell types per pixel, e.g. 3 or 4).
RCTD &lt;- run.RCTD(RCTD, doublet_mode = &quot;full&quot;)</code></pre>
<pre><code>fitBulk: decomposing bulk</code></pre>
<pre><code>chooseSigma: using initial Q_mat with sigma =  1</code></pre>
<pre><code>Likelihood value: 5738467.83621614</code></pre>
<pre><code>Sigma value:  0.84</code></pre>
<pre><code>Likelihood value: 5607917.30224369</code></pre>
<pre><code>Sigma value:  0.69</code></pre>
<pre><code>Likelihood value: 5495530.44451824</code></pre>
<pre><code>Sigma value:  0.61</code></pre>
<pre><code>Likelihood value: 5444226.35974223</code></pre>
<pre><code>Sigma value:  0.53</code></pre>
<pre><code>Likelihood value: 5403328.10936985</code></pre>
<pre><code>Sigma value:  0.45</code></pre>
<pre><code>Likelihood value: 5378205.55128649</code></pre>
<pre><code>Sigma value:  0.41</code></pre>
<pre><code>Likelihood value: 5373906.73229557</code></pre>
<pre><code>Sigma value:  0.41</code></pre>
<pre class="r"><code># qsave the object
qsave(RCTD, &quot;output/visium/rctd_decon_out.qs&quot;)</code></pre>
<p>Results of RCTD are stored in <code>@results</code>. There are lots
of information in there. Let’s focus on the <code>weights</code> which
gives estimated proportions of each cell type in each spot. It needs to
be normalised so they sum up to 1.</p>
<pre class="r"><code>norm_weights &lt;- normalize_weights(RCTD@results$weights) 
head(norm_weights)</code></pre>
<pre><code>6 x 9 Matrix of class &quot;dgeMatrix&quot;
                        B cells  Endothelial   Fibroblast Intestinal Epithelial
AACAATGTGCTCCGAG-1 7.823063e-05 7.823063e-05 7.823466e-05          7.823063e-05
AACACCATTCGCATAC-1 8.735751e-02 5.150152e-02 3.932158e-01          2.359593e-03
AACACGACAACGGAGT-1 7.020012e-05 2.057911e-03 7.020012e-05          1.471526e-04
AACACGCAGATAACAA-1 2.110692e-04 3.183124e-01 3.336937e-01          4.281214e-03
AACACTCGTGAGCTTC-1 9.750831e-05 7.962796e-03 3.004770e-01          9.750831e-05
AACAGCCTCCTGACTA-1 4.867774e-05 2.294518e-02 4.867774e-05          5.190048e-05
                        Myeloid    Neuronal Smooth Muscle      T cells
AACAATGTGCTCCGAG-1 7.823063e-05 0.001602927  7.823063e-05 7.823063e-05
AACACCATTCGCATAC-1 2.737532e-01 0.020714188  3.778596e-02 1.087397e-01
AACACGACAACGGAGT-1 7.020012e-05 0.000183992  7.020012e-05 7.020012e-05
AACACGCAGATAACAA-1 1.098895e-01 0.073236796  1.265166e-01 3.364758e-02
AACACTCGTGAGCTTC-1 3.956164e-03 0.010070445  3.451159e-02 6.982922e-02
AACAGCCTCCTGACTA-1 1.198421e-02 0.002564929  4.867774e-05 4.867774e-05
                          Tumor
AACAATGTGCTCCGAG-1 0.9978494547
AACACCATTCGCATAC-1 0.0245724532
AACACGACAACGGAGT-1 0.9972599437
AACACGCAGATAACAA-1 0.0002110692
AACACTCGTGAGCTTC-1 0.5729978100
AACAGCCTCCTGACTA-1 0.9622590714</code></pre>
<p>We can then attach it to our Seurat object and interrogate it.</p>
<pre class="r"><code>colnames(norm_weights) &lt;- paste0(&quot;decon_&quot;, colnames(norm_weights))
dat &lt;- AddMetaData(dat, metadata = norm_weights)
head(dat[[]])</code></pre>
<pre><code>                      orig.ident nCount_Spatial nFeature_Spatial patient_id
AACAATGTGCTCCGAG-1 SeuratProject         159478            12671         P2
AACACCATTCGCATAC-1 SeuratProject           5719             3393         P2
AACACGACAACGGAGT-1 SeuratProject         244734            13369         P2
AACACGCAGATAACAA-1 SeuratProject          12387             6003         P2
AACACTCGTGAGCTTC-1 SeuratProject          55548            10947         P2
AACAGCCTCCTGACTA-1 SeuratProject         270289            13723         P2
                   remove_spot_from_loupe percent_mt nCount_SCT nFeature_SCT
AACAATGTGCTCCGAG-1                  FALSE   3.507023      78899        11013
AACACCATTCGCATAC-1                  FALSE   2.080783      71715        12165
AACACGACAACGGAGT-1                  FALSE   3.377939      78551        10577
AACACGCAGATAACAA-1                  FALSE   2.882054      71631        12235
AACACTCGTGAGCTTC-1                  FALSE   2.385324      76821        10931
AACAGCCTCCTGACTA-1                  FALSE   3.331212      79104        10992
                      Cluster1    Cluster2   Cluster3    Cluster4    Cluster5
AACAATGTGCTCCGAG-1 -0.48331890  0.78324471 -0.8296040 -0.52937960 -0.35567254
AACACCATTCGCATAC-1  0.41389772 -0.04350017  1.5741999 -0.05605231  0.11432363
AACACGACAACGGAGT-1  0.22576106  1.09322111 -0.1339130 -0.30391066 -0.35184381
AACACGCAGATAACAA-1  1.21610364 -0.16191345  1.8153155 -0.23086647 -0.08703452
AACACTCGTGAGCTTC-1 -0.03386751  0.45520543  1.6151196 -0.06641174  0.37234615
AACAGCCTCCTGACTA-1  0.36436415  1.03264208  0.1069596 -0.17954790 -0.38298954
                       Cluster6 endothelial      tumour fibroblast       bcell
AACAATGTGCTCCGAG-1 -0.913391613 -0.48331890  0.78324471 -0.8296040 -0.52937960
AACACCATTCGCATAC-1  0.179063881  0.41389772 -0.04350017  1.5741999 -0.05605231
AACACGACAACGGAGT-1 -0.377623635  0.22576106  1.09322111 -0.1339130 -0.30391066
AACACGCAGATAACAA-1  0.504274609  1.21610364 -0.16191345  1.8153155 -0.23086647
AACACTCGTGAGCTTC-1 -0.114673812 -0.03386751  0.45520543  1.6151196 -0.06641174
AACAGCCTCCTGACTA-1  0.007111593  0.36436415  1.03264208  0.1069596 -0.17954790
                         tcell      myeloid decon_B cells decon_Endothelial
AACAATGTGCTCCGAG-1 -0.35567254 -0.913391613  7.823063e-05      7.823063e-05
AACACCATTCGCATAC-1  0.11432363  0.179063881  8.735751e-02      5.150152e-02
AACACGACAACGGAGT-1 -0.35184381 -0.377623635  7.020012e-05      2.057911e-03
AACACGCAGATAACAA-1 -0.08703452  0.504274609  2.110692e-04      3.183124e-01
AACACTCGTGAGCTTC-1  0.37234615 -0.114673812  9.750831e-05      7.962796e-03
AACAGCCTCCTGACTA-1 -0.38298954  0.007111593  4.867774e-05      2.294518e-02
                   decon_Fibroblast decon_Intestinal Epithelial decon_Myeloid
AACAATGTGCTCCGAG-1     7.823466e-05                7.823063e-05  7.823063e-05
AACACCATTCGCATAC-1     3.932158e-01                2.359593e-03  2.737532e-01
AACACGACAACGGAGT-1     7.020012e-05                1.471526e-04  7.020012e-05
AACACGCAGATAACAA-1     3.336937e-01                4.281214e-03  1.098895e-01
AACACTCGTGAGCTTC-1     3.004770e-01                9.750831e-05  3.956164e-03
AACAGCCTCCTGACTA-1     4.867774e-05                5.190048e-05  1.198421e-02
                   decon_Neuronal decon_Smooth Muscle decon_T cells
AACAATGTGCTCCGAG-1    0.001602927        7.823063e-05  7.823063e-05
AACACCATTCGCATAC-1    0.020714188        3.778596e-02  1.087397e-01
AACACGACAACGGAGT-1    0.000183992        7.020012e-05  7.020012e-05
AACACGCAGATAACAA-1    0.073236796        1.265166e-01  3.364758e-02
AACACTCGTGAGCTTC-1    0.010070445        3.451159e-02  6.982922e-02
AACAGCCTCCTGACTA-1    0.002564929        4.867774e-05  4.867774e-05
                    decon_Tumor
AACAATGTGCTCCGAG-1 0.9978494547
AACACCATTCGCATAC-1 0.0245724532
AACACGACAACGGAGT-1 0.9972599437
AACACGCAGATAACAA-1 0.0002110692
AACACTCGTGAGCTTC-1 0.5729978100
AACAGCCTCCTGACTA-1 0.9622590714</code></pre>
<pre class="r"><code>SpatialFeaturePlot(
  object = dat, 
  features = colnames(norm_weights),
  pt.size.factor = 5,
  image.alpha = 0.3,
  ncol = 4
)</code></pre>
<p><img src="figure/visium_04.Rmd/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>We could also draw UMAP coloured by the proportion of different cell
types:</p>
<pre class="r"><code>dat &lt;- RunPCA(dat, assay = &quot;SCT&quot;, verbose = FALSE)
dat &lt;- RunUMAP(dat, reduction = &quot;pca&quot;, dims = 1:10)</code></pre>
<pre><code>Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to &#39;umap-learn&#39; and metric to &#39;correlation&#39;
This message will be shown once per session</code></pre>
<pre><code>17:48:03 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
<pre><code>17:48:03 Read 3959 rows and found 10 numeric columns</code></pre>
<pre><code>17:48:03 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
<pre><code>17:48:03 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
<pre><code>0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
<pre><code>[----|----|----|----|----|----|----|----|----|----|</code></pre>
<pre><code>**************************************************|
17:48:04 Writing NN index file to temp file /var/folders/7q/lqx0fzgn1zqfmlc5tx1c0n0r0004z0/T//Rtmpjd23iu/file16884ab2f331
17:48:04 Searching Annoy index using 1 thread, search_k = 3000
17:48:04 Annoy recall = 100%
17:48:05 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30
17:48:05 Initializing from normalized Laplacian + noise (using RSpectra)
17:48:05 Commencing optimization for 500 epochs, with 152802 positive edges
17:48:08 Optimization finished</code></pre>
<pre class="r"><code>FeaturePlot(dat, features = colnames(norm_weights))</code></pre>
<p><img src="figure/visium_04.Rmd/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Where we can deduce that we have spots that contain mostly or wholly
one cell type, but we also have regions that have a mix of cell
types.</p>
<p>We can assign labels to the spots if the majority of the cell types
(&gt;80%) in the spot are made of one cell type:</p>
<pre class="r"><code>dat[[]]$majority_decon_per_spot &lt;- apply(norm_weights, 1, function(x) {
    # Find the maximum value in the row
    max_prop &lt;- max(x)
    if (max_prop &gt; 0.8) {
      return(gsub(&quot;decon_&quot;, &quot;&quot;, names(which.max(x))))
    } else {
      return(&quot;mix&quot;)
    }
    
})
DimPlot(dat, group.by = &quot;majority_decon_per_spot&quot;)</code></pre>
<p><img src="figure/visium_04.Rmd/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>SpatialDimPlot(dat, group.by = &quot;majority_decon_per_spot&quot;, pt.size.factor = 5, image.alpha = 0.5)</code></pre>
<p><img src="figure/visium_04.Rmd/unnamed-chunk-12-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="cross-checking-module-scoring-and-deconvolution"
class="section level2">
<h2>Cross checking module scoring and deconvolution</h2>
<p>For cell types that we have matching module scoring for, let’s
compare them against the results of RCTD:</p>
<pre class="r"><code>SpatialFeaturePlot(
  object = dat, 
  features = c(
    &quot;endothelial&quot;, &quot;decon_Endothelial&quot;,
    &quot;tumour&quot;, &quot;decon_Tumor&quot;,
    &quot;fibroblast&quot;, &quot;decon_Fibroblast&quot;,
    &quot;bcell&quot;, &quot;decon_B cells&quot;,
    &quot;tcell&quot;, &quot;decon_T cells&quot;,
    &quot;myeloid&quot;, &quot;decon_Myeloid&quot;,
    &quot;decon_Intestinal Epithelial&quot;, &quot;decon_Smooth Muscle&quot;,
    &quot;decon_Neuronal&quot;
  ),
  pt.size.factor = 6,
  image.alpha = 0.3,
  ncol = 4
)</code></pre>
<p><img src="figure/visium_04.Rmd/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Comparing the two, we can see areas where concordance we get some
concordance, in that areas that shows high module score tend to also
estimated to contain corresponding cell types after deconvolution.</p>
<p>However, we also see some differences. For example, we have a region
on the bottom right that shows high score for tumour module but the
deconvolution detected no tumour at all. Instead, they are deconvolved
into endothelial cells. On the flip side, we also have a region at the
bottom middle where deconvolution detected some T cells but module
scoring showed low signals in T cells module.</p>
<p>We can focus on one difference (say the tumour vs epithelial region)
and investigate further. It is worth noting that we did not have any
epithelial module in our module scoring.</p>
<pre class="r"><code>SpatialFeaturePlot(
  object = dat, 
  features = c(
    &quot;CDH1&quot;, &quot;KRT8&quot;, &quot;KRT19&quot;, &quot;CD44&quot;, &quot;VIM&quot;,
    genes_to_score[[&#39;tumour&#39;]]
  ),
  pt.size.factor = 5,
  image.alpha = 0.3,
  ncol = 5
)</code></pre>
<p><img src="figure/visium_04.Rmd/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>We have high and medium expression of CDH1 which means the epithelial
cells are retaining their cell-cell adhesion molecules. VIM and CD44 are
commonly associated with EMT, but they are lowly expressed. KRT8, KRT19,
and the remaining genes we have in the tumour module can be expressed on
epithelial cells. So together, it may well be that these are normal
epithelial cells rather than tumour cells. But because our tumour module
encapsulates genes that are commonly associated with epithelial cells,
our module scoring led us to briefly believe we have a tumour region
there.</p>
</div>
<div id="spatially-variable-genes" class="section level2">
<h2>Spatially variable genes</h2>
<p>Common analysis perform on spatial data is finding spatially variable
genes. These are genes that exhibit significant variability in
expression levels across different locations within the tissue.</p>
<pre class="r"><code># require HVGs to already be detected, which SCTransform would have done.
# the function below takes a while to run. consider reducing the number of HVGs
dat &lt;- FindSpatiallyVariableFeatures(
  dat,
  assay = &quot;SCT&quot;,
  features = VariableFeatures(dat)[1:500],
  selection.method = &#39;moransi&#39;
)</code></pre>
<pre><code>Computing Moran&#39;s I</code></pre>
<p>Let’s visualise some of them:</p>
<pre class="r"><code># a bit of data wrangling to get SVGs
svgs_meta &lt;- dat[[&#39;SCT&#39;]][[]]
svgs_meta &lt;- subset(svgs_meta, moransi.spatially.variable == TRUE)
svgs_meta &lt;- svgs_meta[order(svgs_meta$moransi.spatially.variable.rank), ]

SpatialFeaturePlot(
  dat,
  features = head(rownames(svgs_meta), 10),
  ncol = 5,
  image.alpha = 0.3,
  pt.size.factor = 5
) + ggtitle(&quot;Top 10 SVGs&quot;)</code></pre>
<p><img src="figure/visium_04.Rmd/unnamed-chunk-16-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>SVGs are different from highly variable genes (HVGs). HVGs are genes
that show significant variability in expression across different cells
or conditions, but this variability does not necessarily correlate with
spatial location. Let’s compare them - TMEM200B, CLEC10A are HVGs but
not SVGs while SYNM, MGP are HVGs:</p>
<pre class="r"><code>SpatialFeaturePlot(
  dat,
  features = c(head(rownames(svgs_meta), 2), &quot;TMEM200B&quot;, &quot;CLEC10A&quot;),
  ncol = 2,
  image.alpha = 0.3,
  pt.size.factor = 5
)</code></pre>
<p><img src="figure/visium_04.Rmd/unnamed-chunk-17-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>SVGs are good for finding genes which expression varies across
different regions while HVGs are good for finding genes which
distinguishes different cell types or groups regardless of the genes’
spatial context.</p>
<p>We can inspect SVGs against the cell type deconvolution to see if
there are relationship between the different cell types and the
genes:</p>
<pre class="r"><code>SpatialFeaturePlot(
  dat,
  features = c(
    head(rownames(svgs_meta), 9), colnames(norm_weights)
  ),
  ncol = 5,
  image.alpha = 0.3,
  pt.size.factor = 5
)</code></pre>
<p><img src="figure/visium_04.Rmd/unnamed-chunk-18-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This plot indicates that not all tumour regions may be equal. The top
and right regions seem to express CXCL2, CXCL3, GPX2, and SPINK1. The
one in the left middle, are lacking CXCL2, CXCL3, and SPINK1. Notably,
it is surrounded by several immune cells, which altogether hint at the
possibility of the two tumour regions may be undergoing different
biological processes.</p>
<p>We can also grab top few genes and run pathway analysis to see which
biological processes are commonly associated with those genes. There are
many softwares to do this, <a
href="https://metascape.org/">Metascape</a> is one of them.</p>
<p><img src="assets/visium_metascape_go.png" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-visium_metascape_go">
Past versions of visium_metascape_go.png
</button>
</p>
<div id="fig-visium_metascape_go" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/051b760847b01faf5c93970eed5dbc49d28c741c/docs/assets/visium_metascape_go.png" target="_blank">051b760</a>
</td>
<td>
Givanna Putri
</td>
<td>
2024-11-12
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="visual-inferrence-of-cell-cell-communications"
class="section level3">
<h3>Visual inferrence of cell-cell communications</h3>
<p>Now that we have somewhat determine what cell types reside in which
region, we can use SpatialFeaturePlot to
<strong><em>shallowly</em></strong> infer cell-cell communication by
visual inspection.</p>
<p>For example, SIRPA and CD47 is a common receptor ligand pair between
macrophages and tumour cells. CD47 ligand is commonly sent out by tumour
cells to tell macrophages to not eat them. This is quite a well known
immune mechanism employed by tumour cells.</p>
<pre class="r"><code>SpatialFeaturePlot(
  dat,
  features = c(
    &quot;decon_Tumor&quot;, &quot;decon_Myeloid&quot;,
    &quot;CD68&quot;, &quot;CD14&quot;, &quot;CD163&quot;,
    &quot;SIRPA&quot;, &quot;CD47&quot;
  ),
  ncol = 5,
  pt.size.factor = 5,
  image.alpha = 0.3
)</code></pre>
<p><img src="figure/visium_04.Rmd/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>From the plot, we can infer that there are regions where macrophages
and tumour cells co-exist SIRPA and CD47 are expressed, which shows that
macrophages’ inhibitory receptor is activated, signalling it against
phagocytosis.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Australia/Sydney
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] spacexr_2.2.1      scales_1.3.0       ggplot2_3.5.1      qs_0.27.2         
[5] Seurat_5.1.0       SeuratObject_5.0.2 sp_2.1-4           workflowr_1.7.1   

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3     rstudioapi_0.16.0      jsonlite_1.8.9        
  [4] magrittr_2.0.3         spatstat.utils_3.1-0   farver_2.1.2          
  [7] rmarkdown_2.28         fs_1.6.5               vctrs_0.6.5           
 [10] ROCR_1.0-11            spatstat.explore_3.3-2 htmltools_0.5.8.1     
 [13] sass_0.4.9             sctransform_0.4.1      parallelly_1.38.0     
 [16] KernSmooth_2.23-24     bslib_0.8.0            htmlwidgets_1.6.4     
 [19] ica_1.0-3              plyr_1.8.9             plotly_4.10.4         
 [22] zoo_1.8-12             cachem_1.1.0           whisker_0.4.1         
 [25] igraph_2.0.3           iterators_1.0.14       mime_0.12             
 [28] lifecycle_1.0.4        pkgconfig_2.0.3        Matrix_1.7-0          
 [31] R6_2.5.1               fastmap_1.2.0          fitdistrplus_1.2-1    
 [34] future_1.34.0          shiny_1.9.1            digest_0.6.37         
 [37] colorspace_2.1-1       patchwork_1.3.0        ps_1.8.0              
 [40] rprojroot_2.0.4        tensor_1.5             RSpectra_0.16-2       
 [43] irlba_2.3.5.1          labeling_0.4.3         RcppZiggurat_0.1.6    
 [46] progressr_0.14.0       fansi_1.0.6            spatstat.sparse_3.1-0 
 [49] httr_1.4.7             polyclip_1.10-7        abind_1.4-8           
 [52] compiler_4.4.1         doParallel_1.0.17      withr_3.0.1           
 [55] fastDummies_1.7.4      highr_0.11             MASS_7.3-61           
 [58] tools_4.4.1            lmtest_0.9-40          httpuv_1.6.15         
 [61] future.apply_1.11.2    Rfast2_0.1.5.2         goftest_1.2-3         
 [64] quadprog_1.5-8         glue_1.7.0             callr_3.7.6           
 [67] nlme_3.1-166           promises_1.3.0         grid_4.4.1            
 [70] Rtsne_0.17             getPass_0.2-4          cluster_2.1.6         
 [73] reshape2_1.4.4         generics_0.1.3         gtable_0.3.5          
 [76] spatstat.data_3.1-2    tidyr_1.3.1            RApiSerialize_0.1.4   
 [79] data.table_1.16.0      stringfish_0.16.0      utf8_1.2.4            
 [82] spatstat.geom_3.3-3    RcppAnnoy_0.0.22       foreach_1.5.2         
 [85] ggrepel_0.9.6          RANN_2.6.2             pillar_1.9.0          
 [88] stringr_1.5.1          spam_2.10-0            RcppHNSW_0.6.0        
 [91] later_1.3.2            splines_4.4.1          dplyr_1.1.4           
 [94] lattice_0.22-6         survival_3.7-0         deldir_2.0-4          
 [97] tidyselect_1.2.1       Rnanoflann_0.0.3       miniUI_0.1.1.1        
[100] pbapply_1.7-2          knitr_1.48             git2r_0.33.0          
[103] gridExtra_2.3          scattermore_1.2        xfun_0.47             
[106] matrixStats_1.4.1      stringi_1.8.4          lazyeval_0.2.2        
[109] yaml_2.3.10            evaluate_1.0.0         codetools_0.2-20      
[112] tibble_3.2.1           cli_3.6.3              RcppParallel_5.1.9    
[115] uwot_0.2.2             xtable_1.8-4           reticulate_1.39.0     
[118] munsell_0.5.1          processx_3.8.4         jquerylib_0.1.4       
[121] Rcpp_1.0.13            globals_0.16.3         spatstat.random_3.3-2 
[124] png_0.1-8              Rfast_2.1.0            spatstat.univar_3.0-1 
[127] parallel_4.4.1         dotCall64_1.1-1        listenv_0.9.1         
[130] viridisLite_0.4.2      ggridges_0.5.6         leiden_0.4.3.1        
[133] purrr_1.0.2            rlang_1.1.4            cowplot_1.1.3         </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
