<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2024-11-18" />

<title>10x Xenium Analysis Part 1</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">ASI Spatial Workshop 2024</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Analysis Files
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="setup.html">Setup</a>
    </li>
    <li>
      <a href="stomics_01_intro.html">STOmics intro</a>
    </li>
    <li>
      <a href="stomics_02_basics.html">STOmics basics</a>
    </li>
    <li>
      <a href="stomics_03_analysis.html">STOmics analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ghar1821/asi_spatial_workshop_2024">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">10x Xenium Analysis Part 1</h1>
<h4 class="date">2024-11-18</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-11-20
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>asi_spatial_workshop_2024/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240925code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240925)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240925code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240925)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomghar1821asispatialworkshop2024treee5a7744281c1a6e4d467b8dcb5d0ac085406865etargetblanke5a7744a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/tree/e5a7744281c1a6e4d467b8dcb5d0ac085406865e" target="_blank">e5a7744</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomghar1821asispatialworkshop2024treee5a7744281c1a6e4d467b8dcb5d0ac085406865etargetblanke5a7744a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/tree/e5a7744281c1a6e4d467b8dcb5d0ac085406865e" target="_blank">e5a7744</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    stomics/
    Ignored:    visium/

Untracked files:
    Untracked:  RCTD_Plots/
    Untracked:  xenium/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/Xenium_Part1.Rmd</code>) and HTML
(<code>docs/Xenium_Part1.html</code>) files. If you’ve configured a
remote Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/e5a7744281c1a6e4d467b8dcb5d0ac085406865e/analysis/Xenium_Part1.Rmd" target="_blank">e5a7744</a>
</td>
<td>
Cathal King
</td>
<td>
2024-11-20
</td>
<td>
updating Xenium analysis files
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghar1821/asi_spatial_workshop_2024/blob/8df3368293913e1169ef5c1bc9992309ac92983e/analysis/Xenium_Part1.Rmd" target="_blank">8df3368</a>
</td>
<td>
Cathal King
</td>
<td>
2024-11-18
</td>
<td>
Xenium analysis part 1
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>This part of the analysis will read in raw data and construct a data
object containing 10x Xenium data. It will also explore the Xenium
outputs explore the data object</p>
<div id="load-libraries" class="section level1">
<h1>Load libraries</h1>
</div>
<div id="read-in-xenium-data" class="section level1">
<h1>Read in Xenium data</h1>
<ul>
<li>Locate the Xenium output bundle on your computer and set it to
<code>dir</code></li>
<li>Might have to unzip first</li>
<li>Check the Xenium output bundle in a tree-like format</li>
</ul>
<pre class="r"><code># Xenium_V1_Human_Colon_Cancer_P5_CRC_Add_on_FFPE_outs.zip

# read data into an SE object
# dir should point to the Xenium outputs
# this output folder contains data from patient 5
dir &lt;- &quot;xenium/raw/&quot;

# examine files in Xenium output bundle
fs::dir_tree(dir)</code></pre>
<pre><code>xenium/raw/
├── Xenium_V1_Human_Colon_Cancer_P5_CRC_Add_on_FFPE_outs.zip
├── analysis.tar.gz
├── analysis.zarr.zip
├── analysis_summary.html
├── aux_outputs.tar.gz
├── cell_boundaries.csv.gz
├── cell_boundaries.parquet
├── cell_boundaries_sf.parquet
├── cell_feature_matrix.h5
├── cell_feature_matrix.tar.gz
├── cell_feature_matrix.zarr.zip
├── cells.csv
├── cells.parquet
├── cells.zarr.zip
├── experiment.xenium
├── gene_panel.json
├── metrics_summary.csv
├── morphology.ome.tif
├── morphology_focus
│   ├── morphology_focus_0000.ome.tif
│   ├── morphology_focus_0001.ome.tif
│   ├── morphology_focus_0002.ome.tif
│   └── morphology_focus_0003.ome.tif
├── nucleus_boundaries.csv.gz
├── nucleus_boundaries.parquet
├── nucleus_boundaries_sf.parquet
├── transcripts.csv.gz
├── transcripts.parquet
└── transcripts.zarr.zip</code></pre>
<pre class="r"><code># the main files needed for construction of SE are:
#cell_feature_matrix.h5
#cells.parquet
#experiment.xenium

# create data object and call it se
se &lt;- SpatialFeatureExperiment::readXenium(data_dir = dir, 
                                                    sample_id = &quot;p5&quot;,
                                           row.names = &quot;symbol&quot;)</code></pre>
<pre><code>&gt;&gt;&gt; Preprocessed sf segmentations found
&gt;&gt;&gt; Reading cell and nucleus segmentations</code></pre>
<pre><code>&gt;&gt;&gt; Reading cell metadata -&gt; `cells.parquet`</code></pre>
<pre><code>&gt;&gt;&gt; Reading h5 gene count matrix</code></pre>
<pre><code>&gt;&gt;&gt; filtering cellSeg geometries to match 275998 cells with counts &gt; 0</code></pre>
<pre><code>&gt;&gt;&gt; filtering nucSeg geometries to match 256552 cells with counts &gt; 0</code></pre>
<pre class="r"><code># Read in data with readRDS instead
# se &lt;- readRDS(~/data/se_xenium.rds)</code></pre>
</div>
<div id="exploring-the-xenium-se-object" class="section level1">
<h1>Exploring the Xenium SE object</h1>
<ul>
<li>Getting, setting, and plotting fields of SE objects</li>
<li>Common operations on SE objects</li>
<li>getters, setters, and plotting functions</li>
<li>Find the gene count matrix</li>
</ul>
<pre class="r"><code># summary
se</code></pre>
<pre><code>class: SpatialFeatureExperiment 
dim: 541 275998 
metadata(1): Samples
assays(1): counts
rownames(541): ABCC8 ACP5 ... UnassignedCodeword_0330
  UnassignedCodeword_0338
rowData names(3): ID Symbol Type
colnames(275998): aaaadchh-1 aaaaglla-1 ... oinihnka-1 oiniiibd-1
colData names(9): transcript_counts control_probe_counts ...
  nucleus_area sample_id
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
spatialCoords names(2) : x_centroid y_centroid
imgData names(4): sample_id image_id data scaleFactor

unit: micron
Geometries:
colGeometries: centroids (POINT), cellSeg (MULTIPOLYGON), nucSeg (MULTIPOLYGON) 

Graphs:
p5: </code></pre>
<pre class="r"><code># dimensions
dim(se)</code></pre>
<pre><code>[1]    541 275998</code></pre>
<pre class="r"><code># where are the counts?
# use the &#39;counts&#39; getter function 
m &lt;- counts(se)
head(m)</code></pre>
<pre><code>&lt;6 x 275998&gt; sparse DelayedMatrix object of type &quot;integer&quot;:
        aaaadchh-1 aaaaglla-1 aaaanjng-1 ... oiniecfb-1 oinihnka-1 oiniiibd-1
ABCC8            0          0          0   .          0          0          0
ACP5             1          0          0   .          0          0          0
ACTA2            1          0          3   .          0          1          0
ADH1C            0          0          0   .          0          0          0
ADRA2A           0          0          0   .          0          0          0
AFAP1L2          0          0          0   .          0          0          0</code></pre>
</div>
<div id="explore-metadata-within-the-xenium-se-data-object"
class="section level1">
<h1>Explore metadata within the Xenium SE data object</h1>
<ul>
<li>The gene count matrix has metadata about the cells and genes.</li>
<li>This metadata is stored within various slots of the SE object.</li>
<li>Use the <code>colData</code> getter function to get cell metadata
and <code>rowData</code> to get gene metadata.</li>
</ul>
<pre class="r"><code>colData(se) # cell metadata</code></pre>
<pre><code>DataFrame with 275998 rows and 9 columns
           transcript_counts control_probe_counts control_codeword_counts
                   &lt;integer&gt;            &lt;integer&gt;               &lt;integer&gt;
aaaadchh-1                56                    0                       0
aaaaglla-1                51                    0                       0
aaaanjng-1                45                    0                       0
aaabakie-1                37                    0                       0
aaabalkl-1               112                    0                       0
...                      ...                  ...                     ...
oiniccld-1                20                    0                       0
oinicmgd-1                 8                    0                       0
oiniecfb-1                44                    0                       0
oinihnka-1                 6                    0                       0
oiniiibd-1                56                    0                       0
           unassigned_codeword_counts deprecated_codeword_counts total_counts
                            &lt;integer&gt;                  &lt;integer&gt;    &lt;integer&gt;
aaaadchh-1                          0                          0           56
aaaaglla-1                          0                          0           51
aaaanjng-1                          0                          0           45
aaabakie-1                          0                          0           37
aaabalkl-1                          0                          0          112
...                               ...                        ...          ...
oiniccld-1                          0                          0           20
oinicmgd-1                          0                          0            8
oiniecfb-1                          0                          0           44
oinihnka-1                          0                          0            6
oiniiibd-1                          0                          0           56
           cell_area nucleus_area   sample_id
           &lt;numeric&gt;    &lt;numeric&gt; &lt;character&gt;
aaaadchh-1   49.9880      31.3384          p5
aaaaglla-1   44.5241      25.8294          p5
aaaanjng-1   54.0069      33.6866          p5
aaabakie-1   49.4461      31.3384          p5
aaabalkl-1   74.2369      35.7638          p5
...              ...          ...         ...
oiniccld-1  20.81703     16.57234          p5
oinicmgd-1  18.73984     18.73984          p5
oiniecfb-1  42.58235     42.58235          p5
oinihnka-1   7.54109      7.54109          p5
oiniiibd-1 177.96079     21.17828          p5</code></pre>
<pre class="r"><code>rowData(se) # gene metadata</code></pre>
<pre><code>DataFrame with 541 rows and 3 columns
                                            ID                 Symbol
                                   &lt;character&gt;            &lt;character&gt;
ABCC8                          ENSG00000006071                  ABCC8
ACP5                           ENSG00000102575                   ACP5
ACTA2                          ENSG00000107796                  ACTA2
ADH1C                          ENSG00000248144                  ADH1C
ADRA2A                         ENSG00000150594                 ADRA2A
...                                        ...                    ...
UnassignedCodeword_0289 UnassignedCodeword_0.. UnassignedCodeword_0..
UnassignedCodeword_0292 UnassignedCodeword_0.. UnassignedCodeword_0..
UnassignedCodeword_0298 UnassignedCodeword_0.. UnassignedCodeword_0..
UnassignedCodeword_0330 UnassignedCodeword_0.. UnassignedCodeword_0..
UnassignedCodeword_0338 UnassignedCodeword_0.. UnassignedCodeword_0..
                                       Type
                                &lt;character&gt;
ABCC8                       Gene Expression
ACP5                        Gene Expression
ACTA2                       Gene Expression
ADH1C                       Gene Expression
ADRA2A                      Gene Expression
...                                     ...
UnassignedCodeword_0289 Unassigned Codeword
UnassignedCodeword_0292 Unassigned Codeword
UnassignedCodeword_0298 Unassigned Codeword
UnassignedCodeword_0330 Unassigned Codeword
UnassignedCodeword_0338 Unassigned Codeword</code></pre>
<pre class="r"><code>colGeometries(se)</code></pre>
<pre><code>List of length 3
names(3): centroids cellSeg nucSeg</code></pre>
<pre class="r"><code># use the $ operator to access slops within the object
# se$

# save file as SE object
# re-use this later as needed
#saveRDS(object = se, file = &quot;/DIR/se_p5.rds&quot;)</code></pre>
</div>
<div id="plot-xenium-cells" class="section level1">
<h1>Plot Xenium cells</h1>
<ul>
<li>Extract cell segmentation results to a data frame</li>
</ul>
<pre class="r"><code>(cells &lt;- colGeometry(se, &quot;cellSeg&quot;))</code></pre>
<pre><code>Simple feature collection with 275998 features and 1 field
Geometry type: MULTIPOLYGON
Dimension:     XY
Bounding box:  xmin: 0 ymin: -6554.138 xmax: 6677.812 ymax: -51.6375
CRS:           NA
First 10 features:
           label_id                       geometry
aaaadchh-1        1 MULTIPOLYGON (((492.3625 -2...
aaaaglla-1        2 MULTIPOLYGON (((500.8625 -2...
aaaanjng-1        3 MULTIPOLYGON (((490.45 -231...
aaabakie-1        4 MULTIPOLYGON (((497.8875 -2...
aaabalkl-1        5 MULTIPOLYGON (((485.775 -23...
aaabbekk-1        6 MULTIPOLYGON (((514.8875 -2...
aaabbkil-1        7 MULTIPOLYGON (((528.0625 -2...
aaabgamd-1        8 MULTIPOLYGON (((531.8875 -2...
aaaboeji-1        9 MULTIPOLYGON (((517.0125 -2...
aaacgkkm-1       10 MULTIPOLYGON (((484.5 -2358...</code></pre>
<ul>
<li>Plot cell geometries using the <code>colGeometry</code>
function</li>
</ul>
<pre class="r"><code>ggplot(cells) + geom_sf() + ggtitle(&quot;Xenium cells - p5&quot;)</code></pre>
<p><img src="figure/Xenium_Part1.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="plot-xenium-cells-custom" class="section level1">
<h1>Plot Xenium cells custom</h1>
<ul>
<li>Plot the cell segmentation results. Also plot with the dark
theme.</li>
<li>Use the <code>plotGeometry</code> function.</li>
</ul>
<pre class="r"><code># &quot;type&quot; or &quot;colGeometryName&quot; depending on version of the package
plotGeometry(se, colGeometryName = &quot;cellSeg&quot;, show_axes = TRUE) +
    plotGeometry(se, colGeometryName = &quot;cellSeg&quot;, show_axes = TRUE, dark = TRUE)</code></pre>
<p><img src="figure/Xenium_Part1.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="save-checkpoint" class="section level1">
<h1>Save checkpoint</h1>
<pre class="r"><code>saveRDS(object = se, file = &quot;xenium/data/se_xenium.rds&quot;)</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.0 (2024-04-24)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.7.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Australia/Adelaide
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] plotly_4.10.4                  scales_1.3.0                  
 [3] terra_1.7-78                   EBImage_4.48.0                
 [5] sf_1.0-19                      bluster_1.16.0                
 [7] fs_1.6.5                       scran_1.34.0                  
 [9] BiocParallel_1.40.0            scater_1.34.0                 
[11] ggplot2_3.5.1                  scuttle_1.16.0                
[13] SingleCellExperiment_1.28.1    SummarizedExperiment_1.36.0   
[15] Biobase_2.66.0                 GenomicRanges_1.58.0          
[17] GenomeInfoDb_1.42.0            IRanges_2.40.0                
[19] S4Vectors_0.44.0               BiocGenerics_0.52.0           
[21] MatrixGenerics_1.18.0          matrixStats_1.4.1             
[23] RBioFormats_1.6.0              Voyager_1.8.1                 
[25] SpatialFeatureExperiment_1.8.2 workflowr_1.7.1               

loaded via a namespace (and not attached):
  [1] later_1.3.2               bitops_1.0-9             
  [3] tibble_3.2.1              R.oo_1.27.0              
  [5] lifecycle_1.0.4           edgeR_4.4.0              
  [7] rprojroot_2.0.4           processx_3.8.4           
  [9] lattice_0.22-6            magrittr_2.0.3           
 [11] limma_3.62.1              sass_0.4.9               
 [13] rmarkdown_2.29            jquerylib_0.1.4          
 [15] yaml_2.3.10               metapod_1.14.0           
 [17] httpuv_1.6.15             sp_2.1-4                 
 [19] DBI_1.2.3                 abind_1.4-8              
 [21] zlibbioc_1.52.0           purrr_1.0.2              
 [23] R.utils_2.12.3            RCurl_1.98-1.16          
 [25] git2r_0.35.0              GenomeInfoDbData_1.2.13  
 [27] ggrepel_0.9.6             irlba_2.3.5.1            
 [29] units_0.8-5               RSpectra_0.16-2          
 [31] dqrng_0.4.1               DelayedMatrixStats_1.28.0
 [33] codetools_0.2-20          DropletUtils_1.26.0      
 [35] DelayedArray_0.32.0       xml2_1.3.6               
 [37] tidyselect_1.2.1          UCSC.utils_1.2.0         
 [39] memuse_4.2-3              farver_2.1.2             
 [41] ScaledMatrix_1.14.0       viridis_0.6.5            
 [43] jsonlite_1.8.9            BiocNeighbors_2.0.0      
 [45] e1071_1.7-16              tools_4.4.0              
 [47] ggnewscale_0.5.0          sfarrow_0.4.1            
 [49] Rcpp_1.0.13-1             glue_1.8.0               
 [51] gridExtra_2.3             SparseArray_1.6.0        
 [53] xfun_0.49                 dplyr_1.1.4              
 [55] HDF5Array_1.34.0          withr_3.0.2              
 [57] fastmap_1.2.0             boot_1.3-31              
 [59] rhdf5filters_1.18.0       fansi_1.0.6              
 [61] spData_2.3.3              callr_3.7.6              
 [63] digest_0.6.37             rsvd_1.0.5               
 [65] R6_2.5.1                  colorspace_2.1-1         
 [67] wk_0.9.4                  jpeg_0.1-10              
 [69] R.methodsS3_1.8.2         utf8_1.2.4               
 [71] tidyr_1.3.1               generics_0.1.3           
 [73] data.table_1.16.2         class_7.3-22             
 [75] httr_1.4.7                htmlwidgets_1.6.4        
 [77] S4Arrays_1.6.0            whisker_0.4.1            
 [79] spdep_1.3-6               pkgconfig_2.0.3          
 [81] scico_1.5.0               rJava_1.0-11             
 [83] gtable_0.3.6              XVector_0.46.0           
 [85] htmltools_0.5.8.1         fftwtools_0.9-11         
 [87] png_0.1-8                 SpatialExperiment_1.16.0 
 [89] knitr_1.49                rstudioapi_0.17.1        
 [91] rjson_0.2.23              proxy_0.4-27             
 [93] cachem_1.1.0              rhdf5_2.50.0             
 [95] stringr_1.5.1             KernSmooth_2.23-24       
 [97] parallel_4.4.0            vipor_0.4.7              
 [99] arrow_17.0.0.1            s2_1.1.7                 
[101] pillar_1.9.0              grid_4.4.0               
[103] vctrs_0.6.5               promises_1.3.0           
[105] BiocSingular_1.22.0       beachmat_2.22.0          
[107] sfheaders_0.4.4           cluster_2.1.6            
[109] beeswarm_0.4.0            evaluate_1.0.1           
[111] zeallot_0.1.0             magick_2.8.5             
[113] cli_3.6.3                 locfit_1.5-9.10          
[115] compiler_4.4.0            rlang_1.1.4              
[117] crayon_1.5.3              classInt_0.4-10          
[119] ps_1.8.1                  getPass_0.2-4            
[121] ggbeeswarm_0.7.2          stringi_1.8.4            
[123] viridisLite_0.4.2         deldir_2.0-4             
[125] assertthat_0.2.1          munsell_0.5.1            
[127] lazyeval_0.2.2            tiff_0.1-12              
[129] Matrix_1.7-1              patchwork_1.3.0          
[131] bit64_4.5.2               sparseMatrixStats_1.18.0 
[133] Rhdf5lib_1.28.0           statmod_1.5.0            
[135] igraph_2.1.1              bslib_0.8.0              
[137] bit_4.5.0                </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
