---
title: "Setup"
author: "Givanna Putri, Thomas O'Neil, John Salamon, Cathal King"
date: "2024-11-04"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

Please ensure you follow the instructions below **prior** to attending the workshop.

You can either create a new R script and copy paste the content of the code blocks
or download [this Rmd file](https://github.com/ghar1821/asi_spatial_workshop_2024/blob/master/analysis/setup.Rmd)
and run each code block step by step.

***Do not just blindly run the code! Make sure you read the description above each
of the code block to understand what the code is doing before running them!***
Sourcing the entire Rmd file will not run anything as every code block has been 
deliberately annotated such that they won't execute when the whole file is sourced.


If you have any problems, please reach out to either:

- [Felix Marsh-Wakefield](mailto:felix.marsh-wakefield@sydney.edu.au)
- [Thomas O'Neil](mailto:thomas.oneil@sydney.edu.au)
- [Givanna Putri](mailto:putri.g@wehi.edu.au)
- [Cathal King](mailto:cathal.king@sahmri.com)

## Packages

Please install the packages that we will be using in the workshop by running
the code block below.

```{r, eval=FALSE}

cran_packages <- c(
  "BiocManager", "Seurat", "ggplot2", "scales", "patchwork", "qs",
  "hdf5r", "viridis", "devtools", "cli", "fs", "sf", "terra", "scales",
  "plotly", "hexbin", "raster"
)

install.packages(cran_packages)

bioc_packages <- c(
  "glmGamPoi", "scater", "scran", "SpatialExperiment", "ggspavis", "scran",
  "nnSVG", "SpatialFeatureExperiment", "Voyager", "RBioFormats", "BiocParallel",
  "bluster", "EBImage"
)

BiocManager::install(bioc_packages)

# RCTD only available via devtools install
devtools::install_github("dmcable/spacexr", build_vignettes = FALSE)
```

Running this chunk will let you know if the packages have been installed properly. 

```{r eval=FALSE}
checkSetup <- function() {
  library(cli)
  cat("\n--------------------------------------\n")
  cat(style_bold(col_magenta("\n***Installing General Packages***\n\n")))
  not <- c(); not2 <- c()
  packages1 <- c(cran_packages, bioc_packages)#, "Test")
  for (i in 1:length(packages1)){
    if(requireNamespace(packages1[i], quietly = TRUE)==F) {
      cat(paste(style_bold(col_red(packages1[i])), "has not been installed\n"))
      not <- c(not,i)
    } else {
      suppressWarnings(suppressMessages(library(as.character(packages1[i]), character.only = TRUE)))
      cat(col_yellow(packages1[i]), "is loaded!\n")
    }
  }
  cat("\n--------------------------------------\n")
  if (length(not) > 0){
    cat(style_bold(bg_red("\n  **IMPORTANT**  ")),
        style_bold(col_yellow("\n\nYou need to install: \n")),
        paste(paste(c(packages1[not]), collapse=", ")),
        "\n\n--------------------------------------",
        "\n\n Use:\n - install.packages(),\n - BiocManager::install() or, \n - use Google to find installation instructions.\n\n", style_bold(col_green("Then run this function again!\n\n")))
  } else {
    cat("",col_green(style_bold("\n All packages are loaded!\n\n Happy Coding! :)\n\n")))
  }
}
checkSetup()
```

***Note:***

You may encounter error running the code block above when it tries to load the *RBioFormats* package.
This is because *RBioFormats* loads [rJava](https://github.com/s-u/rJava) package which relies on
having Java Run Time (JRE) installed in your system.

**For Mac users,** if you have [homebrew](https://brew.sh/) installed, you can install JRE
through OpenJDK using homebrew by running `brew install openjdk` on terminal.
Afterwards, you will need to set the `JAVA_HOME` environment value in rstudio to 
the location where homebrew installed openjdk.
For example, if OpenJDK is installed under `/opt/homebrew/opt/openjdk/libexec/openjdk.jdk/Contents/Home`,
and set `JAVA_HOME` to point to that directory
using the following code:

```{bash eval=FALSE}
Sys.setenv(JAVA_HOME="/opt/homebrew/opt/openjdk/libexec/openjdk.jdk/Contents/Home")
```

If you are not sure how to get JRE installed into your system to load *RBioFormats*
package, please reach out to your IT support for assistance.

## Folder structure

The code block below will create a bunch of folders in the location where
this script is stored.

```{r, eval=FALSE}
# This sets the working directory for all subsequent code chunks to be run
# to the location of this file.
knitr::opts_knit$set(root.dir = dirname(rstudioapi::getActiveDocumentContext()$path))

if(!dir.exists("visium")){dir.create("visium")}
if(!dir.exists("visium/raw")){dir.create("visium/raw")}
if(!dir.exists("visium/data")){dir.create("visium/data")}

if(!dir.exists("xenium")){dir.create("xenium")}
if(!dir.exists("xenium/raw")){dir.create("xenium/raw")}
if(!dir.exists("xenium/data")){dir.create("xenium/data")}

if(!dir.exists("stomics")){dir.create("stomics")}
if(!dir.exists("stomics/raw")){dir.create("stomics/raw")}
if(!dir.exists("stomics/data")){dir.create("stomics/data")}

```

## Data Download{.tabset}

Please source the code block below to load a function to check the integrity of the downloaded files.

```{r eval=FALSE}
check_md5 <- function(files_expected_md5sum) {
    # files_expected_md5sum must be a named vector where the name is the path to the file
    # and the value is the expected md5sum
    library(cli)
    library(tools)
    
    error <- combine_ansi_styles("red", "bold")
    
    need_redownload <- FALSE
    for (filename in names(files_expected_md5sum)) {
        actual_md5sum <- md5sum(filename)
        expected_md5sum <- files_expected_md5sum[filename]
        if (actual_md5sum != expected_md5sum) {
            cat(error(paste(
                filename, "is corrupted. Actual md5sum:", actual_md5sum, "!= Expected md5sum:", expected_md5sum, "\n"
                )
            ))
            
            need_redownload <- TRUE
        }
    }
    if (need_redownload) {
        cat(error("Some files are corrupted. Please redownload"))
    }
}
```

***Note***: if you encounter a timeout error (after 60 seconds) while downloading the files,
you can increase the timeout to say 6,000 seconds to give R more time
to download the files by running `options(timeout = max(6000, getOption("timeout")))`

### Visium

The code block below will download the count matrix and spatial information
required to load the Visium data in.

```{r eval=FALSE}
download.file("https://cf.10xgenomics.com/samples/spatial-exp/3.0.0/Visium_V2_Human_Colon_Cancer_P2/Visium_V2_Human_Colon_Cancer_P2_filtered_feature_bc_matrix.h5", 
              destfile = "visium/raw/Visium_V2_Human_Colon_Cancer_P2_filtered_feature_bc_matrix.h5", 
              method = "curl")


download.file("https://cf.10xgenomics.com/samples/spatial-exp/3.0.0/Visium_V2_Human_Colon_Cancer_P2/Visium_V2_Human_Colon_Cancer_P2_spatial.tar.gz", 
              destfile = "visium/raw/spatial.tar.gz", 
              method = "curl")
```

The code below will check the integrity of the downloaded files.
If you get any red warning, please re-download the files that were found to be corrupted using the code block above.

```{r eval=FALSE}

# check md5sum of the downloaded files.
md5sum_str <- c(
    "visium/raw/Visium_V2_Human_Colon_Cancer_P2_filtered_feature_bc_matrix.h5" = "0a9733cf0dfd3dfaa3b6c1c5c24bcbed",
    "visium/raw/spatial.tar.gz" = "a32edc825ab7089fca7848fc80cd5113"
)
check_md5(md5sum_str)

```

If there are no warning from the code block above, proceed to untar the `spatial.tar.gz` file 
using the code block below.

```{r eval=FALSE}
untar("visium/raw/spatial.tar.gz", exdir ="visium/raw")
```


#### Optional loupe browser file

The following file is optional and is only required if you would like to try out
the demo we will be showing on how to use loupe browser to make some QC step simpler.

```{r eval=FALSE}
download.file(
    "https://cf.10xgenomics.com/samples/spatial-exp/3.0.0/Visium_V2_Human_Colon_Cancer_P2/Visium_V2_Human_Colon_Cancer_P2_cloupe.cloupe",
    destfile = "visium/raw/Visium_V2_Human_Colon_Cancer_P2_cloupe.cloupe",
    method = "curl"
)
md5sum_str <- c(
    "visium/raw/Visium_V2_Human_Colon_Cancer_P2_cloupe.cloupe" = "c5726e203c21a2fc3432831ea7c74252"
)
check_md5(md5sum_str)
```

Install Loupe Browser at this link:
https://www.10xgenomics.com/support/software/loupe-browser/latest

### STOmics

```{r eval=FALSE}
# LISTA paper, processed tissue slice (222MB)
download.file(
    "https://filesender.aarnet.edu.au/download.php?token=ec632836-0700-4921-9699-668bfe9657d2&files_ids=23608389",
    destfile = "stomics/raw/DY1_D0_stereo-seq.h5ad",
    method = 'curl',
    extra = '-L'
)

# LISTA single cell dataset (540MB)
download.file(
    "https://filesender.aarnet.edu.au/download.php?token=ec632836-0700-4921-9699-668bfe9657d2&files_ids=23608386",
    destfile = "stomics/raw/Homeostasis_hepatic_cell_scRNAseq.h5ad",
    method = 'curl',
    extra = '-L'
)
```

```{r eval=FALSE}
# check md5sum of the downloaded files.
md5sum_str <- c(
    "stomics/raw/DY1_D0_stereo-seq.h5ad" = "f63df815d75548034cb659bc3491f51a",
    "stomics/raw/Homeostasis_hepatic_cell_scRNAseq.h5ad" = "b1a7fd3ffa58e0420083fa8651cb7fec"
)
check_md5(md5sum_str)

```


#### OPTIONAL raw data download from SAW

File size: 4.6GB

```{r eval=FALSE}
# Raw data from SAW (4.6GB, optional download!)
download.file(
    "https://filesender.aarnet.edu.au/download.php?token=32ffcb63-24df-47e2-bdce-7341fe86082a&files_ids=23595450",
    destfile = "stomics/raw/visualization.tar.gz",
    method = "curl",
    extra = '-L'
)

# check md5sum of the downloaded files.
md5sum_str <- c(
    "stomics/raw/visualization.tar.gz" = "7d3daa224de7ca1244711531270546d3"
)
check_md5(md5sum_str)
```

If there are no warning from the code block above, proceed to untar the `visualization.tar.gz` file 
using the code block below.

```{r eval=FALSE}
untar("stomics/raw/visualization.tar.gz", exdir ="stomics/raw")
```

### Xenium

Install Xenium Explorer v3.1.1 at this link:
https://www.10xgenomics.com/support/software/xenium-explorer/downloads

Please also install the raw xenium data by using the below code. 
The file size is large so it might take a few minutes to complete. 

```{r eval=FALSE}
# Sample P5 CRC
# This file is ~12GB so will take a few minutes to download.
download.file(
  url = "https://cf.10xgenomics.com/samples/xenium/2.0.0/Xenium_V1_Human_Colon_Cancer_P5_CRC_Add_on_FFPE/Xenium_V1_Human_Colon_Cancer_P5_CRC_Add_on_FFPE_outs.zip", 
  destfile = "xenium/raw/Xenium_V1_Human_Colon_Cancer_P5_CRC_Add_on_FFPE_outs.zip", 
  method = "curl"
)

# check md5sum of the downloaded files.
md5sum_str <- c(
    "xenium/raw/Xenium_V1_Human_Colon_Cancer_P5_CRC_Add_on_FFPE_outs.zip" = "1069cb6b67a596e0011a4b3f5d3b7c08"
)
check_md5(md5sum_str)

unzip(zipfile = "xenium/raw/Xenium_V1_Human_Colon_Cancer_P5_CRC_Add_on_FFPE_outs.zip")

```

Make sure there are no warning from the code block above.
If there are, please redownload the file by re-running `download.file` command.

