---
title: "Visium FFPE Prostate Tissue Analysis"
author: "Givanna Putri"
date: "2024-10-03"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Introduction

The idea is to use the insight from [this paper](https://www.nature.com/articles/s41467-018-04724-5), and see whether we can extract some gene expressions that are interesting, and perhaps also uncover some immune cells that are critical.

The prostate cancer data has gleason score of 7.

### TIFF file

There is a fluorescence image (IF) comprising of 3 channels: 

* Channel 1: Iba1/AIF-1
* Channel 2: Vimentin
* Channel 3: DAPI

The TIFF file contains all the channels, separated into 3 pages.

> **Helpful Hints:**
>
> For this dataset, the fluorescence image is a grayscale multi-page TIFF.
> Thus, when running spaceranger, we need to specify the `--darkimage` parameter and pass the name of the TIFF file in.
>
> Additionally, we also need to load the output of fiducial alignment process done separately using the Loupe Browser software from 10x.
> More information at the bottom of this section.

Post processing by spaceranger, there will be a folder called `spatial`.
Within the folder, there is a file called `tissue_hires_image.png`.
This file was generated by spaceranger based on the fluorescence image file (which is a grayscale image for this dataset).
It collapses the 3 channels into 1 image and assign different colour to different channel.

```{r echo=FALSE}
knitr::include_graphics("assets/visium_ffpe_prostate_tissue_hires_image.png", error = FALSE)
```

However it is not clear which colour belong to which channel.
To work this out, we can import the TIFF file into ImageJ using the Bio-Formats plugin and 
visually match the colours in the image against the brightest part of the TIFF image's channels.

> **Demo**:
>
> Load the TIFF file using ImageJ's Bio-Formats plugin. 
>
> Once loaded, align the window showing each channel side by side with the high-res image generated by spaceranger. 
> Then look for the bright area on ImageJ and correspond it to the high-res image.
>
>> We can zoom in and out on ImageJ by pressing the + or - keys.
>> We can change channel by using the left and right arrow keys.
>> Clicking on ImageJ will automatically zoom the window in to the area you clicked.

From here, one can determine that:

* Green = Iba1/Channel 1
* Red = Vimentin/Channel 2
* Blue = DAPI


**Notes on fiducial alignment:**
Fiducial alignment is a process of aligning images from different rounds of sequencing or imaging
The purpose of this process is to make sure we correctly align the fluorescence images with the coordinates of the spots.
For visium data containing fluorescence images, the alignment process has to be done manually.
This can be done using [10x Loupe Browser](https://www.10xgenomics.com/support/software/loupe-browser/latest).
The process on how to do fiducial alignment is out of the scope of the workshop.
For more information, you can refer to the guide written by 10x  [here](https://www.10xgenomics.com/support/software/space-ranger/latest/analysis/inputs/image-fiducial-alignment).

> Tips:
More details on what each of the spatial output file means is available [here](https://www.10xgenomics.com/support/software/space-ranger/latest/analysis/outputs/spatial-outputs).

## Setup

```{r}
library(Seurat)
```


## Preprocess

The H5 file and the spatial folder came from SpaceRanger.

The spatial folder is required for seurat to find the following extra information:

* `tissue_lowres_image.png`: the image of the slide
* `scalefactors_json.json`: the scale factors to convert the high-resolution coordinates and the scaled-down coordinates - need to check whether this is the conversion from the high res image to low res.
* `tissue_positions_list.csv`: coordinates for each spot

*NOTE* the function requires `hdf5r` package.

```{r}
dat <- Load10X_Spatial(
  data.dir = 'data/visium_ffpe_human_prostate',
  filename = 'filtered_feature_bc_matrix.h5'
)
```

Quick view of the tissue.

```{r}
SpatialFeaturePlot(
  dat, 
  features = "nCount_Spatial",
  pt.size.factor = 1,
  alpha = c(0.1, 1)
)
```

What function does above is use the 





