---
title: "Visium FFPE Colon Tissue Analysis"
author: "Givanna Putri"
date: "2024-10-30"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Setup

```{r}
library(Seurat)

# for plotting
library(ggplot2)
library(scales)
library(patchwork)

# for exporting object
library(qs)
```


## Loading data

The H5 file and the spatial folder came from SpaceRanger.

The spatial folder is required for seurat to find the following extra information:

* `tissue_lowres_image.png`: the image of the slide
* `scalefactors_json.json`: the scale factors to convert the high-resolution coordinates and the scaled-down coordinates - need to check whether this is the conversion from the high res image to low res.
* `tissue_positions_list.csv`: coordinates for each spot

*NOTE* the function requires `hdf5r` package.

First up, let's import our data into a seurat object.

```{r}
dat <- Load10X_Spatial(
  data.dir = 'data/visium_human_crc',
  filename = 'filtered_feature_bc_matrix.h5'
)

```

## Inspecting Seurat object

***TODO***: add a diagram to show what seurat object looks like, if there is any? 

We can inspect the object to see the number of spots and genes there are in the dataset. 

```{r}
# 4,269 spots - samples in the seurat object description
# 18,085 genes.
dat
```

A Seurat object can contain multiple Assays.
The idea behind assay is so we can store data from different omics technologies in one object.

Within an assay, we can have multiple Layers, each containing one spot by gene matrix.
In subsequent sections, when we perform some transformations to the data, we shall
see how storing the same data but transformed it in different manner 
in different Layers can come in handy.

By default, when we run `Load10x_Spatial`, Seurat will store the raw spot by gene matrix
in `Spatial` assay under `counts` Layer.
You can change the assay name by specifying the `assay` parameter when running the 
`Load10x_Spatial` function.

We can retrieve the spot by gene matrix like so:

```{r}
dat[["Spatial"]]$counts
```

By default, genes (features) are stored as rows and spots (or cells) are stored as columns.

Different assay can have different features, and as of Seurat v5,
different assay and layer can have different spots (or cells). 

### Gene and spot metadata

We can retrieve the name of the genes we have in the data like so:

```{r}
head(rownames(dat), 20)
head(Features(dat), 20)
```

And the unique name assigned to each spot like so:

```{r}
head(colnames(dat), 20)
head(Cells(dat), 20)
```

We can also query the seurat object as if it is a table by using functions
like `ncol`, `nrow`, and `dim`:

```{r}
# how many spots
ncol(dat)

# how many genes
nrow(dat)

# how many genes and spots (rows and cols respectively)
dim(dat)
```

Metadata for the spots can be retrieved like so:

```{r}
head(dat[[]], 20)
```

Here, each row is a spot. 
`nCount_Spatial` shows the total number of transcripts (specifically total number of unique UMIs) for that spot.
`nFeature_Spatial` shows how many unique genes is expressed per spot.
`orig.ident` refers to the "original identity" of the sample. 
By default, the `Load10x_Spatial` will pre-fill this with "SeuratProject".

The metadata in seurat object is actually stored as a data.frame.

```{r}
class(dat[[]])
```

Thus, we can add more information into it using functions that work for data.frame. 
For example, we can add a separate column identifying the patient ID the data originates from:

```{r}
dat$patient_id <- "P2"
head(dat[[]], 20)
```

To access the metadata associated with the genes, we can use similar notation
and also include which assay we want to retrieve the metadata from.

```{r}
dat[['Spatial']][[]]
```

## Visualising the spots

Seurat offers function to visualise the spots captured on the tissue.
One that is very commonly used is `SpatialFeaturePlot` which place the spots
on top of the tissue image and colour the spots by whatever features you have in the Seurat object.

For example, let's visualise the library size of the spots.

```{r}
SpatialFeaturePlot(dat, features = "nCount_Spatial")
```

You can also visualise multiple features at a time, e.g., let's also visualise the number
of genes expressed for each spot as well as the library size:

```{r}
SpatialFeaturePlot(dat, features = c("nCount_Spatial", "nFeature_Spatial"))
```

By default, the plot will focus on the the area of the tissue where spots containing spots.
Disabling this will show entire tissue.

```{r}
SpatialFeaturePlot(dat, features = "nCount_Spatial", crop=FALSE)
```


Often, spots may look too small, like in this case.
We can increase it by increasing the `pt.size.factor` parameter.

```{r}
# store it so we can use the same number later on
pt_size <- 5
SpatialFeaturePlot(dat, features = c("nCount_Spatial", "nFeature_Spatial"), 
                   pt.size.factor = pt_size)
```

To visualise qualitative features, we can use `SpatialDimPlot`.
For example, let's visualise the spot by the patient ID metadata we added before.

```{r}
SpatialDimPlot(dat, group.by = 'patient_id', pt.size.factor = pt_size)
```

We can override the spot colour by overriding the `cols` parameter with
a named vector mapping the discrete category in the data against the colour.

```{r}
# pretend we have three patients in the data
dat[[]]$pretend_patient_id <- c(
  rep("Adam", 3000),
  rep("Jane", 1000),
  rep("Clare", 269)
)

SpatialDimPlot(dat, group.by = 'pretend_patient_id', 
               pt.size.factor = pt_size, 
               cols = c("Adam" = "blue", "Jane" = "orange", "Clare" = "black"))
```

### Tweaking spatial plots

Often, it is helpful to tweak the opacity of the spots or the background image to
help us better visualise the spots and their features.

You can adjust the opacity of the spots by lowering the `alpha` parameter:

```{r}
SpatialFeaturePlot(
  dat, 
  features = "nCount_Spatial",
  pt.size.factor = pt_size,
  alpha = 0.5
)
```

Alpha parameter can take either a single numerical value or a range (a vector with 2 values).
If the latter, it represents the minimum and maximum transparency and default to c(1,1),
i.e. same transparency for both low and high value.
Thus, we can accentuate the spots with high value by lowering the minimum value
we pass as alpha:

```{r}
SpatialFeaturePlot(
  dat, 
  features = "nCount_Spatial",
  pt.size.factor = 5,
  alpha = c(0.1, 1)
)
```

The lower the minimum alpha value, the stronger the high expression will look.

However, it is often the case that the background image is too overpowering, making
it very hard to view the spots.
We can make the background more transparent by lowering the `image.alpha` parameter.

```{r}
SpatialFeaturePlot(
  dat, 
  features = "nCount_Spatial",
  pt.size.factor = pt_size,
  image.alpha = 0.2
)
```

You can also completely forgo the tissue image by setting the `image.alpha` to 0.

```{r}
SpatialFeaturePlot(
  dat, 
  features = "nCount_Spatial",
  pt.size.factor = 4,
  image.alpha = 0
)
```

Both `SpatialFeaturePlot` and `SpatialDimPlot` will return a `ggplot` object,
which we can modify using the functionalities built into `ggplot2` package.

As an example, let's update the plot showing library size so we: 

1. Change the colour scheme to use viridis
2. Rename `nCount_Spatial` to `library_size`
3. Have a black background and axes

Let's update the plot so we have a black background and white letters and axes.

```{r}
plt_dark_bg <- SpatialFeaturePlot(
  dat, 
  features = "nCount_Spatial",
  pt.size.factor = pt_size,
  image.alpha = 0
) + scale_fill_viridis_c(option = 'viridis') + 
  labs(fill='library_size') +
  theme_minimal() + 
  theme(
  legend.text = element_text(colour = "white"),
  legend.title = element_text(colour = "white"),
  legend.background = element_rect(fill = "black"),
  plot.background = element_rect(fill = "black"),
  panel.grid = element_blank(),
  axis.text = element_text(colour = "white"),
  axis.line = element_line(colour = "white")
) 
plt_dark_bg
```

Since the plot is a ggplot object, we can export them using either the
export button in Rstudio's panel, or the ggsave function:

```{r eval=FALSE}
ggsave(filename="output/spatial_plot_darkbg.png", plot=plt_dark_bg)
```

> Good to know:
When there are multiple features drawn using `SpatialFeaturePlot`, 
the resulting object is no longer a ggplot object but rather a "Large patchwork" object, 
an object generated by the [patchwork package](https://patchwork.data-imaginist.com/index.html).
This package is commonly used for combining plots in R.
In actual fact, the large patchwork object itself is a list where each element is a ggplot object.
Thus, to modify the plot, you will just need to override the object like above. 

## QC

We can divide the QC into two sections, QC the spots, and QC the genes.

### QC genes

By default, `Load10X_Spatial` will include all genes in the data.
However, we may have genes that are only expressed in a handful of spots and may well be noise.

```{r}
# Count how many spots each gene is expressed in.
spots_per_gene <- rowSums(dat[["Spatial"]]$counts > 0)

# Plot as histogram
# Have to convert spots_per_gene to data.frame first, the column is spots
ggplot(data.frame(spots = spots_per_gene), aes(x=spots)) + 
  geom_histogram(binwidth = 1, colour = 'blue') +
  geom_vline(xintercept = 100, colour = 'red') +
  theme_minimal() +
  labs(title = "Number of Spots per Gene",
       x = "Number of Spots",
       y = "Number of Genes") +
  scale_x_continuous(breaks = pretty_breaks(10)) +
  scale_y_continuous(breaks = pretty_breaks(10))

```

From the histogram, we can see that we have some genes that are expressed
only by a handful of spots, e.g., there are 12 genes that are not expressed
by any spots at all.

For this dataset, we will set the threshold to 100, i.e. for a gene to be included,
it needs to be expressed by at least 100 spots, which is around 2.3% of total 
spots we have in the data.

```{r}
# names will give you the gene names
genes_to_keep <- names(spots_per_gene[spots_per_gene >= 100])

# Subset the Seurat object
dat_fil <- subset(dat, features = genes_to_keep)
```

> Good to know:
The threshold 100 is not fixed. 
This will vary depending on the dataset you have on hand.
So adjust this as you see fit.

### QC spots

In addition to QC the genes, we should also QC the spots.
The idea is to make sure the spots overlap with tissue sections that are stained with the fluorescence dyes,
and that we do not have spots that lie outside the tissue.

> Tom O'neil demo using ImageJ

To check whether the spots lie on top of the tissues, we can interactively inspect `SpatialDimPlot`
and note the ID of the spots we need to remove.

The default spot ID is a nucleotide sequence that is hard to remember and note.
Thus, to make life easier, let's assign a new numerical ID to the spots.

```{r}
dat_fil[[]]$numeric_spot_id <- paste0("spot_", seq(ncol(dat_fil)))

# backup the original id
dat_fil[[]]$orig_spot_id <- colnames(dat_fil)

# assign the new numeric id as names of the spots
colnames(dat_fil) <- dat_fil[[]]$numeric_spot_id

head(dat_fil[[]], 20)
```

Let's interactively visualise the tissue and the spot:

```{r eval=FALSE}
# unfortunately, the spot size and colour cannot be changed here. 
SpatialDimPlot(
  dat_fil,
  group.by = "patient_id",
  interactive = TRUE
)
```

Note the ID of the spots we want to remove and then remove them.

```{r}
spots_to_remove <- paste0("spot_", c(
  4030, 1810, 1168, 3110, 44, 513, 2278, 3126, 2803, 3828, 2147, 1879, 3828, 2021, 
  4215, 3631, 3225, 3631, 3843, 4035, 1076, 93, 2615, 2041, 749, 3629, 3155, 3683, 
  560, 2953, 2211, 59, 1502, 1789, 2244, 649, 2720, 3803, 3087, 2717, 663, 2882, 
  2934, 3440, 1540, 3830, 471, 1035, 3221, 3257, 3039, 1570, 3560, 758, 40, 1154, 
  2275, 3841, 3475, 3361, 3760, 3592, 1983, 1862, 671, 4234, 622, 2583, 573, 3832, 3959, 1426,
  3746, 959, 1228, 3391, 2332, 2785, 4243, 2315, 1107, 1504, 1133, 3505, 3504, 670
))

# add it as a metadata
dat_fil[[]]$remove_spot <- Cells(dat_fil) %in% spots_to_remove

dat_fil <- subset(x = dat_fil, remove_spot == FALSE)
```

### Check spots' library size and genes expressed

We should also remove spots which express low library size and/or low number of genes expressed.

Few plots we can draw to check this are violin plots, histogram, and plain scatter plots:

```{r}
plt1 <- ggplot(dat_fil[[]], aes(x=nCount_Spatial)) + geom_histogram(bins = 100) +
  theme_classic() +
  scale_y_continuous(breaks = pretty_breaks(10)) +
  scale_x_continuous(breaks = pretty_breaks(10)) 
plt2 <- ggplot(dat_fil[[]], aes(x=nFeature_Spatial)) + geom_histogram(bins = 100) +
  theme_classic() +
  scale_y_continuous(breaks = pretty_breaks(10)) +
  scale_x_continuous(breaks = pretty_breaks(10)) 

plt3 <- VlnPlot(dat_fil, features = c("nCount_Spatial", "nFeature_Spatial"))

plt4 <- FeatureScatter(dat_fil, feature1 = "nCount_Spatial", feature2 = "nFeature_Spatial") +
  scale_y_continuous(breaks = pretty_breaks(10)) +
  scale_x_continuous(breaks = pretty_breaks(10)) +
  theme_bw()

plt1 + plt2 + plt3 + plt4
```

The histograms and violin plots are great for determining the minimum cutoff for library
size and/or number of genes expressed, while the scatter plot is handy to 
check whether the spots with low library size also has low number of genes expressed, which is a clear
sign of potential noise.

For this workshop, we will exclude spots that express less than 3000 genes.
Notably, this threshold will also automatically exclude spots with small library size.

```{r}
dat_fil <- subset(x = dat_fil, nFeature_Spatial >= 3000)
dat_fil
```

> Good to know:
Unlike single cell in which cells with high library size are often removed because
they may well represent doublets or multiplets, for visium data, spots with high library size 
may not necessarily mean bad spots as visium data is not single cell resolution.
Those spots may well just contain an assortment of cell types.
Removing them may well be removing areas with important biology!

### QC spots based on mitochondria genes expression

Just like in single cell, we can also QC the spots based on the percentage of mitochondria genes expressed.
Spots that have excessively high expression of mitochondria genes may contain
low quality or dying cells.

```{r}
# This function will compute, for each spot, percentage of transcripts that mapped to mitochondria genes
dat_fil[['percent_mt']] <- PercentageFeatureSet(dat_fil, pattern = '^MT-')

# Visualize them
plt1 <- VlnPlot(dat_fil, features = "percent_mt")
plt2 <- FeatureScatter(dat_fil, feature1 = 'nCount_Spatial', feature2 = 'percent_mt')

plt1 + plt2
```

We will remove spots that have > 12% mitochondria genes.

```{r}
dat_fil <- subset(dat_fil, subset = percent_mt < 12)
dat_fil
```

After QC, we ended up with 4,215 spots.

### Export data out

There are two options, save as RDS or using QS library.

```{r}
saveRDS(dat_fil, "output/visium_data_qced.rds")

qsave(dat_fil, "output/visium_data_qced.qs")
```

> Good to know:
There are QC tools purposely developed for QCing spatial data like SpotSweeper 
that can QC spots based on the their neighbouring spots.

## Normalisation

Why is this necessary?
The same reason as any other RNA seq based protocol (including single cell).
Differences in library size may well be due to variances in sampling of RNA molecules.
During library preparation, it is not possible to capture all molecules and sequence them.
Thus, the unevenness of library sizes and differences in gene expression across spots, 
may well be due to variance in sampling.
There are also other technical variations that contributed to the differences that need to be accounted for,
e.g., Probe hybridization efficiency and PCR amplification biases.


Normalisation aims to adjust differences introduced due to technical variations.
In single cell, there are many methods that have been developed to do this (see [here](https://www.nature.com/articles/s41592-023-01814-1) for benchmarking paper).
Admittedly, not all of them will be suitable for visium data.
For this tutorial, we will demonstrate classic log normalisation and SCTransform.

### Running SCTransform

Briefly SCTransform fits a regularised negative binomial model to the raw count matrix, 
using sequencing depth as a covariate.

Note, SCT requires glmGamPoi package: `BiocManager::install('glmGamPoi')`

```{r}
# Might need to increase the size limit for global variables so parallel workers
# can access it. 2GB should be sufficient.
options(future.globals.maxSize = 2000 * 1024^2)


# Note, return.only.var.genes can be set to TRUE if we want to get back 
# the expression of only highly variable genes in scale.data
dat_fil <- SCTransform(dat_fil, assay = "Spatial", verbose = TRUE, return.only.var.genes = FALSE)
# Inspect after
dat_fil
```

By default, SCTransform will store the output in a different assay called "SCT".

```{r}
dat_fil[['SCT']]$data
```

> Note on layers:
As previously discussed in the beginning, an assay can contain multiple layers.
After running SCTransfrom, you shall notice a new assay call SCT containing SCTransformed counts.
Specifically, counts layer contains corrected UMI counts,
data contains log1p - log(1 + counts) of the corrected UMI counts,
scale.data contains the counts in the data layer transformed using pearson residuals 
calculated by SCTransform.

### Comparison with log normalisation

Log normalisation is commonly used for single cell data.
It is also widely used for normalising visium data where for each spot and gene,
we divide the expression by the spot's library size and multiply the value by a size factor
(say 10,000).
Then we perform log1p operation where we run logarithmic transformation after adding value of 1.
Why the addition, because log of 0 is infinity.

```{r}
# By default, this will run log CPTT
# The data layer will contain the normalised counts.
# The counts layer will contain the unnormalised count.
dat_fil <- NormalizeData(dat_fil, verbose = FALSE, assay = "Spatial")
dat_fil
```

Upon running this, you will notice an additional layer in the Spatial assay call `data`
which contains the output of log normalisation function.

We can plot the output of SCTransform, log normalisation, and raw count 
for a gene (say COL1A1), side by side to see what they look like.

```{r}
# Important to change assay before plotting to make sure it is the right one.

DefaultAssay(object = dat_fil) <- "SCT"
plt_sct <- SpatialFeaturePlot(
  dat_fil,
  features = "COL1A1",
  image.alpha = 0.5,
  pt.size.factor = 4
) + ggtitle("SCT")


DefaultAssay(object = dat_fil) <- "Spatial"
plt_lognorm <- SpatialFeaturePlot(
  dat_fil,
  features = "COL1A1",
  image.alpha = 0.5,
  pt.size.factor = 4
) + ggtitle("Log Norm")

# Plot raw data for comparison
plt_raw <- SpatialFeaturePlot(
  dat_fil,
  features = "COL1A1",
  image.alpha = 0.5,
  pt.size.factor = 4,
  slot = 'counts'
) + ggtitle("Raw")

# Put all 3 plots side by side for comparison
plt_raw + plt_sct + plt_lognorm

```

> Good to know:
There are pros and cons on choosing either SCT or log normalisation.
From the plot, you can see they are different.
Literatures say SCT mitigates the correlation between library size and normalised counts
which log normalisation failed to do so.
On the other hand, log normalisation is very simple (you can compute it by hand if you like!)
and fast to compute, very well established, intuitive, and compatible with a variety of analysis tools and pipelines.
While SCT may be more robust for data with high variability in sequencing depth,
at the end of the day, choose whichever method that you are comfortable with.

For this tutorial, we will use the normalised counts generated by SCTransform.

```{r}
# Set the default assay to SCT
DefaultAssay(dat_fil) <- "SCT"

# save so we can reload if we need to
qsave(dat_fil, "output/visium_data_qced.qs")
# dat_fil <- qread("output/visium_data_qced.qs")
```

## Identifying cell types in different spots

There are many ways to do this, e.g., clustering or deconvolution.
Clustering which can be controversial for spot based data like visium as a spot
may contain an assortment of cell types.
Deconvolution will estimate how many cell types in a spot using a single cell data
as a reference, which is not always readily available.
Another simple yet efficient method is using module scoring (`AddModuleScore` function).

`AddModuleScore` calculates the average expression levels of predefined set of genes (module) for each spot, 
subtracting the background expression of control genes, which by default, is set to the whole library. 
This results in a "module score" for each module in each spot. 
High score suggests high likelihood of a given cell type (defined by a set of marker genes in a module)
being present in that spot.

Module scoring is a simple yet powerful way of resolving what cell types are likely
to be present in which spot if you know for sure what the marker genes for those cell types are.
A spot that scores high in multiple modules is likely to contain a mixture of those cell types,
while a score scoring high in only a single module, is likely to contain mostly a cell type.

Care need to be taken when interpreting the module score as the specificity of 
a given module is highly dependent on the quality of the gene lists.
Moreover, it does not provide absolute quantification of cell types (exactly how many cell
types are present in a spot).

```{r}
# for finding genes
# Features(dat_fil)[grep("S100B.*", Features(dat_fil))]

# The following marker genes are obtained from combination of
# https://doi.org/10.1038/s43018-024-00807-z
# https://doi.org/10.1038/s41467-024-49916-4
# and perplexity AI.
genes_to_score <- list(
  ec = c("PLVAP", "VWF", "PECAM1", "ERG"),
  malignant = c("EPCAM", "EGFR"),
  glial = c("CLU", "CRYAB", "S100B", "GFAP", "SOX10"),
  fibroblast = c("COL1A1", "COL1A2", "VIM", "FAP", "PDGFRA", "THY1", "S100A4"),
  bcell = c("PTPRC", "MS4A1", "CD79A", "MZB1"),
  tcell = c("PTPRC", "CD3D", "CD3E", "CD3G", "CCL5", "GZMA", "TRBC2", "CD2", "TRAC", "CD7", "KLRB1", "GNLY"),
  myeloid = c("PTPRC", "ITGAX", "ITGAM", "CD14", "LYZ", "LILRA4", "IRF7"),
  hev_ec = c("CXCL10", "CXCL11", "GBP1", "CXCL9", "ISG15", "GBP4", "WARS", "IL32", "CCL2", "CTSS", 
             "IGFBP5", "PECAM1", "VWF"),
  macro = c("CD68", "FCER1G", "CD14", "TYROBP")
)

dat_fil <- AddModuleScore(
  object = dat_fil, features = genes_to_score,
  name = "cell_scores"
)

# rename the modules (basically copy to new column..)
names(dat_fil[[]])[names(dat_fil[[]]) == 'cell_scores1'] <- "ec_score"
names(dat_fil[[]])[names(dat_fil[[]]) == 'cell_scores2'] <- "malignant_score"
names(dat_fil[[]])[names(dat_fil[[]]) == 'cell_scores3'] <- "glial_score"
names(dat_fil[[]])[names(dat_fil[[]]) == 'cell_scores4'] <- "fibroblast_score"
names(dat_fil[[]])[names(dat_fil[[]]) == 'cell_scores5'] <- "bcell_score"
names(dat_fil[[]])[names(dat_fil[[]]) == 'cell_scores6'] <- "tcell_score"
names(dat_fil[[]])[names(dat_fil[[]]) == 'cell_scores7'] <- "myeloid_score"
names(dat_fil[[]])[names(dat_fil[[]]) == 'cell_scores8'] <- "hev_ec_score"
names(dat_fil[[]])[names(dat_fil[[]]) == 'cell_scores9'] <- "macro_score"
```


After scoring the modules, we can look at a variety of different biology of interest,
including those that are derived from previous publication, e.g., from [this paper](https://doi.org/10.1038/s43018-024-00807-z).

### Co-localisation of different cell types

For example, the paper found the presence of rare endothelial cells (HEV-CXCL10) expressing HEV marker (SELP)
and CXCL9/10/11 that play important role in recruiting T cells.
By plotting the module scores for Endothelial cells (`ec_score`), HEV-CXCL10 endothelial cells
(`hev_ec_score`), and T cells (`tcell_score`), we can see whether there exists spots
in this CRC tissue are have high score for all modules or whether there are spots
with high `hev_ec_score` next to spots with high `tcell_score`.
The former will indicate a spot where there may be a mix of T cells and HEV-CXCL10 endothelial cells
while the former will indicate there are regions where they are close to one another.

```{r}
SpatialFeaturePlot(
  dat_fil,
  features = c(
    "ec_score", "hev_ec_score", "tcell_score"
  ),
  pt.size.factor = pt_size,
  image.alpha = 0.3
)
```

We do have some overlap on the top left and middle right. 
Having generic endothelial module ensure that the HEV-CXCL10 subsets
lie on the appropriate region (where endothelial cells are).

The paper also discovered myeloid regions neighbouring fibroblast regions which 
suggests positive correlation between macrophages and fibroblast in tumour samples.
Also positive correlation between macro-ISG15 and macro-SPP1 with tumour cells.

```{r}
SpatialFeaturePlot(
  dat_fil,
  features = c(
    "malignant_score", "fibroblast_score", "macro_score", "ISG15", "SPP1"
  ),
  pt.size.factor = pt_size,
  image.alpha = 0.3
)
```

### Inferring cell-cell communications

SIRPA and CD47 is a common receptor ligand pair between macrophages and tumour cells.
CD47 ligand is commonly sent out by tumour cells to tell macrophages to not eat them.
This is quite a well known immune mechanism employed by tumour cells.

```{r}
SpatialFeaturePlot(
  dat_fil,
  features = c(
    "macro_score", "malignant_score", "SIRPA", "CD47"
  ),
  ncol = 2,
  pt.size.factor = pt_size,
  image.alpha = 0.3
)
```

From the plot, we can infer the following:

1. There are regions where macrophages and tumour cells co-exist (top right)
where there are high SIRPA and CD47 expression.
As the SIRPA signal is highly expressed, macrophages' inhibitory receptor is activated, 
signalling it against phagocytosis.
2. There are regions in middle bottom (the strip) where there are tumour cells with
low expression of CD47, and macrophages with low expression of SIRPA.
This is the region where phagocytosis might happen because the tumour cells are
lacking the CD47 ligand.

### PCA and UMAP

Spatial plot may not be optimal to show regions where different cell types colocalised in a spot.
To help with this, we can draw a UMAP plot coloured by the module scores we have 
previously calculated.
Spots (one dot in a UMAP) with high score for two modules may indicate colocalisation
of two different subsets.

```{r}
dat_fil <- RunPCA(dat_fil, assay='SCT')

# just to pick how many PCs to use to calculate UMAP.
ElbowPlot(dat_fil, ndims=50)

dat_fil <- RunUMAP(dat_fil, reduction = 'pca', dims=1:20)
```

```{r}
FeaturePlot(
  dat_fil, 
  features = c(
    "macro_score", "malignant_score", "SIRPA", "CD47"
  ),
  cols = viridis::viridis(50)
)
```

## Extra stuff

Maybe spatially variable genes? and deconvolution - if there is a nice single cell data we can use.
Maybe the h5ad cancer data from [this paper](https://www.nature.com/articles/s43018-024-00807-z) can be used.
Also merging two tissues?









